<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.251">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>DeepMoA: method to predict the mechanism of action of cancer drugs</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@*/dist/embed-amd.js" crossorigin="anonymous"></script>


</head>

<body class="fullcontent">

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">DeepMoA: method to predict the mechanism of action of cancer drugs</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="select-data-and-import-libraries" class="level1">
<h1>Select data and import libraries</h1>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys <span class="co"># we require code from other folders</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> itertools</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>os.environ[<span class="st">'KMP_DUPLICATE_LIB_OK'</span>]<span class="op">=</span><span class="st">'True'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>CB_color_cycle <span class="op">=</span> [<span class="st">'#EECC16'</span>, <span class="st">'#62BB35'</span>, <span class="st">'#FDAE33'</span>,<span class="st">'#208EA3'</span>, <span class="st">'#EA4E9D'</span>, <span class="st">'#984ea3'</span>,<span class="st">'#999999'</span>, <span class="st">'#e41a1c'</span>, <span class="st">'#dede00'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.font_manager <span class="im">as</span> fm</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>font_files <span class="op">=</span> fm.findSystemFonts()</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>plt.rcdefaults()</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Go through and add each to Matplotlib's font cache.</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> font_file <span class="kw">in</span> font_files:</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    fm.fontManager.addfont(font_file)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>plt.rc(<span class="st">'font'</span>, family<span class="op">=</span><span class="st">'Roboto'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="61">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>plt.rc(<span class="st">'font'</span>, family<span class="op">=</span><span class="st">'Roboto'</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'font.family'</span>] <span class="op">=</span> <span class="st">'sans-serif'</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'font.sans-serif'</span>] <span class="op">=</span> <span class="st">'Roboto'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="62">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>config InlineBackend.figure_format<span class="op">=</span><span class="st">'retina'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="63">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pytorch relates imports</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.nn <span class="im">as</span> nn</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.optim <span class="im">as</span> optim</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># imports from captum library</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> captum.attr <span class="im">import</span> LayerDeepLift</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="64">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for combobox</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> __future__ <span class="im">import</span> print_function</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> ipywidgets <span class="im">import</span> interact, interactive, fixed, interact_manual</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> ipywidgets <span class="im">as</span> widgets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="65">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.express <span class="im">as</span> px</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.graph_objects <span class="im">as</span> go</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.io <span class="im">as</span> pio</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="66">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>pd.options.display.min_rows <span class="op">=</span> <span class="dv">20000</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>pd.set_option(<span class="st">'max_colwidth'</span>, <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="67">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>pd.options.display.max_rows <span class="op">=</span> <span class="dv">20000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="68">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>pd.set_option(<span class="st">'min_rows'</span>, <span class="dv">20000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>mac <span class="op">=</span> <span class="st">"/Users/katyna/Library/CloudStorage/OneDrive-Tecnun/"</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>windows <span class="op">=</span> <span class="st">"C:/Users/ksada/OneDrive - Tecnun/"</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>computer <span class="op">=</span> mac <span class="co"># CHANGE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>sys.path.append(computer <span class="op">+</span> <span class="st">"SparseGO_code/code"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> util</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> util <span class="im">import</span> <span class="op">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To make histograms</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> histogram(dataframe, color, title, ylabel,n_bins):</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    N, bins, patches <span class="op">=</span> plt.hist(dataframe, color<span class="op">=</span>color,bins<span class="op">=</span>n_bins, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(bins)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> bins[i]<span class="op">&lt;</span><span class="fl">0.05</span>:</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>            patches[i].set_facecolor(CB_color_cycle[<span class="dv">2</span>])</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">"P-value"</span>, fontsize<span class="op">=</span><span class="dv">16</span>)  </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(ylabel, fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    plt.title(title, fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    plt.xticks(fontsize<span class="op">=</span><span class="dv">14</span>)  </span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    plt.yticks(fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> plt.subplot(<span class="dv">111</span>)  </span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)  </span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>inputdir <span class="op">=</span> computer<span class="op">+</span><span class="st">"SparseGO_code/data/train_sparseGO/allsamples/"</span> <span class="co"># CHANGE</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>dir1<span class="op">=</span>computer<span class="op">+</span><span class="st">"Tesis/Codigo/VariableImportance/"</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>dir2<span class="op">=</span>computer<span class="op">+</span><span class="st">"SparseGO_code/results/weights&amp;biases/SparseGO_Standard_Expression/samples1/"</span> <span class="co"># CHANGE</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>resultsdir<span class="op">=</span>dir2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>gene2id <span class="op">=</span> inputdir<span class="op">+</span><span class="st">"gene2ind.txt"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>cell2id<span class="op">=</span>inputdir<span class="op">+</span><span class="st">"cell2ind.txt"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>drug2id<span class="op">=</span>inputdir<span class="op">+</span><span class="st">"drug2ind.txt"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>drug2fingerprint<span class="op">=</span>inputdir<span class="op">+</span><span class="st">"drug2fingerprint.txt"</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>load<span class="op">=</span>resultsdir<span class="op">+</span><span class="st">"last_model.pt"</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>onto <span class="op">=</span> inputdir<span class="op">+</span><span class="st">"sparseGO_ont.txt"</span>  <span class="co"># CHANGE </span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>genotype<span class="op">=</span>inputdir<span class="op">+</span><span class="st">"cell2expression.txt"</span>  <span class="co"># CHANGE </span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>num_neurons_per_GO <span class="op">=</span> <span class="dv">6</span> <span class="co"># CHANGE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="deeplift" class="level1">
<h1>DeepLIFT</h1>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>gene2id_mapping <span class="op">=</span> load_mapping(gene2id)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>dG, terms_pairs, genes_terms_pairs <span class="op">=</span> load_ontology(onto, gene2id_mapping)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>sorted_pairs, level_list, level_number <span class="op">=</span> sort_pairs(genes_terms_pairs, terms_pairs, dG, gene2id_mapping)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>layer_connections <span class="op">=</span> pairs_in_layers(sorted_pairs, level_list, level_number) </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>cell_features <span class="op">=</span> np.genfromtxt(genotype, delimiter<span class="op">=</span><span class="st">','</span>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>drug_features <span class="op">=</span> np.genfromtxt(drug2fingerprint, delimiter<span class="op">=</span><span class="st">','</span>)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>drug2id_mapping <span class="op">=</span> load_mapping(drug2id)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>cell2id_mapping <span class="op">=</span> load_mapping(cell2id)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>num_genes <span class="op">=</span> <span class="bu">len</span>(gene2id_mapping)</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>drug_dim <span class="op">=</span> <span class="bu">len</span>(drug_features[<span class="dv">0</span>,:])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>There are 14457 genes
There are 1 roots: GO:0008150
There are 4840 terms
There are 1 connected components</code></pre>
</div>
</div>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#model = torch.load(load, map_location='cuda:%d' % 0)</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> torch.load(load, map_location<span class="op">=</span>torch.device(<span class="st">'cpu'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="14">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>sparseGO_nn(
  (genes_terms_sparse_linear_1): SparseLinearNew(
    in_features=14457, out_features=29040, bias=True, sparsity=0.001853544595343158, connectivity=tensor([[ 3102,  3103,  3104,  ..., 24099, 24100, 24101],
            [    0,     0,     0,  ..., 14456, 14456, 14456]]), small_world=False
  )
  (genes_terms_batchnorm): BatchNorm1d(14457, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (genes_terms_tanh): Tanh()
  (drop_0): Dropout(p=0.15, inplace=False)
  (GO_terms_sparse_linear_1): SparseLinearNew(
    in_features=29040, out_features=13656, bias=True, sparsity=0.0010022839837906142, connectivity=tensor([[    0,     1,     2,  ...,  5367,  5368,  5369],
            [    0,     0,     0,  ..., 29039, 29039, 29039]]), small_world=False
  )
  (drop_1): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_1): Tanh()
  (GO_terms_batchnorm_1): BatchNorm1d(29040, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (GO_terms_sparse_linear_2): SparseLinearNew(
    in_features=13656, out_features=7806, bias=True, sparsity=0.0013316105361699598, connectivity=tensor([[ 1458,  1459,  1460,  ...,  7803,  7804,  7805],
            [    0,     0,     0,  ..., 13655, 13655, 13655]]), small_world=False
  )
  (drop_2): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_2): Tanh()
  (GO_terms_batchnorm_2): BatchNorm1d(13656, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (GO_terms_sparse_linear_3): SparseLinearNew(
    in_features=7806, out_features=4644, bias=True, sparsity=0.0021370958932405404, connectivity=tensor([[ 456,  457,  458,  ..., 4563, 4564, 4565],
            [   0,    0,    0,  ..., 7805, 7805, 7805]]), small_world=False
  )
  (drop_3): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_3): Tanh()
  (GO_terms_batchnorm_3): BatchNorm1d(7806, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (GO_terms_sparse_linear_4): SparseLinearNew(
    in_features=4644, out_features=2640, bias=True, sparsity=0.0034971811134601833, connectivity=tensor([[  12,   13,   14,  ..., 2547, 2548, 2549],
            [   0,    0,    0,  ..., 4643, 4643, 4643]]), small_world=False
  )
  (drop_4): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_4): Tanh()
  (GO_terms_batchnorm_4): BatchNorm1d(4644, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (GO_terms_sparse_linear_5): SparseLinearNew(
    in_features=2640, out_features=1416, bias=True, sparsity=0.00614406779661017, connectivity=tensor([[ 732,  733,  734,  ..., 1413, 1414, 1415],
            [   0,    0,    0,  ..., 2639, 2639, 2639]]), small_world=False
  )
  (drop_5): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_5): Tanh()
  (GO_terms_batchnorm_5): BatchNorm1d(2640, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (GO_terms_sparse_linear_6): SparseLinearNew(
    in_features=1416, out_features=702, bias=True, sparsity=0.012856728958423875, connectivity=tensor([[  90,   91,   92,  ...,  579,  580,  581],
            [   0,    0,    0,  ..., 1415, 1415, 1415]]), small_world=False
  )
  (drop_6): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_6): Tanh()
  (GO_terms_batchnorm_6): BatchNorm1d(1416, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (GO_terms_sparse_linear_7): SparseLinearNew(
    in_features=702, out_features=288, bias=True, sparsity=0.02938034188034188, connectivity=tensor([[  0,   1,   2,  ..., 177, 178, 179],
            [  0,   0,   0,  ..., 701, 701, 701]]), small_world=False
  )
  (drop_7): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_7): Tanh()
  (GO_terms_batchnorm_7): BatchNorm1d(702, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (GO_terms_sparse_linear_8): SparseLinearNew(
    in_features=288, out_features=30, bias=True, sparsity=1.0, connectivity=tensor([[  0,   1,   2,  ...,  27,  28,  29],
            [  0,   0,   0,  ..., 287, 287, 287]]), small_world=False
  )
  (drop_8): Dropout(p=0.15, inplace=False)
  (GO_terms_tanh_8): Tanh()
  (GO_terms_batchnorm_8): BatchNorm1d(288, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (drug_linear_layer_1): Linear(in_features=2048, out_features=200, bias=True)
  (drug_drop_1): Dropout(p=0.15, inplace=False)
  (drug_tanh_1): Tanh()
  (drug_batchnorm_layer_1): BatchNorm1d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (drug_linear_layer_2): Linear(in_features=200, out_features=100, bias=True)
  (drug_drop_2): Dropout(p=0.15, inplace=False)
  (drug_tanh_2): Tanh()
  (drug_batchnorm_layer_2): BatchNorm1d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (drug_linear_layer_3): Linear(in_features=100, out_features=50, bias=True)
  (drug_drop_3): Dropout(p=0.15, inplace=False)
  (drug_tanh_3): Tanh()
  (drug_batchnorm_layer_3): BatchNorm1d(100, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (final_batchnorm_layer): BatchNorm1d(80, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (drop_final): Dropout(p=0, inplace=False)
  (final_linear_layer): Linear(in_features=80, out_features=40, bias=True)
  (final_tanh): Tanh()
  (final_aux_batchnorm_layer): BatchNorm1d(40, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (drop_aux_final): Dropout(p=0, inplace=False)
  (final_aux_linear_layer): Linear(in_features=40, out_features=1, bias=True)
  (final_aux_tanh): Tanh()
  (final_linear_layer_output): Linear(in_features=1, out_features=1, bias=True)
)</code></pre>
</div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="15">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save layers to be analyzed</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>model_layers <span class="op">=</span> []</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.genes_terms_sparse_linear_1)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.GO_terms_sparse_linear_1)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.GO_terms_sparse_linear_2)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.GO_terms_sparse_linear_3)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.GO_terms_sparse_linear_4)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.GO_terms_sparse_linear_5)</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.GO_terms_sparse_linear_6)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>model_layers.append(model.GO_terms_sparse_linear_7)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="go-terms-info" class="level2">
<h2 class="anchored" data-anchor-id="go-terms-info">GO terms info</h2>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Go term names</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>gene_ontology <span class="op">=</span> pd.read_excel(<span class="st">'all_go_terms_info.xlsx'</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>gene_ontology <span class="op">=</span> gene_ontology.drop_duplicates(subset<span class="op">=</span>[<span class="st">'id'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get all layers’ GO term with the neuron number</p>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>all_terms_ids <span class="op">=</span> {}</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>all_terms_names <span class="op">=</span> {}</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>all_layers_non_virtual <span class="op">=</span> {} <span class="co"># store only terms that are part of the layer (remove virtual), those are the important attribuitions</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>all_layers_non_virtual_names <span class="op">=</span> {}</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>num_neurons_per_GO <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> layer_number <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(layer_connections)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    layer_pairs <span class="op">=</span> layer_connections[layer_number] </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>    terms_ids <span class="op">=</span> []</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    names <span class="op">=</span> []</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    output_id <span class="op">=</span> create_index(layer_pairs[:,<span class="dv">0</span>]) <span class="co"># first 6 neurons correspond to the term with key 0</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># all_terms_ids IS CREATED IN THE SAME GO TERM ORDER IN WHICH THE CORRESPONDING NETWORK LAYER WAS CREATED (when calculating the XAI score, this lets us know which GO term it corresponds to).</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> term <span class="kw">in</span> output_id.keys():</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>        <span class="co">#name = gene_ontology.loc[gene_ontology['GO_term'] == term].to_numpy()[0,3].replace("_"," ").capitalize()</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>        name <span class="op">=</span> gene_ontology.loc[gene_ontology[<span class="st">'id'</span>] <span class="op">==</span> term].to_numpy()[<span class="dv">0</span>,<span class="dv">1</span>].capitalize()</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,num_neurons_per_GO<span class="op">+</span><span class="dv">1</span>): <span class="co"># vector que tiene GO:0000038_1, GO:0000038_2 ... GO:0000038_6 y asi luego concatenar con las attributions</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>            terms_ids.append(term<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(i))</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>            names.append(name<span class="op">+</span><span class="st">" ("</span><span class="op">+</span><span class="bu">str</span>(i)<span class="op">+</span><span class="st">")"</span>)</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>    all_terms_ids[layer_number] <span class="op">=</span> np.array(terms_ids)</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>    all_terms_names[layer_number] <span class="op">=</span> np.array(names)</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>    non_virtual <span class="op">=</span> [] <span class="co"># store the terms part of that layer</span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>    non_virtual_names <span class="op">=</span> []</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> term <span class="kw">in</span> level_list[layer_number<span class="op">+</span><span class="dv">1</span>]:</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>        nv_name <span class="op">=</span> gene_ontology.loc[gene_ontology[<span class="st">'id'</span>] <span class="op">==</span> term].to_numpy()[<span class="dv">0</span>,<span class="dv">1</span>].capitalize()</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">7</span>):</span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>            non_virtual.append(term<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(i))</span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>            non_virtual_names.append(nv_name<span class="op">+</span><span class="st">" ("</span><span class="op">+</span><span class="bu">str</span>(i)<span class="op">+</span><span class="st">")"</span>)</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>    all_layers_non_virtual[layer_number] <span class="op">=</span> non_virtual</span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>    all_layers_non_virtual_names[layer_number] <span class="op">=</span> non_virtual_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All GO terms part of a layer (non-virtual) with their corresponding name and layer number…</p>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>real_go_info <span class="op">=</span> pd.DataFrame({<span class="st">"GO_term"</span>:[],<span class="st">"Name"</span>:[],<span class="st">"layer_number"</span>:[]})</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> layer_number <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(layer_connections)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    layer_go_info <span class="op">=</span> pd.DataFrame({<span class="st">"GO_term"</span>:all_layers_non_virtual[layer_number],<span class="st">"Name"</span>:all_layers_non_virtual_names[layer_number],<span class="st">"layer_number"</span>:(layer_number)})</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>    real_go_info <span class="op">=</span> pd.concat((real_go_info,layer_go_info))</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>real_go_info.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>Name</th>
      <th>layer_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>GO:0006264_1</td>
      <td>Mitochondrial dna replication (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GO:0006264_2</td>
      <td>Mitochondrial dna replication (2)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GO:0006264_3</td>
      <td>Mitochondrial dna replication (3)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>GO:0006264_4</td>
      <td>Mitochondrial dna replication (4)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>GO:0006264_5</td>
      <td>Mitochondrial dna replication (5)</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="drugs-info" class="level2">
<h2 class="anchored" data-anchor-id="drugs-info">Drugs info</h2>
<div class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_compound_names(file_name):</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    compounds <span class="op">=</span> []</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(file_name, <span class="st">'r'</span>) <span class="im">as</span> fi:</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> fi:</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>            tokens <span class="op">=</span> line.strip().split(<span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>            compounds.append([tokens[<span class="dv">1</span>],tokens[<span class="dv">2</span>]])</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> compounds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>drugs <span class="op">=</span> get_compound_names(inputdir<span class="op">+</span><span class="st">"compound_names.txt"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>drugs.pop(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<pre><code>['SMILE', 'Name']</code></pre>
</div>
</div>
</section>
<section id="deeplift-for-vnn" class="level2">
<h2 class="anchored" data-anchor-id="deeplift-for-vnn">DeepLIFT for VNN</h2>
<p><em>Reference activation…</em> (baseline)</p>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>median_cell_features <span class="op">=</span> np.median(cell_features,axis<span class="op">=</span><span class="dv">0</span>) <span class="co"># to use as a reference</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>median_drug_features <span class="op">=</span> np.genfromtxt(computer<span class="op">+</span><span class="st">"SparseGO_code/data/glucose_fingerprint.txt"</span>, delimiter<span class="op">=</span><span class="st">','</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="attribution-function-sum" class="level3">
<h3 class="anchored" data-anchor-id="attribution-function-sum">Attribution function: sum</h3>
<div class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_layer_attribution(layer_number,input_data,baseline,selected_drug_data):</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    dl <span class="op">=</span> LayerDeepLift(model, model_layers[layer_number],multiply_by_inputs <span class="op">=</span> <span class="va">True</span>) <span class="co"># CHOOSE LAYER TO STUDY</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    dl_attr_test <span class="op">=</span> dl.attribute(input_data,baseline)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    dl_attr_test_sum <span class="op">=</span> dl_attr_test.cpu().detach().numpy().<span class="bu">sum</span>(<span class="dv">0</span>) <span class="co"># the attributions are summed for each sample</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># all_terms_ids WAS CREATED IN THE SAME GO TERM ORDER IN WHICH THE CORRESPONDING NETWORK LAYER WAS CREATED (when calculating the XAI score, this lets us know which GO term it corresponds to).</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>    attribution_data <span class="op">=</span> pd.DataFrame(np.column_stack((all_terms_ids[layer_number],dl_attr_test_sum)), columns<span class="op">=</span>[<span class="st">"GO_term"</span>,selected_drug_data[<span class="dv">1</span>]])</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>    attribution_data[[selected_drug_data[<span class="dv">1</span>]]] <span class="op">=</span> attribution_data[[selected_drug_data[<span class="dv">1</span>]]].<span class="bu">apply</span>(pd.to_numeric).<span class="bu">round</span>(<span class="dv">10</span>)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>    attribution_data <span class="op">=</span> attribution_data.loc[attribution_data[<span class="st">'GO_term'</span>].isin(all_layers_non_virtual[layer_number])] <span class="co"># only the keep the non virtual terms</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> attribution_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="deeplift-for-all-drugs" class="level3">
<h3 class="anchored" data-anchor-id="deeplift-for-all-drugs">DeepLIFT for all drugs</h3>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all <span class="op">=</span> pd.DataFrame()</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain the top GO terms on all layers for each drug</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> selected_drug_data <span class="kw">in</span> drugs:</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>    selected_drug <span class="op">=</span>selected_drug_data[<span class="dv">0</span>] <span class="co"># DRUG smile</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    selected_drug_features <span class="op">=</span> []</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>    drug_specific_features<span class="op">=</span>drug_features[drug2id_mapping[selected_drug]] <span class="co"># features of drug</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(cell2id_mapping)): <span class="co"># make all combinations of selected_drug and cell types </span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>        selected_drug_features.append(np.concatenate((cell_features[i], drug_specific_features), axis<span class="op">=</span><span class="va">None</span>))</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    selected_drug_features <span class="op">=</span> torch.FloatTensor(np.array(selected_drug_features))</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Data for deeplift...</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>    input_data <span class="op">=</span> torch.autograd.Variable(selected_drug_features.cuda(<span class="dv">0</span>))</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># baseline is the median of the expression data and drug features </span></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>    baseline <span class="op">=</span> torch.FloatTensor(np.concatenate((median_cell_features, median_drug_features), axis<span class="op">=</span><span class="va">None</span>))</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>    baseline <span class="op">=</span> torch.reshape(baseline, (<span class="dv">1</span>, baseline.size()[<span class="dv">0</span>]))</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>    baseline <span class="op">=</span> torch.autograd.Variable(baseline.cuda(<span class="dv">0</span>))</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>    attribution_data_drug <span class="op">=</span> <span class="bu">list</span>(<span class="bu">map</span>(get_layer_attribution,<span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(model_layers)),itertools.repeat(input_data, <span class="bu">len</span>(model_layers)),itertools.repeat(baseline, <span class="bu">len</span>(model_layers)),itertools.repeat(selected_drug_data, <span class="bu">len</span>(model_layers)))) <span class="co"># get the attribution for each layer (map is similar to apply)</span></span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>    attribution_data_drug <span class="op">=</span> pd.concat(attribution_data_drug) <span class="co"># concatenate attribution of all layers</span></span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>    attribution_data_all <span class="op">=</span> pd.concat([attribution_data_all,attribution_data_drug.iloc[:,<span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(selected_drug_data[<span class="dv">1</span>])</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>attribution_data_all <span class="op">=</span> pd.concat([attribution_data_drug.iloc[:,<span class="dv">0</span>],attribution_data_all], axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all <span class="op">=</span> attribution_data_all.set_index(<span class="st">"GO_term"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="24">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>(-)-Parthenolide</th>
      <th>(5Z)-7-Oxozeaenol</th>
      <th>5-Fluorouracil</th>
      <th>681640</th>
      <th>A-443654</th>
      <th>A-484954</th>
      <th>A-770041</th>
      <th>A-83-01</th>
      <th>Acadesine</th>
      <th>ACY-1215</th>
      <th>...</th>
      <th>Vemurafenib</th>
      <th>VER-155008</th>
      <th>Vorapaxar</th>
      <th>vorinostat:carboplatin (1:1 mol/mol)</th>
      <th>vorinostat:navitoclax (4:1 mol/mol)</th>
      <th>VU0155056</th>
      <th>WZ4002</th>
      <th>WZ8040</th>
      <th>YL54</th>
      <th>Zebularine</th>
    </tr>
    <tr>
      <th>GO_term</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:0000012_1</th>
      <td>0.009143</td>
      <td>0.012238</td>
      <td>0.014801</td>
      <td>0.008749</td>
      <td>0.028462</td>
      <td>0.002892</td>
      <td>0.021746</td>
      <td>0.003400</td>
      <td>0.002582</td>
      <td>0.015812</td>
      <td>...</td>
      <td>0.010666</td>
      <td>0.008861</td>
      <td>0.014170</td>
      <td>0.006866</td>
      <td>0.016116</td>
      <td>0.010692</td>
      <td>0.023399</td>
      <td>0.022375</td>
      <td>0.010787</td>
      <td>0.007544</td>
    </tr>
    <tr>
      <th>GO:0000012_2</th>
      <td>-0.000285</td>
      <td>-0.000928</td>
      <td>-0.000579</td>
      <td>0.000208</td>
      <td>-0.001000</td>
      <td>-0.000165</td>
      <td>-0.002064</td>
      <td>-0.000199</td>
      <td>-0.000089</td>
      <td>-0.000651</td>
      <td>...</td>
      <td>-0.001053</td>
      <td>-0.000468</td>
      <td>-0.000831</td>
      <td>-0.000112</td>
      <td>-0.000545</td>
      <td>-0.000622</td>
      <td>-0.001353</td>
      <td>-0.001063</td>
      <td>-0.000853</td>
      <td>-0.000307</td>
    </tr>
    <tr>
      <th>GO:0000012_3</th>
      <td>0.000593</td>
      <td>0.001505</td>
      <td>0.003211</td>
      <td>0.000860</td>
      <td>0.002106</td>
      <td>-0.000137</td>
      <td>0.003661</td>
      <td>-0.000340</td>
      <td>-0.000157</td>
      <td>0.002306</td>
      <td>...</td>
      <td>0.000033</td>
      <td>0.001873</td>
      <td>-0.000379</td>
      <td>0.001317</td>
      <td>-0.000275</td>
      <td>0.000724</td>
      <td>0.003970</td>
      <td>0.005052</td>
      <td>-0.000554</td>
      <td>0.000269</td>
    </tr>
    <tr>
      <th>GO:0000012_4</th>
      <td>-0.000407</td>
      <td>-0.000310</td>
      <td>-0.000145</td>
      <td>-0.000331</td>
      <td>-0.000294</td>
      <td>-0.000306</td>
      <td>-0.000686</td>
      <td>-0.000259</td>
      <td>-0.000302</td>
      <td>-0.000689</td>
      <td>...</td>
      <td>-0.000542</td>
      <td>-0.000382</td>
      <td>0.000087</td>
      <td>-0.000564</td>
      <td>-0.002327</td>
      <td>-0.000219</td>
      <td>0.000312</td>
      <td>0.000633</td>
      <td>-0.000277</td>
      <td>-0.000085</td>
    </tr>
    <tr>
      <th>GO:0000012_5</th>
      <td>-0.000152</td>
      <td>-0.000454</td>
      <td>-0.000350</td>
      <td>0.000121</td>
      <td>-0.000196</td>
      <td>-0.000090</td>
      <td>-0.000926</td>
      <td>-0.000056</td>
      <td>-0.000057</td>
      <td>-0.000293</td>
      <td>...</td>
      <td>-0.000339</td>
      <td>-0.000374</td>
      <td>-0.000076</td>
      <td>-0.000136</td>
      <td>-0.000653</td>
      <td>-0.000179</td>
      <td>-0.000458</td>
      <td>-0.000314</td>
      <td>-0.000226</td>
      <td>-0.000019</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 790 columns</p>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="chembl-drug-target-slim" class="level1">
<h1>ChEMBL Drug Target Slim</h1>
<div class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> chembl_webresource_client.new_client <span class="im">import</span> new_client</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Import SparseGO drugs</p>
<div class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get names </span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_compound_names(file_name):</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    compounds <span class="op">=</span> []</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(file_name, <span class="st">'r'</span>) <span class="im">as</span> fi:</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> fi:</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>            tokens <span class="op">=</span> line.strip().split(<span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>            compounds.append(tokens[<span class="dv">2</span>].lower())</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> compounds</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>names <span class="op">=</span> get_compound_names(computer<span class="op">+</span><span class="st">"SparseGO_code/data/compound_names.txt"</span>)</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>names.pop(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">
<pre><code>'name'</code></pre>
</div>
</div>
<section id="cheml-ids" class="level2">
<h2 class="anchored" data-anchor-id="cheml-ids">chEML IDs</h2>
<p>Get chembl IDs of drugs if available (there are always 684 drugs, the compounds2ids object can be reused)</p>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all chembl IDs -- takes long </span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>molecule <span class="op">=</span> new_client.molecule</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>compounds2ids <span class="op">=</span> {}</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i,drug <span class="kw">in</span> <span class="bu">enumerate</span>(names):</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">" + "</span> <span class="kw">in</span> drug:</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>        drug_split <span class="op">=</span> drug.split(<span class="st">" + "</span>, <span class="dv">1</span>)</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>        ID1 <span class="op">=</span> <span class="bu">list</span>(molecule.<span class="bu">filter</span>(pref_name__iexact<span class="op">=</span>drug_split[<span class="dv">0</span>]).only(<span class="st">'molecule_chembl_id'</span>))</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>        ID2 <span class="op">=</span> <span class="bu">list</span>(molecule.<span class="bu">filter</span>(pref_name__iexact<span class="op">=</span>drug_split[<span class="dv">1</span>]).only(<span class="st">'molecule_chembl_id'</span>))</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(ID1)<span class="op">&gt;</span><span class="dv">0</span> <span class="kw">and</span> <span class="bu">len</span>(ID2)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>            compounds2ids[drug]<span class="op">=</span>[ID1[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>],ID2[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>]]</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> <span class="bu">len</span>(ID1)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>            compounds2ids[drug]<span class="op">=</span>ID1[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>] </span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> <span class="bu">len</span>(ID2)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>            compounds2ids[drug]<span class="op">=</span>ID2[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>] </span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(drug,i)</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> <span class="st">":"</span> <span class="kw">in</span> drug: <span class="co"># combinations</span></span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>        drug_split <span class="op">=</span> drug.split(<span class="st">":"</span>, <span class="dv">1</span>)</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>        ID1 <span class="op">=</span> <span class="bu">list</span>(molecule.<span class="bu">filter</span>(pref_name__iexact<span class="op">=</span>drug_split[<span class="dv">0</span>]).only(<span class="st">'molecule_chembl_id'</span>))</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a>        ID2 <span class="op">=</span> <span class="bu">list</span>(molecule.<span class="bu">filter</span>(pref_name__iexact<span class="op">=</span>drug_split[<span class="dv">1</span>].split(<span class="st">" ("</span>, <span class="dv">1</span>)[<span class="dv">0</span>]).only(<span class="st">'molecule_chembl_id'</span>))</span>
<span id="cb38-24"><a href="#cb38-24" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(ID1)<span class="op">&gt;</span><span class="dv">0</span> <span class="kw">and</span> <span class="bu">len</span>(ID2)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-25"><a href="#cb38-25" aria-hidden="true" tabindex="-1"></a>            compounds2ids[drug]<span class="op">=</span>[ID1[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>],ID2[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>]]</span>
<span id="cb38-26"><a href="#cb38-26" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> <span class="bu">len</span>(ID1)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-27"><a href="#cb38-27" aria-hidden="true" tabindex="-1"></a>            compounds2ids[drug]<span class="op">=</span>ID1[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>] </span>
<span id="cb38-28"><a href="#cb38-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> <span class="bu">len</span>(ID2)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-29"><a href="#cb38-29" aria-hidden="true" tabindex="-1"></a>            compounds2ids[drug]<span class="op">=</span>ID2[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>] </span>
<span id="cb38-30"><a href="#cb38-30" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb38-31"><a href="#cb38-31" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(drug,i)</span>
<span id="cb38-32"><a href="#cb38-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-33"><a href="#cb38-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb38-34"><a href="#cb38-34" aria-hidden="true" tabindex="-1"></a>        ID <span class="op">=</span> <span class="bu">list</span>(molecule.<span class="bu">filter</span>(pref_name__iexact<span class="op">=</span>drug).only(<span class="st">'molecule_chembl_id'</span>))</span>
<span id="cb38-35"><a href="#cb38-35" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(ID)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-36"><a href="#cb38-36" aria-hidden="true" tabindex="-1"></a>            ID <span class="op">=</span> ID[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>]</span>
<span id="cb38-37"><a href="#cb38-37" aria-hidden="true" tabindex="-1"></a>            compounds2ids[drug]<span class="op">=</span>ID</span>
<span id="cb38-38"><a href="#cb38-38" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb38-39"><a href="#cb38-39" aria-hidden="true" tabindex="-1"></a>            <span class="co"># for drugs that have the chembl ID as the name!!</span></span>
<span id="cb38-40"><a href="#cb38-40" aria-hidden="true" tabindex="-1"></a>            ID <span class="op">=</span> <span class="bu">list</span>(molecule.<span class="bu">filter</span>(chembl_id<span class="op">=</span>drug).only(<span class="st">'molecule_chembl_id'</span>)) </span>
<span id="cb38-41"><a href="#cb38-41" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">len</span>(ID)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-42"><a href="#cb38-42" aria-hidden="true" tabindex="-1"></a>                ID <span class="op">=</span> ID[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>]</span>
<span id="cb38-43"><a href="#cb38-43" aria-hidden="true" tabindex="-1"></a>                compounds2ids[drug]<span class="op">=</span>ID</span>
<span id="cb38-44"><a href="#cb38-44" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb38-45"><a href="#cb38-45" aria-hidden="true" tabindex="-1"></a>                <span class="co"># in case it is not found by pref_name</span></span>
<span id="cb38-46"><a href="#cb38-46" aria-hidden="true" tabindex="-1"></a>                ID <span class="op">=</span> <span class="bu">list</span>(molecule.<span class="bu">filter</span>(molecule_synonyms__molecule_synonym__iexact<span class="op">=</span>drug).only(<span class="st">'molecule_chembl_id'</span>))</span>
<span id="cb38-47"><a href="#cb38-47" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="bu">len</span>(ID)<span class="op">&gt;</span><span class="dv">0</span>:</span>
<span id="cb38-48"><a href="#cb38-48" aria-hidden="true" tabindex="-1"></a>                    ID <span class="op">=</span> ID[<span class="dv">0</span>][<span class="st">'molecule_chembl_id'</span>]</span>
<span id="cb38-49"><a href="#cb38-49" aria-hidden="true" tabindex="-1"></a>                    compounds2ids[drug]<span class="op">=</span>ID</span>
<span id="cb38-50"><a href="#cb38-50" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb38-51"><a href="#cb38-51" aria-hidden="true" tabindex="-1"></a>                    <span class="bu">print</span>(drug,i)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for CTRPv2 and GDSC2</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"(-)-parthenolide"</span>]<span class="op">=</span><span class="st">"CHEMBL540445"</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"bleomycin a2"</span>]<span class="op">=</span><span class="st">"CHEMBL403664"</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"blebbistatin"</span>]<span class="op">=</span><span class="st">"CHEMBL1328324"</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"bax channel blocker"</span>]<span class="op">=</span><span class="st">"CHEMBL1402326"</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"am-580"</span>]<span class="op">=</span><span class="st">"CHEMBL69367"</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"bibr-1532"</span>]<span class="op">=</span><span class="st">"CHEMBL27323"</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"cyanoquinoline 11"</span>]<span class="op">=</span><span class="st">"CHEMBL436817"</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"etp-46464"</span>]<span class="op">=</span><span class="st">"CHEMBL3334621"</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"importazole"</span>]<span class="op">=</span><span class="st">"CHEMBL1498173"</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"isoevodiamine"</span>]<span class="op">=</span><span class="st">"CHEMBL81925"</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"aminofurazanyl-azabenzimidazole 6n"</span>]<span class="op">=</span><span class="st">"CHEMBL220241"</span></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"bms-345541 free base"</span>]<span class="op">=</span><span class="st">"CHEMBL249697"</span></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"mim1"</span>]<span class="op">=</span><span class="st">"CHEMBL3265288"</span></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"ml323"</span>]<span class="op">=</span><span class="st">"CHEMBL3182437"</span></span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>compounds2ids[<span class="st">"sabutoclax"</span>]<span class="op">=</span><span class="st">"CHEMBL1094250"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="188">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(compounds2ids)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="188">
<pre><code>411</code></pre>
</div>
</div>
</section>
<section id="chembl-moa-targets" class="level2">
<h2 class="anchored" data-anchor-id="chembl-moa-targets">chEMBL MoA (targets)</h2>
<p>Get the molecule targets of each drug (if available)</p>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>compounds2targets <span class="op">=</span> <span class="bu">dict</span>() <span class="co"># required to store the drug targets </span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> drug <span class="kw">in</span> compounds2ids.keys():</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    compounds2targets[drug] <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>chembl_ids <span class="op">=</span> <span class="bu">list</span>(compounds2ids.values()) <span class="co"># Chembl IDs of drugs</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> drug <span class="kw">in</span> compounds2ids:</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># we jump from compounds to targets through activities:</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    activities <span class="op">=</span> new_client.mechanism.<span class="bu">filter</span>(parent_molecule_chembl_id__in<span class="op">=</span>compounds2ids[drug]).only(</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>        [<span class="st">'parent_molecule_chembl_id'</span>, <span class="st">'target_chembl_id'</span>])</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># extracting target ChEMBL IDs from activities:</span></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> act <span class="kw">in</span> activities:</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>        compounds2targets[drug].add(act[<span class="st">'target_chembl_id'</span>])</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(drug)</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a><span class="co"># We now know all targets for some drug</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>compounds2targets <span class="op">=</span> {k: v <span class="cf">for</span> k, v <span class="kw">in</span> compounds2targets.items() <span class="cf">if</span> <span class="bu">len</span>(v) <span class="op">!=</span> <span class="dv">0</span> <span class="kw">and</span> <span class="bu">len</span>([x <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">list</span>(v) <span class="cf">if</span> x <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>]) <span class="op">!=</span> <span class="dv">0</span> }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="189">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(compounds2targets)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="189">
<pre><code>267</code></pre>
</div>
</div>
</section>
<section id="drug-slim-go-terms" class="level2">
<h2 class="anchored" data-anchor-id="drug-slim-go-terms">Drug slim GO terms</h2>
<p>Get the GO terms of each target</p>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>compounds_GOterms <span class="op">=</span> {}</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(compounds2targets.keys())):</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    compound <span class="op">=</span> <span class="bu">list</span>(compounds2targets.keys())[i]</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    GOterms_list <span class="op">=</span> []</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(<span class="bu">list</span>(compounds2targets[compound]))):   </span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>        target <span class="op">=</span> <span class="bu">list</span>(compounds2targets[compound])[j]</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>        all_cross_references <span class="op">=</span> <span class="bu">list</span>(new_client.target.<span class="bu">filter</span>(target_chembl_id<span class="op">=</span>target).only([<span class="st">'target_components'</span>]).only([<span class="st">'target_components_xrefs'</span>]))[<span class="dv">0</span>][<span class="st">'target_components'</span>]</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(all_cross_references)<span class="op">&gt;</span><span class="dv">0</span>: <span class="co"># not all targets have annotated go_terms</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(all_cross_references)):</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>                GOterms <span class="op">=</span> all_cross_references[i]</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>                GOterms <span class="op">=</span> pd.DataFrame(GOterms[<span class="st">'target_component_xrefs'</span>])</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a>                GOterms <span class="op">=</span> pd.concat([GOterms,pd.Series([target]).repeat(<span class="bu">len</span>(GOterms)).reset_index().pop(<span class="dv">0</span>)],axis<span class="op">=</span><span class="dv">1</span>) <span class="co"># add target ID to dataframe </span></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>                GOterms_list<span class="op">=</span> GOterms_list <span class="op">+</span> GOterms.values.tolist()</span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>    compounds_GOterms[compound] <span class="op">=</span>  pd.DataFrame(GOterms_list).drop_duplicates()</span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(compound)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="190">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(compounds_GOterms)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="190">
<pre><code>270</code></pre>
</div>
</div>
<p>Add GO terms found in CTRPv2</p>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>CTRPv2_terms <span class="op">=</span> pd.read_excel(<span class="st">'ctrp_goterms_drugs.xlsx'</span>)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add GO terms of drugs with or without annotations</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> drug <span class="kw">in</span> CTRPv2_terms[<span class="st">"Drug"</span>].unique():</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> drug <span class="kw">not</span> <span class="kw">in</span> <span class="bu">list</span>(compounds_GOterms.keys()): <span class="co"># some drugs had no previous data, no annotations from chembl</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>        compounds_GOterms[drug] <span class="op">=</span> pd.DataFrame() <span class="co"># create empty dataframe</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> term <span class="kw">in</span> <span class="bu">list</span>(CTRPv2_terms.loc[CTRPv2_terms[<span class="st">"Drug"</span>]<span class="op">==</span>drug][<span class="st">"Field"</span>]):</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>        compounds_GOterms[drug] <span class="op">=</span> pd.concat([compounds_GOterms[drug],pd.DataFrame([term,<span class="st">""</span>,<span class="st">"GoProcess"</span>,<span class="st">""</span>]).transpose()])</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>    compounds_GOterms[drug] <span class="op">=</span> compounds_GOterms[drug].drop_duplicates() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Delete drugs with no GOterms (some targets have no annotated GO terms)</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>compounds_GOterms <span class="op">=</span> {k: v <span class="cf">for</span> k, v <span class="kw">in</span> compounds_GOterms.items() <span class="cf">if</span> <span class="bu">len</span>(v) <span class="op">!=</span> <span class="dv">0</span> } </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="30">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(compounds_GOterms)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="30">
<pre><code>270</code></pre>
</div>
</div>
</section>
<section id="match-go-terms" class="level2">
<h2 class="anchored" data-anchor-id="match-go-terms">Match GO terms</h2>
<p>Find all terms that match, terms that are part of both, the sparseGO graph and the drug slim results…</p>
<div class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> load_ontology_extra_output(ontology_file, gene2id_mapping):</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Creates the directed graph of the GO terms and stores the connected elements in arrays.</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="co">        Output</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="co">        ------</span></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a><span class="co">        dG: networkx.classes.digraph.DiGraph</span></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="co">            Directed graph of all terms</span></span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a><span class="co">        terms_pairs: numpy.ndarray</span></span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a><span class="co">            Store the connection between a term and a term</span></span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a><span class="co">        genes_terms_pairs: numpy.ndarray</span></span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a><span class="co">            Store the connection between a gene and a term</span></span>
<span id="cb54-15"><a href="#cb54-15" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb54-16"><a href="#cb54-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-17"><a href="#cb54-17" aria-hidden="true" tabindex="-1"></a>    dG <span class="op">=</span> nx.DiGraph() <span class="co"># Directed graph class</span></span>
<span id="cb54-18"><a href="#cb54-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-19"><a href="#cb54-19" aria-hidden="true" tabindex="-1"></a>    file_handle <span class="op">=</span> <span class="bu">open</span>(ontology_file) <span class="co">#  Open the file that has genes and go terms</span></span>
<span id="cb54-20"><a href="#cb54-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-21"><a href="#cb54-21" aria-hidden="true" tabindex="-1"></a>    terms_pairs <span class="op">=</span> [] <span class="co"># store the pairs between a term and a term</span></span>
<span id="cb54-22"><a href="#cb54-22" aria-hidden="true" tabindex="-1"></a>    genes_terms_pairs <span class="op">=</span> [] <span class="co"># store the pairs between a gene and a term</span></span>
<span id="cb54-23"><a href="#cb54-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-24"><a href="#cb54-24" aria-hidden="true" tabindex="-1"></a>    gene_set <span class="op">=</span> <span class="bu">set</span>() <span class="co"># create a set (elements can't repeat)</span></span>
<span id="cb54-25"><a href="#cb54-25" aria-hidden="true" tabindex="-1"></a>    term_direct_gene_map <span class="op">=</span> {}</span>
<span id="cb54-26"><a href="#cb54-26" aria-hidden="true" tabindex="-1"></a>    term_size_map <span class="op">=</span> {}</span>
<span id="cb54-27"><a href="#cb54-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-28"><a href="#cb54-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-29"><a href="#cb54-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> file_handle:</span>
<span id="cb54-30"><a href="#cb54-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-31"><a href="#cb54-31" aria-hidden="true" tabindex="-1"></a>        line <span class="op">=</span> line.rstrip().split() <span class="co"># delete spaces and transform to list, line has 3 elements</span></span>
<span id="cb54-32"><a href="#cb54-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-33"><a href="#cb54-33" aria-hidden="true" tabindex="-1"></a>        <span class="co"># No me hace falta el if, no tengo que separar las parejas</span></span>
<span id="cb54-34"><a href="#cb54-34" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> line[<span class="dv">2</span>] <span class="op">==</span> <span class="st">'default'</span>: <span class="co"># si el tercer elemento es default entonces se conectan los terms en el grafo</span></span>
<span id="cb54-35"><a href="#cb54-35" aria-hidden="true" tabindex="-1"></a>            dG.add_edge(line[<span class="dv">0</span>], line[<span class="dv">1</span>]) <span class="co"># Add an edge between line[0] and line[1]</span></span>
<span id="cb54-36"><a href="#cb54-36" aria-hidden="true" tabindex="-1"></a>            terms_pairs.append([line[<span class="dv">0</span>], line[<span class="dv">1</span>]]) <span class="co"># Add the pair to the list</span></span>
<span id="cb54-37"><a href="#cb54-37" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb54-38"><a href="#cb54-38" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> line[<span class="dv">1</span>] <span class="kw">not</span> <span class="kw">in</span> gene2id_mapping: <span class="co"># se salta el gen si no es parte de los que estan en gene2id_mapping</span></span>
<span id="cb54-39"><a href="#cb54-39" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(line[<span class="dv">1</span>])</span>
<span id="cb54-40"><a href="#cb54-40" aria-hidden="true" tabindex="-1"></a>                <span class="cf">continue</span></span>
<span id="cb54-41"><a href="#cb54-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-42"><a href="#cb54-42" aria-hidden="true" tabindex="-1"></a>            genes_terms_pairs.append([line[<span class="dv">0</span>], line[<span class="dv">1</span>]]) <span class="co"># add the pair</span></span>
<span id="cb54-43"><a href="#cb54-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-44"><a href="#cb54-44" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> line[<span class="dv">0</span>] <span class="kw">not</span> <span class="kw">in</span> term_direct_gene_map: <span class="co"># si el termino todavia no esta en el diccionario lo agrega</span></span>
<span id="cb54-45"><a href="#cb54-45" aria-hidden="true" tabindex="-1"></a>                term_direct_gene_map[ line[<span class="dv">0</span>] ] <span class="op">=</span> <span class="bu">set</span>() <span class="co"># crea un set</span></span>
<span id="cb54-46"><a href="#cb54-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-47"><a href="#cb54-47" aria-hidden="true" tabindex="-1"></a>            term_direct_gene_map[line[<span class="dv">0</span>]].add(gene2id_mapping[line[<span class="dv">1</span>]]) <span class="co"># añadimos el gen al set de ese term</span></span>
<span id="cb54-48"><a href="#cb54-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-49"><a href="#cb54-49" aria-hidden="true" tabindex="-1"></a>            gene_set.add(line[<span class="dv">1</span>]) <span class="co"># añadimos el gen al set total de genes</span></span>
<span id="cb54-50"><a href="#cb54-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-51"><a href="#cb54-51" aria-hidden="true" tabindex="-1"></a>    terms_pairs <span class="op">=</span> np.array(terms_pairs) <span class="co"># convert to 2d array</span></span>
<span id="cb54-52"><a href="#cb54-52" aria-hidden="true" tabindex="-1"></a>    genes_terms_pairs <span class="op">=</span> np.array(genes_terms_pairs) <span class="co"># convert to 2d array</span></span>
<span id="cb54-53"><a href="#cb54-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-54"><a href="#cb54-54" aria-hidden="true" tabindex="-1"></a>    file_handle.close()</span>
<span id="cb54-55"><a href="#cb54-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-56"><a href="#cb54-56" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'There are'</span>, <span class="bu">len</span>(gene_set), <span class="st">'genes'</span>)</span>
<span id="cb54-57"><a href="#cb54-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-58"><a href="#cb54-58" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> term <span class="kw">in</span> dG.nodes(): <span class="co"># hacemos esto para cada uno de los GO terms</span></span>
<span id="cb54-59"><a href="#cb54-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-60"><a href="#cb54-60" aria-hidden="true" tabindex="-1"></a>        term_gene_set <span class="op">=</span> <span class="bu">set</span>() <span class="co"># se crea un set</span></span>
<span id="cb54-61"><a href="#cb54-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-62"><a href="#cb54-62" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> term <span class="kw">in</span> term_direct_gene_map:</span>
<span id="cb54-63"><a href="#cb54-63" aria-hidden="true" tabindex="-1"></a>            term_gene_set <span class="op">=</span> term_direct_gene_map[term] <span class="co"># genes conectados al term</span></span>
<span id="cb54-64"><a href="#cb54-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-65"><a href="#cb54-65" aria-hidden="true" tabindex="-1"></a>        deslist <span class="op">=</span> nxadag.descendants(dG, term) <span class="co">#regresa todos sus GO terms descendientes (biological processes tiene 2085 descendientes, todos menos el mismo)</span></span>
<span id="cb54-66"><a href="#cb54-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-67"><a href="#cb54-67" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> child <span class="kw">in</span> deslist:</span>
<span id="cb54-68"><a href="#cb54-68" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> child <span class="kw">in</span> term_direct_gene_map: <span class="co"># añadir los genes de sus descendientes</span></span>
<span id="cb54-69"><a href="#cb54-69" aria-hidden="true" tabindex="-1"></a>                term_gene_set <span class="op">=</span> term_gene_set <span class="op">|</span> term_direct_gene_map[child] <span class="co"># union of both sets, ahora tiene todos los genes los suyos y los de sus descendientes</span></span>
<span id="cb54-70"><a href="#cb54-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-71"><a href="#cb54-71" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(term_gene_set) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb54-72"><a href="#cb54-72" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="st">'There is empty terms, please delete term:'</span>, term)</span>
<span id="cb54-73"><a href="#cb54-73" aria-hidden="true" tabindex="-1"></a>            sys.exit(<span class="dv">1</span>)</span>
<span id="cb54-74"><a href="#cb54-74" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb54-75"><a href="#cb54-75" aria-hidden="true" tabindex="-1"></a>            <span class="co"># por ahora esta variable no me hace falta</span></span>
<span id="cb54-76"><a href="#cb54-76" aria-hidden="true" tabindex="-1"></a>            term_size_map[term] <span class="op">=</span> <span class="bu">len</span>(term_gene_set) <span class="co"># cantidad de genes en ese term  (tomando en cuenta sus descendientes)</span></span>
<span id="cb54-77"><a href="#cb54-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-78"><a href="#cb54-78" aria-hidden="true" tabindex="-1"></a>    leaves <span class="op">=</span> [n <span class="cf">for</span> n <span class="kw">in</span> dG.nodes <span class="cf">if</span> dG.in_degree(n) <span class="op">==</span> <span class="dv">0</span>] <span class="co"># buscar la raiz</span></span>
<span id="cb54-79"><a href="#cb54-79" aria-hidden="true" tabindex="-1"></a>    <span class="co">#leaves = [n for n,d in dG.in_degree() if d==0]</span></span>
<span id="cb54-80"><a href="#cb54-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-81"><a href="#cb54-81" aria-hidden="true" tabindex="-1"></a>    uG <span class="op">=</span> dG.to_undirected() <span class="co"># Returns an undirected representation of the digraph</span></span>
<span id="cb54-82"><a href="#cb54-82" aria-hidden="true" tabindex="-1"></a>    connected_subG_list <span class="op">=</span> <span class="bu">list</span>(nxacc.connected_components(uG)) <span class="co">#list of all GO terms</span></span>
<span id="cb54-83"><a href="#cb54-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-84"><a href="#cb54-84" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Verify my graph makes sense...</span></span>
<span id="cb54-85"><a href="#cb54-85" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'There are'</span>, <span class="bu">len</span>(leaves), <span class="st">'roots:'</span>, leaves[<span class="dv">0</span>])</span>
<span id="cb54-86"><a href="#cb54-86" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'There are'</span>, <span class="bu">len</span>(dG.nodes()), <span class="st">'terms'</span>)</span>
<span id="cb54-87"><a href="#cb54-87" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'There are'</span>, <span class="bu">len</span>(connected_subG_list), <span class="st">'connected components'</span>)</span>
<span id="cb54-88"><a href="#cb54-88" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(leaves) <span class="op">&gt;</span> <span class="dv">1</span>:</span>
<span id="cb54-89"><a href="#cb54-89" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">'There are more than 1 root of ontology. Please use only one root.'</span>)</span>
<span id="cb54-90"><a href="#cb54-90" aria-hidden="true" tabindex="-1"></a>        sys.exit(<span class="dv">1</span>)</span>
<span id="cb54-91"><a href="#cb54-91" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(connected_subG_list) <span class="op">&gt;</span> <span class="dv">1</span>:</span>
<span id="cb54-92"><a href="#cb54-92" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>( <span class="st">'There are more than connected components. Please connect them.'</span>)</span>
<span id="cb54-93"><a href="#cb54-93" aria-hidden="true" tabindex="-1"></a>        sys.exit(<span class="dv">1</span>)</span>
<span id="cb54-94"><a href="#cb54-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-95"><a href="#cb54-95" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> dG, terms_pairs, genes_terms_pairs, term_direct_gene_map, term_size_map</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sparsego-graph" class="level3">
<h3 class="anchored" data-anchor-id="sparsego-graph">SparseGO graph</h3>
<div class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import SparseGO graph (to extract all nodes/terms)... </span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load ontology: create the graph of connected GO terms</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>dG, terms_pairs, genes_terms_pairs, term_direct_gene_map, term_size_map <span class="op">=</span> load_ontology_extra_output(onto, gene2id_mapping)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="co">####</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>sparseGO_terms <span class="op">=</span> <span class="bu">list</span>(dG.nodes())</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>sparseGO_terms.remove(<span class="st">"GO:0008150"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>There are 14457 genes
There are 1 roots: GO:0008150
There are 4840 terms
There are 1 connected components</code></pre>
</div>
</div>
</section>
<section id="full-go-graph" class="level3">
<h3 class="anchored" data-anchor-id="full-go-graph">Full GO graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import full graph (to find parents)...</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> obonet</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="co">#import networkx as nx</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>url <span class="op">=</span> <span class="st">'http://purl.obolibrary.org/obo/go/go-basic.obo'</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>full_graph <span class="op">=</span> obonet.read_obo(url)</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>full_graph <span class="op">=</span> full_graph.reverse() <span class="co"># change the direction of nodes</span></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>[n <span class="cf">for</span> n <span class="kw">in</span> full_graph.nodes <span class="cf">if</span> full_graph.in_degree(n) <span class="op">==</span> <span class="dv">0</span>] <span class="co"># graph contains the 3 roots (BP,MF,CC)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="40">
<pre><code>['GO:0003674', 'GO:0005575', 'GO:0008150']</code></pre>
</div>
</div>
</section>
<section id="match-terms" class="level3">
<h3 class="anchored" data-anchor-id="match-terms">Match terms!</h3>
<p>Find all terms that match, terms that are part of both, the sparseGO graph and the drug slim results… if the slim terms’ ascendants are a match, they are also added</p>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Each model has DIFFERENT matches (the graph is different)</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>compounds_GOterms_matches <span class="op">=</span> {}</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> drug <span class="kw">in</span> compounds_GOterms.keys():</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># choose drug</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>    drug_df <span class="op">=</span> compounds_GOterms[drug]</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>    drug_slim_GOterms <span class="op">=</span> <span class="bu">set</span>(drug_df.loc[drug_df[<span class="dv">2</span>] <span class="op">==</span> <span class="st">"GoProcess"</span>][<span class="dv">0</span>]) <span class="co"># only GO processes</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">#set(sparseGO_terms) &amp; set(drug_slim_GOterms)</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>    drug_matches <span class="op">=</span> [] <span class="co"># store all directly matched terms and matches with all parents</span></span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> term <span class="kw">in</span> drug_slim_GOterms: <span class="co"># term ='GO:1902669' # buen ejemplo </span></span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> term <span class="kw">in</span> sparseGO_terms: <span class="co"># is the term in the sparseGO terms? </span></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>            drug_matches.append([<span class="dv">1</span>,term]) <span class="co"># add to list</span></span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>            <span class="co">#1: same term, 2:not direct match  (esto igual despues...the number indicates how direct is the relationship 0:same term, 1: parent, 2: grandpa, 3:...)</span></span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a>        <span class="co"># are its ascendants in the sparseGO terms? </span></span>
<span id="cb59-16"><a href="#cb59-16" aria-hidden="true" tabindex="-1"></a>        parents <span class="op">=</span> [source <span class="cf">for</span> source, _ <span class="kw">in</span>  full_graph.in_edges(term)] <span class="co"># parents of term</span></span>
<span id="cb59-17"><a href="#cb59-17" aria-hidden="true" tabindex="-1"></a>        relationship <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb59-18"><a href="#cb59-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">while</span>(<span class="bu">len</span>(parents)<span class="op">&gt;</span><span class="dv">0</span>): <span class="co"># check all ascendants </span></span>
<span id="cb59-19"><a href="#cb59-19" aria-hidden="true" tabindex="-1"></a>            <span class="co">#relationship+=1</span></span>
<span id="cb59-20"><a href="#cb59-20" aria-hidden="true" tabindex="-1"></a>            parents <span class="op">=</span> [source <span class="cf">for</span> source, _ <span class="kw">in</span>  full_graph.in_edges(parents)] <span class="co"># parents of parents</span></span>
<span id="cb59-21"><a href="#cb59-21" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb59-22"><a href="#cb59-22" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> parent_term <span class="kw">in</span> parents: <span class="co"># add parents that match sparseGO terms </span></span>
<span id="cb59-23"><a href="#cb59-23" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> parent_term <span class="kw">in</span> sparseGO_terms:</span>
<span id="cb59-24"><a href="#cb59-24" aria-hidden="true" tabindex="-1"></a>                    drug_matches.append([relationship, parent_term])</span>
<span id="cb59-25"><a href="#cb59-25" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb59-26"><a href="#cb59-26" aria-hidden="true" tabindex="-1"></a>        drug_matches <span class="op">=</span> (pd.DataFrame(drug_matches).drop_duplicates()).values.tolist() <span class="co"># remove duplicates</span></span>
<span id="cb59-27"><a href="#cb59-27" aria-hidden="true" tabindex="-1"></a>        compounds_GOterms_matches[drug] <span class="op">=</span> drug_matches</span>
<span id="cb59-28"><a href="#cb59-28" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(drug)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># delete drugs that have no matches</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>compounds_GOterms_matches <span class="op">=</span> {i:j <span class="cf">for</span> i,j <span class="kw">in</span> compounds_GOterms_matches.items() <span class="cf">if</span> j <span class="op">!=</span> []}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(compounds_GOterms_matches)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="33">
<pre><code>265</code></pre>
</div>
</div>
</section>
</section>
<section id="sparsego-terms-x-drugslim-terms-matrix" class="level2">
<h2 class="anchored" data-anchor-id="sparsego-terms-x-drugslim-terms-matrix">SparseGO terms x drugSlim terms matrix</h2>
<div class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all.columns <span class="op">=</span> attribution_data_all.columns.<span class="bu">str</span>.lower() <span class="co"># in order to match the term</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="35">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>(-)-parthenolide</th>
      <th>(5z)-7-oxozeaenol</th>
      <th>5-fluorouracil</th>
      <th>681640</th>
      <th>a-443654</th>
      <th>a-484954</th>
      <th>a-770041</th>
      <th>a-83-01</th>
      <th>acadesine</th>
      <th>acy-1215</th>
      <th>...</th>
      <th>vemurafenib</th>
      <th>ver-155008</th>
      <th>vorapaxar</th>
      <th>vorinostat:carboplatin (1:1 mol/mol)</th>
      <th>vorinostat:navitoclax (4:1 mol/mol)</th>
      <th>vu0155056</th>
      <th>wz4002</th>
      <th>wz8040</th>
      <th>yl54</th>
      <th>zebularine</th>
    </tr>
    <tr>
      <th>GO_term</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:0000012_1</th>
      <td>0.009143</td>
      <td>0.012238</td>
      <td>0.014801</td>
      <td>0.008749</td>
      <td>0.028462</td>
      <td>0.002892</td>
      <td>0.021746</td>
      <td>0.003400</td>
      <td>0.002582</td>
      <td>0.015812</td>
      <td>...</td>
      <td>0.010666</td>
      <td>0.008861</td>
      <td>0.014170</td>
      <td>0.006866</td>
      <td>0.016116</td>
      <td>0.010692</td>
      <td>0.023399</td>
      <td>0.022375</td>
      <td>0.010787</td>
      <td>0.007544</td>
    </tr>
    <tr>
      <th>GO:0000012_2</th>
      <td>-0.000285</td>
      <td>-0.000928</td>
      <td>-0.000579</td>
      <td>0.000208</td>
      <td>-0.001000</td>
      <td>-0.000165</td>
      <td>-0.002064</td>
      <td>-0.000199</td>
      <td>-0.000089</td>
      <td>-0.000651</td>
      <td>...</td>
      <td>-0.001053</td>
      <td>-0.000468</td>
      <td>-0.000831</td>
      <td>-0.000112</td>
      <td>-0.000545</td>
      <td>-0.000622</td>
      <td>-0.001353</td>
      <td>-0.001063</td>
      <td>-0.000853</td>
      <td>-0.000307</td>
    </tr>
    <tr>
      <th>GO:0000012_3</th>
      <td>0.000593</td>
      <td>0.001505</td>
      <td>0.003211</td>
      <td>0.000860</td>
      <td>0.002106</td>
      <td>-0.000137</td>
      <td>0.003661</td>
      <td>-0.000340</td>
      <td>-0.000157</td>
      <td>0.002306</td>
      <td>...</td>
      <td>0.000033</td>
      <td>0.001873</td>
      <td>-0.000379</td>
      <td>0.001317</td>
      <td>-0.000275</td>
      <td>0.000724</td>
      <td>0.003970</td>
      <td>0.005052</td>
      <td>-0.000554</td>
      <td>0.000269</td>
    </tr>
    <tr>
      <th>GO:0000012_4</th>
      <td>-0.000407</td>
      <td>-0.000310</td>
      <td>-0.000145</td>
      <td>-0.000331</td>
      <td>-0.000294</td>
      <td>-0.000306</td>
      <td>-0.000686</td>
      <td>-0.000259</td>
      <td>-0.000302</td>
      <td>-0.000689</td>
      <td>...</td>
      <td>-0.000542</td>
      <td>-0.000382</td>
      <td>0.000087</td>
      <td>-0.000564</td>
      <td>-0.002327</td>
      <td>-0.000219</td>
      <td>0.000312</td>
      <td>0.000633</td>
      <td>-0.000277</td>
      <td>-0.000085</td>
    </tr>
    <tr>
      <th>GO:0000012_5</th>
      <td>-0.000152</td>
      <td>-0.000454</td>
      <td>-0.000350</td>
      <td>0.000121</td>
      <td>-0.000196</td>
      <td>-0.000090</td>
      <td>-0.000926</td>
      <td>-0.000056</td>
      <td>-0.000057</td>
      <td>-0.000293</td>
      <td>...</td>
      <td>-0.000339</td>
      <td>-0.000374</td>
      <td>-0.000076</td>
      <td>-0.000136</td>
      <td>-0.000653</td>
      <td>-0.000179</td>
      <td>-0.000458</td>
      <td>-0.000314</td>
      <td>-0.000226</td>
      <td>-0.000019</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 790 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="36">
<pre><code>(29034, 790)</code></pre>
</div>
</div>
<p>Only keep drugs that have annotated GO terms</p>
<div class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>attribution_data_annotated <span class="op">=</span> attribution_data_all[<span class="bu">list</span>(compounds_GOterms_matches.keys())]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>attribution_data_annotated.shape <span class="co"># 230 DRUGS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="38">
<pre><code>(29034, 265)</code></pre>
</div>
</div>
<section id="build-drugslim-moa-matrix" class="level3">
<h3 class="anchored" data-anchor-id="build-drugslim-moa-matrix">Build drugSlim (MoA) matrix</h3>
<div class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>slim_matrix <span class="op">=</span> attribution_data_annotated.copy() <span class="co"># copy dataframe in order to build a similar matrix </span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> col <span class="kw">in</span> slim_matrix.columns:</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>    slim_matrix[col].values[:] <span class="op">=</span> <span class="dv">0</span> <span class="co"># empty matrix </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> drug <span class="kw">in</span> compounds_GOterms_matches.keys():</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>    drug_matches <span class="op">=</span> compounds_GOterms_matches[drug]</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>    drug_matches_names <span class="op">=</span> <span class="bu">list</span>(pd.DataFrame(drug_matches)[<span class="dv">1</span>])</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>    drug_matches_names_duplicated <span class="op">=</span> []</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> term <span class="kw">in</span> <span class="bu">set</span>(drug_matches_names):</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">7</span>):</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>            drug_matches_names_duplicated.append(term<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(i))</span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>    slim_matrix[drug][drug_matches_names_duplicated] <span class="op">=</span> <span class="dv">1</span> <span class="co"># add a 1 if term is annotated to drug</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>slim_matrix.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="41">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>5-fluorouracil</th>
      <th>acadesine</th>
      <th>acy-1215</th>
      <th>afatinib</th>
      <th>alectinib</th>
      <th>alisertib</th>
      <th>alvocidib</th>
      <th>amuvatinib</th>
      <th>apitolisib</th>
      <th>arry-520</th>
      <th>...</th>
      <th>sch-79797</th>
      <th>rita</th>
      <th>pifithrin-mu</th>
      <th>l-685458</th>
      <th>nutlin-3</th>
      <th>pi-103</th>
      <th>neopeltolide</th>
      <th>avrainvillamide</th>
      <th>indisulam</th>
      <th>piperlongumine:mst-312 (1:1 mol/mol)</th>
    </tr>
    <tr>
      <th>GO_term</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:0000012_1</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>GO:0000012_2</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>GO:0000012_3</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>GO:0000012_4</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>GO:0000012_5</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 265 columns</p>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="svm" class="level1">
<h1>SVM</h1>
<div class="cell" data-jp-markdownheadingcollapsed="true" data-tags="[]" data-execution_count="42">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> train_test_split</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn <span class="im">import</span> metrics</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.linear_model <span class="im">import</span> LogisticRegression</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn <span class="im">import</span> svm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>slim_matrix_single_neuron <span class="op">=</span> pd.DataFrame(<span class="dv">0</span>, index<span class="op">=</span>sparseGO_terms, columns<span class="op">=</span>slim_matrix.columns)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>preds_svm_matrix <span class="op">=</span> pd.DataFrame(<span class="dv">0</span>, index<span class="op">=</span>sparseGO_terms, columns<span class="op">=</span>slim_matrix.columns)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>platt_matrix <span class="op">=</span> pd.DataFrame(<span class="dv">0</span>, index<span class="op">=</span>sparseGO_terms, columns<span class="op">=</span>slim_matrix.columns)</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>distance_matrix <span class="op">=</span> pd.DataFrame(<span class="dv">0</span>, index<span class="op">=</span>sparseGO_terms, columns<span class="op">=</span>slim_matrix.columns)</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>delta_logits_matrix  <span class="op">=</span> pd.DataFrame(<span class="dv">0</span>, index<span class="op">=</span>sparseGO_terms, columns<span class="op">=</span>slim_matrix.columns)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="create-models" class="level2">
<h2 class="anchored" data-anchor-id="create-models">Create models</h2>
<p><strong>SVM models…</strong></p>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="49">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Dictionaries to store results</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm <span class="op">=</span> {}</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm <span class="op">=</span> {}</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>GO_terms_precision_svm <span class="op">=</span> {}</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_delta_logits <span class="op">=</span> {}</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform logistic</span></span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> goterm <span class="kw">in</span> sparseGO_terms:</span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if (real_go_info[real_go_info["GO_term"]==goterm+"_1"]["layer_number"]).values &gt;3:</span></span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     continue</span></span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># store results of each cross validation</span></span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a>    all_y_test <span class="op">=</span> []</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a>    all_y_pred_proba <span class="op">=</span> []</span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>    all_y_pred_proba_dis <span class="op">=</span> []</span>
<span id="cb76-17"><a href="#cb76-17" aria-hidden="true" tabindex="-1"></a>    all_y_pred <span class="op">=</span> []</span>
<span id="cb76-18"><a href="#cb76-18" aria-hidden="true" tabindex="-1"></a>    all_y_names <span class="op">=</span> []</span>
<span id="cb76-19"><a href="#cb76-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-20"><a href="#cb76-20" aria-hidden="true" tabindex="-1"></a>    goterm_drugs <span class="op">=</span> slim_matrix.loc[[goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(<span class="dv">1</span>)]].values.flatten()</span>
<span id="cb76-21"><a href="#cb76-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-22"><a href="#cb76-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">sum</span>(goterm_drugs) <span class="op">&lt;=</span> <span class="dv">16</span>: <span class="co"># at least 2 annotated drugs in each group</span></span>
<span id="cb76-23"><a href="#cb76-23" aria-hidden="true" tabindex="-1"></a>            <span class="cf">continue</span></span>
<span id="cb76-24"><a href="#cb76-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-25"><a href="#cb76-25" aria-hidden="true" tabindex="-1"></a>    list_nodes <span class="op">=</span> []</span>
<span id="cb76-26"><a href="#cb76-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">7</span>):</span>
<span id="cb76-27"><a href="#cb76-27" aria-hidden="true" tabindex="-1"></a>        list_nodes.append(goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(i))</span>
<span id="cb76-28"><a href="#cb76-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-29"><a href="#cb76-29" aria-hidden="true" tabindex="-1"></a>    score <span class="op">=</span> attribution_data_annotated.loc[list_nodes].T</span>
<span id="cb76-30"><a href="#cb76-30" aria-hidden="true" tabindex="-1"></a>    score_mod <span class="op">=</span> score.divide(score.std()).fillna(<span class="dv">0</span>) <span class="co"># AFECTA MUCHO</span></span>
<span id="cb76-31"><a href="#cb76-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-32"><a href="#cb76-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Separate drugs in 4 groups for cross-validation -----</span></span>
<span id="cb76-33"><a href="#cb76-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-34"><a href="#cb76-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Split data in 2 groups (with train_test_split in order to have 0s in both groups)</span></span>
<span id="cb76-35"><a href="#cb76-35" aria-hidden="true" tabindex="-1"></a>    X_part1,X_part2,y_part1,y_part2<span class="op">=</span>train_test_split(score_mod,goterm_drugs,test_size<span class="op">=</span><span class="fl">0.50</span>,random_state<span class="op">=</span><span class="dv">0</span>,stratify<span class="op">=</span>goterm_drugs)</span>
<span id="cb76-36"><a href="#cb76-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Split data again in 4 groups (split data previously split)</span></span>
<span id="cb76-37"><a href="#cb76-37" aria-hidden="true" tabindex="-1"></a>    X_group1,X_group2,y_group1,y_group2<span class="op">=</span>train_test_split(X_part1,y_part1,test_size<span class="op">=</span><span class="fl">0.50</span>,random_state<span class="op">=</span><span class="dv">0</span>,stratify<span class="op">=</span>y_part1)</span>
<span id="cb76-38"><a href="#cb76-38" aria-hidden="true" tabindex="-1"></a>    X_group3,X_group4,y_group3,y_group4<span class="op">=</span>train_test_split(X_part2,y_part2,test_size<span class="op">=</span><span class="fl">0.50</span>,random_state<span class="op">=</span><span class="dv">0</span>,stratify<span class="op">=</span>y_part2)</span>
<span id="cb76-39"><a href="#cb76-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-40"><a href="#cb76-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span>  <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">5</span>):</span>
<span id="cb76-41"><a href="#cb76-41" aria-hidden="true" tabindex="-1"></a>        vector <span class="op">=</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="dv">5</span>)</span>
<span id="cb76-42"><a href="#cb76-42" aria-hidden="true" tabindex="-1"></a>        group_number <span class="op">=</span> <span class="bu">str</span>(i)</span>
<span id="cb76-43"><a href="#cb76-43" aria-hidden="true" tabindex="-1"></a>        X_test <span class="op">=</span> <span class="bu">globals</span>()[<span class="st">"X_group"</span><span class="op">+</span>group_number]</span>
<span id="cb76-44"><a href="#cb76-44" aria-hidden="true" tabindex="-1"></a>        y_test <span class="op">=</span> <span class="bu">globals</span>()[<span class="st">"y_group"</span><span class="op">+</span>group_number]</span>
<span id="cb76-45"><a href="#cb76-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-46"><a href="#cb76-46" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Use the other 3 groups for training </span></span>
<span id="cb76-47"><a href="#cb76-47" aria-hidden="true" tabindex="-1"></a>        keep <span class="op">=</span> <span class="bu">list</span>({<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>}<span class="op">-</span>{<span class="bu">int</span>(group_number)}) <span class="co"># remove group number of current test </span></span>
<span id="cb76-48"><a href="#cb76-48" aria-hidden="true" tabindex="-1"></a>        X_train <span class="op">=</span> pd.concat((<span class="bu">globals</span>()[<span class="st">"X_group"</span><span class="op">+</span><span class="bu">str</span>(keep[<span class="dv">0</span>])],<span class="bu">globals</span>()[<span class="st">"X_group"</span><span class="op">+</span><span class="bu">str</span>(keep[<span class="dv">1</span>])],<span class="bu">globals</span>()[<span class="st">"X_group"</span><span class="op">+</span><span class="bu">str</span>(keep[<span class="dv">2</span>])]))</span>
<span id="cb76-49"><a href="#cb76-49" aria-hidden="true" tabindex="-1"></a>        y_train <span class="op">=</span> np.concatenate((<span class="bu">globals</span>()[<span class="st">"y_group"</span><span class="op">+</span><span class="bu">str</span>(keep[<span class="dv">0</span>])],<span class="bu">globals</span>()[<span class="st">"y_group"</span><span class="op">+</span><span class="bu">str</span>(keep[<span class="dv">1</span>])],<span class="bu">globals</span>()[<span class="st">"y_group"</span><span class="op">+</span><span class="bu">str</span>(keep[<span class="dv">2</span>])]))</span>
<span id="cb76-50"><a href="#cb76-50" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb76-51"><a href="#cb76-51" aria-hidden="true" tabindex="-1"></a>        <span class="co">#gamma = 1/(X_train.shape[1]*X_train.to_numpy().var())</span></span>
<span id="cb76-52"><a href="#cb76-52" aria-hidden="true" tabindex="-1"></a>        gamma <span class="op">=</span> <span class="st">"scale"</span></span>
<span id="cb76-53"><a href="#cb76-53" aria-hidden="true" tabindex="-1"></a>        C<span class="op">=</span><span class="dv">1</span></span>
<span id="cb76-54"><a href="#cb76-54" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb76-55"><a href="#cb76-55" aria-hidden="true" tabindex="-1"></a>        svm_model <span class="op">=</span> svm.SVC(C<span class="op">=</span>C,gamma<span class="op">=</span>gamma, kernel<span class="op">=</span><span class="st">'rbf'</span>,</span>
<span id="cb76-56"><a href="#cb76-56" aria-hidden="true" tabindex="-1"></a>                           class_weight<span class="op">=</span><span class="st">"balanced"</span>,</span>
<span id="cb76-57"><a href="#cb76-57" aria-hidden="true" tabindex="-1"></a>                            tol<span class="op">=</span><span class="fl">0.001</span>,</span>
<span id="cb76-58"><a href="#cb76-58" aria-hidden="true" tabindex="-1"></a>                            probability<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb76-59"><a href="#cb76-59" aria-hidden="true" tabindex="-1"></a>                            random_state<span class="op">=</span><span class="dv">1234</span>)</span>
<span id="cb76-60"><a href="#cb76-60" aria-hidden="true" tabindex="-1"></a>        <span class="co"># svm_model = svm.SVC(gamma='auto', kernel='rbf',class_weight="balanced",probability=True)</span></span>
<span id="cb76-61"><a href="#cb76-61" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb76-62"><a href="#cb76-62" aria-hidden="true" tabindex="-1"></a>        <span class="co"># fit the model with data</span></span>
<span id="cb76-63"><a href="#cb76-63" aria-hidden="true" tabindex="-1"></a>        svm_model.fit(X_train,y_train)</span>
<span id="cb76-64"><a href="#cb76-64" aria-hidden="true" tabindex="-1"></a>        y_pred<span class="op">=</span>svm_model.predict(X_test)</span>
<span id="cb76-65"><a href="#cb76-65" aria-hidden="true" tabindex="-1"></a>        y_pred_proba <span class="op">=</span> svm_model.predict_proba(X_test)[::,<span class="dv">1</span>] <span class="co"># platt values</span></span>
<span id="cb76-66"><a href="#cb76-66" aria-hidden="true" tabindex="-1"></a>        y_pred_proba_dis <span class="op">=</span> svm_model.decision_function(X_test) <span class="co"># An SVM returns a real-valued prediction for each of the input data samples, which corresponds to its distance from the separating hyperplane.</span></span>
<span id="cb76-67"><a href="#cb76-67" aria-hidden="true" tabindex="-1"></a>        <span class="co">#  decision_function SORTS the results from most probable class to the least probable one.</span></span>
<span id="cb76-68"><a href="#cb76-68" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb76-69"><a href="#cb76-69" aria-hidden="true" tabindex="-1"></a>        all_y_test.append(y_test)</span>
<span id="cb76-70"><a href="#cb76-70" aria-hidden="true" tabindex="-1"></a>        all_y_pred_proba.append(y_pred_proba)</span>
<span id="cb76-71"><a href="#cb76-71" aria-hidden="true" tabindex="-1"></a>        all_y_pred_proba_dis.append(y_pred_proba_dis)</span>
<span id="cb76-72"><a href="#cb76-72" aria-hidden="true" tabindex="-1"></a>        all_y_pred.append(y_pred)</span>
<span id="cb76-73"><a href="#cb76-73" aria-hidden="true" tabindex="-1"></a>        all_y_names.append(X_test.index)</span>
<span id="cb76-74"><a href="#cb76-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-75"><a href="#cb76-75" aria-hidden="true" tabindex="-1"></a>    all_y_test <span class="op">=</span> np.concatenate(all_y_test)</span>
<span id="cb76-76"><a href="#cb76-76" aria-hidden="true" tabindex="-1"></a>    all_y_pred_proba <span class="op">=</span> np.concatenate(all_y_pred_proba)</span>
<span id="cb76-77"><a href="#cb76-77" aria-hidden="true" tabindex="-1"></a>    all_y_pred_proba_dis <span class="op">=</span> np.concatenate(all_y_pred_proba_dis)</span>
<span id="cb76-78"><a href="#cb76-78" aria-hidden="true" tabindex="-1"></a>    all_y_names <span class="op">=</span> np.concatenate(all_y_names)</span>
<span id="cb76-79"><a href="#cb76-79" aria-hidden="true" tabindex="-1"></a>    all_y_pred <span class="op">=</span> np.concatenate(all_y_pred)</span>
<span id="cb76-80"><a href="#cb76-80" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb76-81"><a href="#cb76-81" aria-hidden="true" tabindex="-1"></a>    percentage_go_annotations <span class="op">=</span> <span class="bu">sum</span>(all_y_test)<span class="op">/</span><span class="bu">len</span>(all_y_test)</span>
<span id="cb76-82"><a href="#cb76-82" aria-hidden="true" tabindex="-1"></a>    logits_apriori<span class="op">=</span>np.log(percentage_go_annotations<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>percentage_go_annotations))</span>
<span id="cb76-83"><a href="#cb76-83" aria-hidden="true" tabindex="-1"></a>    logits_apost<span class="op">=</span> np.log(all_y_pred_proba<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>all_y_pred_proba))</span>
<span id="cb76-84"><a href="#cb76-84" aria-hidden="true" tabindex="-1"></a>    delta_logits <span class="op">=</span> logits_apost<span class="op">-</span>logits_apriori</span>
<span id="cb76-85"><a href="#cb76-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-86"><a href="#cb76-86" aria-hidden="true" tabindex="-1"></a>    platt_matrix.loc[goterm,all_y_names] <span class="op">=</span> all_y_pred_proba</span>
<span id="cb76-87"><a href="#cb76-87" aria-hidden="true" tabindex="-1"></a>    distance_matrix.loc[goterm,all_y_names] <span class="op">=</span> all_y_pred_proba_dis</span>
<span id="cb76-88"><a href="#cb76-88" aria-hidden="true" tabindex="-1"></a>    slim_matrix_single_neuron.loc[goterm,all_y_names] <span class="op">=</span> all_y_test</span>
<span id="cb76-89"><a href="#cb76-89" aria-hidden="true" tabindex="-1"></a>    preds_svm_matrix.loc[goterm,all_y_names] <span class="op">=</span> all_y_pred</span>
<span id="cb76-90"><a href="#cb76-90" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb76-91"><a href="#cb76-91" aria-hidden="true" tabindex="-1"></a>    delta_logits_matrix.loc[goterm,all_y_names] <span class="op">=</span> delta_logits</span>
<span id="cb76-92"><a href="#cb76-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-93"><a href="#cb76-93" aria-hidden="true" tabindex="-1"></a>    GO_terms_auc_delta_logits[goterm] <span class="op">=</span> metrics.roc_auc_score(all_y_test, delta_logits)</span>
<span id="cb76-94"><a href="#cb76-94" aria-hidden="true" tabindex="-1"></a>    GO_terms_auc_svm[goterm] <span class="op">=</span> metrics.roc_auc_score(all_y_test, all_y_pred_proba)</span>
<span id="cb76-95"><a href="#cb76-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-96"><a href="#cb76-96" aria-hidden="true" tabindex="-1"></a>    precision, recall, thresholds <span class="op">=</span> metrics.precision_recall_curve(all_y_test, all_y_pred_proba)</span>
<span id="cb76-97"><a href="#cb76-97" aria-hidden="true" tabindex="-1"></a>    GO_terms_aupr_svm[goterm] <span class="op">=</span> metrics.auc(recall, precision)</span>
<span id="cb76-98"><a href="#cb76-98" aria-hidden="true" tabindex="-1"></a>    GO_terms_precision_svm[goterm] <span class="op">=</span> metrics.precision_score(all_y_test, all_y_pred)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="50">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># done with platt values</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(GO_terms_auc_svm.items()),columns <span class="op">=</span> [<span class="st">'goterm'</span>,<span class="st">'auc'</span>]).set_index(<span class="st">"goterm"</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df <span class="op">=</span> GO_terms_auc_svm_df.dropna()</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">False</span>).head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="50">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>auc</th>
    </tr>
    <tr>
      <th>goterm</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:1901029</th>
      <td>0.959991</td>
    </tr>
    <tr>
      <th>GO:0044342</th>
      <td>0.956590</td>
    </tr>
    <tr>
      <th>GO:0002326</th>
      <td>0.955918</td>
    </tr>
    <tr>
      <th>GO:0007006</th>
      <td>0.955500</td>
    </tr>
    <tr>
      <th>GO:1902236</th>
      <td>0.954857</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"There are "</span> <span class="op">+</span><span class="bu">str</span>(<span class="bu">len</span>(GO_terms_auc_svm_df))<span class="op">+</span> <span class="st">" svm models."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>There are 1124 svm models.</code></pre>
</div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="52">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># only keep goterms that have a model </span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>platt_matrix <span class="op">=</span> platt_matrix.loc[<span class="bu">list</span>(GO_terms_auc_svm_df.index),:]</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>distance_matrix <span class="op">=</span> distance_matrix.loc[<span class="bu">list</span>(GO_terms_auc_svm_df.index),:]</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>slim_matrix_single_neuron  <span class="op">=</span> slim_matrix_single_neuron.loc[<span class="bu">list</span>(GO_terms_auc_svm_df.index),:]</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>preds_svm_matrix  <span class="op">=</span> preds_svm_matrix.loc[<span class="bu">list</span>(GO_terms_auc_svm_df.index),:]</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>delta_logits_matrix  <span class="op">=</span> delta_logits_matrix.loc[<span class="bu">list</span>(GO_terms_auc_svm_df.index),:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="auc-histogram" class="level2">
<h2 class="anchored" data-anchor-id="auc-histogram">AUC histogram</h2>
<div class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">'figure.figsize'</span>:(<span class="dv">10</span>,<span class="dv">6</span>)})</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>perc <span class="op">=</span> <span class="bu">str</span>(<span class="bu">round</span>((<span class="dv">100</span><span class="op">*</span><span class="bu">len</span>(GO_terms_auc_svm_df[GO_terms_auc_svm_df[<span class="st">"auc"</span>]<span class="op">&gt;</span><span class="fl">0.69</span>])<span class="op">/</span><span class="bu">len</span>(GO_terms_auc_svm_df)),<span class="dv">2</span>))<span class="op">+</span><span class="st">"%"</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>N, bins, patches <span class="op">=</span> plt.hist(GO_terms_auc_svm_df, color<span class="op">=</span>CB_color_cycle[<span class="dv">6</span>],bins<span class="op">=</span><span class="dv">50</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(bins)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> bins[i]<span class="op">&gt;</span><span class="fl">0.69</span>:</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>        patches[i].set_facecolor(CB_color_cycle[<span class="dv">2</span>])</span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>plt.yticks(fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a>plt.xticks(fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb81-15"><a href="#cb81-15" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'left'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb81-16"><a href="#cb81-16" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'bottom'</span>].set_color(<span class="st">'#DDDDDD'</span>)</span>
<span id="cb81-17"><a href="#cb81-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-18"><a href="#cb81-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Second, remove the ticks as well.</span></span>
<span id="cb81-19"><a href="#cb81-19" aria-hidden="true" tabindex="-1"></a>ax.tick_params(bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb81-20"><a href="#cb81-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-21"><a href="#cb81-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Third, add a horizontal grid (but keep the vertical grid hidden).</span></span>
<span id="cb81-22"><a href="#cb81-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Color the lines a light gray as well.</span></span>
<span id="cb81-23"><a href="#cb81-23" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb81-24"><a href="#cb81-24" aria-hidden="true" tabindex="-1"></a>ax.yaxis.grid(<span class="va">True</span>, color<span class="op">=</span><span class="st">'#EEEEEE'</span>)</span>
<span id="cb81-25"><a href="#cb81-25" aria-hidden="true" tabindex="-1"></a>ax.xaxis.grid(<span class="va">False</span>)</span>
<span id="cb81-26"><a href="#cb81-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-27"><a href="#cb81-27" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"AUC value"</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb81-28"><a href="#cb81-28" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Number of GO term models"</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb81-29"><a href="#cb81-29" aria-hidden="true" tabindex="-1"></a>colors2 <span class="op">=</span> {<span class="st">'GO term models with AUC&gt;=0.7'</span>:CB_color_cycle[<span class="dv">2</span>]}  </span>
<span id="cb81-30"><a href="#cb81-30" aria-hidden="true" tabindex="-1"></a>labels <span class="op">=</span> <span class="bu">list</span>(colors2.keys())</span>
<span id="cb81-31"><a href="#cb81-31" aria-hidden="true" tabindex="-1"></a>handles <span class="op">=</span> [plt.Rectangle((<span class="dv">0</span>,<span class="dv">0</span>),<span class="dv">1</span>,<span class="dv">1</span>, color<span class="op">=</span>colors2[label]) <span class="cf">for</span> label <span class="kw">in</span> labels]</span>
<span id="cb81-32"><a href="#cb81-32" aria-hidden="true" tabindex="-1"></a>plt.legend(handles, labels,fontsize<span class="op">=</span><span class="dv">20</span>, loc<span class="op">=</span><span class="st">"lower left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">0.35</span>,<span class="op">-</span><span class="fl">0.35</span>))</span>
<span id="cb81-33"><a href="#cb81-33" aria-hidden="true" tabindex="-1"></a>plt.text(<span class="fl">0.71</span>, <span class="dv">8</span>, <span class="bu">str</span>(perc), fontsize<span class="op">=</span><span class="dv">20</span>,color<span class="op">=</span><span class="st">'#333333'</span>)</span>
<span id="cb81-34"><a href="#cb81-34" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Overall performance of the models using expression"</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span>
<span id="cb81-35"><a href="#cb81-35" aria-hidden="true" tabindex="-1"></a><span class="co"># con el que mejor funciona es con la suma normal del attribution </span></span>
<span id="cb81-36"><a href="#cb81-36" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb81-37"><a href="#cb81-37" aria-hidden="true" tabindex="-1"></a>fig.savefig(resultsdir<span class="op">+</span><span class="st">'modelsAUCsvm.png'</span>, transparent<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-68-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="auc-waterfall-plot" class="level2">
<h2 class="anchored" data-anchor-id="auc-waterfall-plot">AUC waterfall plot</h2>
<div class="cell" data-execution_count="54">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df <span class="op">=</span>GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> (<span class="dv">12</span>, <span class="dv">9</span>)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>drugs <span class="op">=</span> GO_terms_auc_svm_df.index</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>rhos <span class="op">=</span> GO_terms_auc_svm_df[<span class="st">"auc"</span>]</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>percentage <span class="op">=</span> <span class="bu">round</span>((<span class="bu">sum</span>(rhos<span class="op">&gt;</span><span class="fl">0.69</span>)<span class="op">/</span><span class="bu">len</span>(rhos))<span class="op">*</span><span class="dv">100</span>,<span class="dv">1</span>)</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a><span class="co">#colors = ['#208EA3' if (x &lt; 0.5) else '#A4C61A' for x in rhos ]</span></span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>colors <span class="op">=</span> [<span class="st">'#C9C9C9'</span> <span class="cf">if</span> (x <span class="op">&lt;</span> <span class="fl">0.69</span>) <span class="cf">else</span> <span class="st">"#6492CA"</span> <span class="cf">for</span> x <span class="kw">in</span> rhos ]</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>ax.bar(</span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>drugs,</span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a>    height<span class="op">=</span>rhos,</span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a>    edgecolor<span class="op">=</span>colors,</span>
<span id="cb83-14"><a href="#cb83-14" aria-hidden="true" tabindex="-1"></a>    linewidth<span class="op">=</span><span class="dv">2</span></span>
<span id="cb83-15"><a href="#cb83-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb83-16"><a href="#cb83-16" aria-hidden="true" tabindex="-1"></a>plt.xticks([])</span>
<span id="cb83-17"><a href="#cb83-17" aria-hidden="true" tabindex="-1"></a>plt.yticks(fontsize<span class="op">=</span><span class="dv">28</span>)</span>
<span id="cb83-18"><a href="#cb83-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-19"><a href="#cb83-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-20"><a href="#cb83-20" aria-hidden="true" tabindex="-1"></a><span class="co"># First, let's remove the top, right and left spines (figure borders)</span></span>
<span id="cb83-21"><a href="#cb83-21" aria-hidden="true" tabindex="-1"></a><span class="co"># which really aren't necessary for a bar chart.</span></span>
<span id="cb83-22"><a href="#cb83-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Also, make the bottom spine gray instead of black.</span></span>
<span id="cb83-23"><a href="#cb83-23" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb83-24"><a href="#cb83-24" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb83-25"><a href="#cb83-25" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'left'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb83-26"><a href="#cb83-26" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'bottom'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb83-27"><a href="#cb83-27" aria-hidden="true" tabindex="-1"></a><span class="co">#ax.spines['bottom'].set_color('#DDDDDD')</span></span>
<span id="cb83-28"><a href="#cb83-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-29"><a href="#cb83-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Second, remove the ticks as well.</span></span>
<span id="cb83-30"><a href="#cb83-30" aria-hidden="true" tabindex="-1"></a>ax.tick_params(bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb83-31"><a href="#cb83-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-32"><a href="#cb83-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Third, add a horizontal grid (but keep the vertical grid hidden).</span></span>
<span id="cb83-33"><a href="#cb83-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Color the lines a light gray as well.</span></span>
<span id="cb83-34"><a href="#cb83-34" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">False</span>)</span>
<span id="cb83-35"><a href="#cb83-35" aria-hidden="true" tabindex="-1"></a>ax.yaxis.grid(<span class="va">False</span>)</span>
<span id="cb83-36"><a href="#cb83-36" aria-hidden="true" tabindex="-1"></a><span class="co">#ax.yaxis.grid(True, color='#EEEEEE')</span></span>
<span id="cb83-37"><a href="#cb83-37" aria-hidden="true" tabindex="-1"></a>ax.xaxis.grid(<span class="va">False</span>)</span>
<span id="cb83-38"><a href="#cb83-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-39"><a href="#cb83-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-40"><a href="#cb83-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Add labels and a title. Note the use of `labelpad` and `pad` to add some</span></span>
<span id="cb83-41"><a href="#cb83-41" aria-hidden="true" tabindex="-1"></a><span class="co"># extra space between the text and the tick labels.</span></span>
<span id="cb83-42"><a href="#cb83-42" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'SVM models'</span>, labelpad<span class="op">=-</span><span class="dv">30</span>, color<span class="op">=</span><span class="st">'#333333'</span>,fontsize<span class="op">=</span><span class="dv">50</span>)</span>
<span id="cb83-43"><a href="#cb83-43" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'AUC-ROC value'</span>, labelpad<span class="op">=</span><span class="dv">15</span>, color<span class="op">=</span><span class="st">'#333333'</span>,fontsize<span class="op">=</span><span class="dv">50</span>)</span>
<span id="cb83-44"><a href="#cb83-44" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="st">''</span>, color<span class="op">=</span><span class="st">'#333333'</span>,</span>
<span id="cb83-45"><a href="#cb83-45" aria-hidden="true" tabindex="-1"></a>             weight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb83-46"><a href="#cb83-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-47"><a href="#cb83-47" aria-hidden="true" tabindex="-1"></a>colors2 <span class="op">=</span> {<span class="st">'High confidence drugs (r&gt;0.5)'</span>:<span class="st">'#A4C61A'</span>}  </span>
<span id="cb83-48"><a href="#cb83-48" aria-hidden="true" tabindex="-1"></a>labels <span class="op">=</span> <span class="bu">list</span>(colors2.keys())</span>
<span id="cb83-49"><a href="#cb83-49" aria-hidden="true" tabindex="-1"></a>handles <span class="op">=</span> [plt.Rectangle((<span class="dv">0</span>,<span class="dv">0</span>),<span class="dv">1</span>,<span class="dv">1</span>, color<span class="op">=</span>colors2[label]) <span class="cf">for</span> label <span class="kw">in</span> labels]</span>
<span id="cb83-50"><a href="#cb83-50" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.legend(handles, labels,fontsize=40, loc="lower left",bbox_to_anchor=(0, -0.215))</span></span>
<span id="cb83-51"><a href="#cb83-51" aria-hidden="true" tabindex="-1"></a>plt.text(<span class="dv">77</span>, <span class="fl">0.32</span>, <span class="bu">str</span>(percentage)<span class="op">+</span><span class="st">"%"</span>, fontsize<span class="op">=</span><span class="dv">60</span>,color<span class="op">=</span><span class="st">'#000000'</span>)</span>
<span id="cb83-52"><a href="#cb83-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-53"><a href="#cb83-53" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="fl">0.1</span>,<span class="fl">1.1</span>))</span>
<span id="cb83-54"><a href="#cb83-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Make the chart fill out the figure better.</span></span>
<span id="cb83-55"><a href="#cb83-55" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb83-56"><a href="#cb83-56" aria-hidden="true" tabindex="-1"></a>fig.savefig(resultsdir<span class="op">+</span><span class="st">'WaterfallModelsSVM.png'</span>, transparent<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-70-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="auc-boxplot-by-parents" class="level2">
<h2 class="anchored" data-anchor-id="auc-boxplot-by-parents">AUC boxplot by parents</h2>
<div class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add number of parents</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>number_parents <span class="op">=</span> {}</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>levels <span class="op">=</span> {}</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(GO_terms_auc_svm_df.index)):</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>    term <span class="op">=</span> GO_terms_auc_svm_df.index[i]</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>    number_parents[GO_terms_auc_svm_df.index[i]]<span class="op">=</span><span class="bu">len</span>([source <span class="cf">for</span> source, _ <span class="kw">in</span>  dG.in_edges(term)])</span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>    levels[GO_terms_auc_svm_df.index[i]]<span class="op">=</span>level_number[term]<span class="op">-</span><span class="dv">1</span></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>levels <span class="op">=</span> pd.DataFrame.from_dict(levels, orient<span class="op">=</span><span class="st">'index'</span>)</span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>number_parents <span class="op">=</span> pd.DataFrame.from_dict(number_parents, orient<span class="op">=</span><span class="st">'index'</span>)</span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df <span class="op">=</span> pd.concat([GO_terms_auc_svm_df, levels,number_parents], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df.columns <span class="op">=</span> [<span class="st">"auc"</span>,<span class="st">"levels"</span>,<span class="st">"parents"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df <span class="op">=</span> GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"levels"</span>], ascending<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="58">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="58">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>auc</th>
      <th>levels</th>
      <th>parents</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:1901029</th>
      <td>0.959991</td>
      <td>0</td>
      <td>10</td>
    </tr>
    <tr>
      <th>GO:0070374</th>
      <td>0.793109</td>
      <td>0</td>
      <td>3</td>
    </tr>
    <tr>
      <th>GO:0043552</th>
      <td>0.792174</td>
      <td>0</td>
      <td>3</td>
    </tr>
    <tr>
      <th>GO:0003157</th>
      <td>0.790148</td>
      <td>0</td>
      <td>2</td>
    </tr>
    <tr>
      <th>GO:0090314</th>
      <td>0.790000</td>
      <td>0</td>
      <td>7</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="73">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>c <span class="op">=</span> [<span class="st">'#E8384F'</span>, <span class="st">'#FD817D'</span>, <span class="st">'#FDAE33'</span>,<span class="st">'#EECC16'</span>, <span class="st">'#A4C61A'</span>, <span class="st">'#37A862'</span>,<span class="st">"#208EA3"</span>,<span class="st">"#3B6EAB"</span>]</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> px.data.tips()</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> px.box(GO_terms_auc_svm_df, x<span class="op">=</span><span class="st">"levels"</span>, y<span class="op">=</span><span class="st">"auc"</span>,</span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>             color<span class="op">=</span><span class="st">"levels"</span>,</span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>            color_discrete_sequence<span class="op">=</span>c,</span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a>             width <span class="op">=</span><span class="dv">600</span>,</span>
<span id="cb87-8"><a href="#cb87-8" aria-hidden="true" tabindex="-1"></a>             height<span class="op">=</span><span class="dv">400</span>,</span>
<span id="cb87-9"><a href="#cb87-9" aria-hidden="true" tabindex="-1"></a>              template<span class="op">=</span><span class="st">"simple_white"</span>,</span>
<span id="cb87-10"><a href="#cb87-10" aria-hidden="true" tabindex="-1"></a>              labels<span class="op">=</span><span class="bu">dict</span>(levels<span class="op">=</span><span class="st">"Level of GO hierarchy"</span>, auc<span class="op">=</span><span class="st">"AUROC"</span>)</span>
<span id="cb87-11"><a href="#cb87-11" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb87-12"><a href="#cb87-12" aria-hidden="true" tabindex="-1"></a>fig.update_traces(width<span class="op">=</span><span class="fl">0.9</span>)</span>
<span id="cb87-13"><a href="#cb87-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-14"><a href="#cb87-14" aria-hidden="true" tabindex="-1"></a>fig.add_shape( <span class="co"># add a horizontal "target" line</span></span>
<span id="cb87-15"><a href="#cb87-15" aria-hidden="true" tabindex="-1"></a>    <span class="bu">type</span><span class="op">=</span><span class="st">"line"</span>, line_color<span class="op">=</span><span class="st">"salmon"</span>, line_width<span class="op">=</span><span class="dv">3</span>, opacity<span class="op">=</span><span class="dv">1</span>, line_dash<span class="op">=</span><span class="st">"dot"</span>,</span>
<span id="cb87-16"><a href="#cb87-16" aria-hidden="true" tabindex="-1"></a>    x0<span class="op">=</span><span class="dv">0</span>, x1<span class="op">=</span><span class="dv">1</span>, xref<span class="op">=</span><span class="st">"paper"</span>, y0<span class="op">=</span><span class="fl">0.7</span>, y1<span class="op">=</span><span class="fl">0.7</span>, yref<span class="op">=</span><span class="st">"y"</span></span>
<span id="cb87-17"><a href="#cb87-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb87-18"><a href="#cb87-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-19"><a href="#cb87-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-20"><a href="#cb87-20" aria-hidden="true" tabindex="-1"></a>fig.update_layout(</span>
<span id="cb87-21"><a href="#cb87-21" aria-hidden="true" tabindex="-1"></a>   title<span class="op">=</span><span class="bu">dict</span>(text<span class="op">=</span><span class="st">"&lt;b&gt; AUC value grouped by level of GO hierarchy &lt;b&gt;"</span>,</span>
<span id="cb87-22"><a href="#cb87-22" aria-hidden="true" tabindex="-1"></a>             x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb87-23"><a href="#cb87-23" aria-hidden="true" tabindex="-1"></a>             y<span class="op">=</span><span class="fl">0.9</span>,</span>
<span id="cb87-24"><a href="#cb87-24" aria-hidden="true" tabindex="-1"></a>              font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">20</span>),</span>
<span id="cb87-25"><a href="#cb87-25" aria-hidden="true" tabindex="-1"></a>              xanchor<span class="op">=</span><span class="st">'center'</span>,</span>
<span id="cb87-26"><a href="#cb87-26" aria-hidden="true" tabindex="-1"></a>              yanchor<span class="op">=</span><span class="st">'top'</span>),</span>
<span id="cb87-27"><a href="#cb87-27" aria-hidden="true" tabindex="-1"></a>    xaxis<span class="op">=</span><span class="bu">dict</span>(ticks<span class="op">=</span><span class="st">""</span>, showticklabels<span class="op">=</span><span class="va">False</span>, showgrid<span class="op">=</span><span class="va">False</span>, zeroline<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb87-28"><a href="#cb87-28" aria-hidden="true" tabindex="-1"></a>    yaxis<span class="op">=</span><span class="bu">dict</span>(ticks<span class="op">=</span><span class="st">""</span>, showticklabels<span class="op">=</span><span class="va">True</span>, showgrid<span class="op">=</span><span class="va">True</span>, zeroline<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb87-29"><a href="#cb87-29" aria-hidden="true" tabindex="-1"></a> <span class="co">#   yaxis_range=[min(yy.flatten()),max(yy.flatten())],</span></span>
<span id="cb87-30"><a href="#cb87-30" aria-hidden="true" tabindex="-1"></a>  <span class="co">#  xaxis_range=[min(xx.flatten()),max(xx.flatten())],</span></span>
<span id="cb87-31"><a href="#cb87-31" aria-hidden="true" tabindex="-1"></a>    legend<span class="op">=</span><span class="bu">dict</span>(x<span class="op">=</span><span class="fl">1.1</span>, y<span class="op">=</span><span class="dv">1</span>, orientation<span class="op">=</span><span class="st">"v"</span>,font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">16</span>)),</span>
<span id="cb87-32"><a href="#cb87-32" aria-hidden="true" tabindex="-1"></a>    paper_bgcolor<span class="op">=</span><span class="st">'rgba(0,0,0,0)'</span>,</span>
<span id="cb87-33"><a href="#cb87-33" aria-hidden="true" tabindex="-1"></a>    plot_bgcolor<span class="op">=</span><span class="st">'rgba(0,0,0,0)'</span>,</span>
<span id="cb87-34"><a href="#cb87-34" aria-hidden="true" tabindex="-1"></a>    font<span class="op">=</span><span class="bu">dict</span>(family<span class="op">=</span><span class="st">'Roboto'</span>,color<span class="op">=</span> <span class="st">"#36382E"</span>,size<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb87-35"><a href="#cb87-35" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb87-36"><a href="#cb87-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-37"><a href="#cb87-37" aria-hidden="true" tabindex="-1"></a>fig.show()</span>
<span id="cb87-38"><a href="#cb87-38" aria-hidden="true" tabindex="-1"></a><span class="co"># pio.write_image(fig, resultsdir+"AUC_levels.png", width=600, height=400,scale=8)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Unable to display output for mime type(s): application/vnd.plotly.v1+json</code></pre>
</div>
</div>
</section>
<section id="top-15-predicted-go-terms" class="level2">
<h2 class="anchored" data-anchor-id="top-15-predicted-go-terms">TOP 15 PREDICTED GO TERMS</h2>
<div class="cell" data-execution_count="74">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>top15goterms<span class="op">=</span> np.array(GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">False</span>)[<span class="dv">0</span>:<span class="dv">15</span>].index)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get Top GO term names</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>top15goterms_1 <span class="op">=</span> []</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> goterm <span class="kw">in</span> top15goterms:</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>    top15goterms_1.append(goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(<span class="dv">1</span>))</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>real_go_info_mod_best <span class="op">=</span> real_go_info[real_go_info.GO_term.isin(top15goterms_1)]</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>real_go_info_mod_best.GO_term <span class="op">=</span> real_go_info_mod_best.GO_term.<span class="bu">str</span>.replace(<span class="st">"_1"</span>,<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="76">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>top15goterms_auc <span class="op">=</span> GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">False</span>)[<span class="dv">0</span>:<span class="dv">15</span>].reset_index()</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>top15goterms_auc.columns<span class="op">=</span>[<span class="st">"GO_term"</span>,<span class="st">"auc"</span>,<span class="st">"levels"</span>,<span class="st">"parents"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="77">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>top15goterms_auc <span class="op">=</span> top15goterms_auc.merge(real_go_info_mod_best[real_go_info_mod_best[<span class="st">"GO_term"</span>].isin(top15goterms)], on<span class="op">=</span><span class="st">"GO_term"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="78">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>top15goterms_auc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="78">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>auc</th>
      <th>levels</th>
      <th>parents</th>
      <th>Name</th>
      <th>layer_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>GO:1901029</td>
      <td>0.959991</td>
      <td>0</td>
      <td>10</td>
      <td>Negative regulation of mitochondrial outer membrane permeabilization involved in apoptotic signaling pathway (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GO:0044342</td>
      <td>0.956590</td>
      <td>1</td>
      <td>1</td>
      <td>Type b pancreatic cell proliferation (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GO:0002326</td>
      <td>0.955918</td>
      <td>0</td>
      <td>2</td>
      <td>B cell lineage commitment (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>GO:0007006</td>
      <td>0.955500</td>
      <td>3</td>
      <td>2</td>
      <td>Mitochondrial membrane organization (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>GO:1902236</td>
      <td>0.954857</td>
      <td>0</td>
      <td>6</td>
      <td>Negative regulation of endoplasmic reticulum stress-induced intrinsic apoptotic signaling pathway (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>GO:1900118</td>
      <td>0.953573</td>
      <td>0</td>
      <td>3</td>
      <td>Negative regulation of execution phase of apoptosis (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>GO:0030279</td>
      <td>0.950408</td>
      <td>1</td>
      <td>3</td>
      <td>Negative regulation of ossification (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>GO:0051452</td>
      <td>0.949592</td>
      <td>1</td>
      <td>1</td>
      <td>Intracellular ph reduction (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GO:0033033</td>
      <td>0.944694</td>
      <td>0</td>
      <td>3</td>
      <td>Negative regulation of myeloid cell apoptotic process (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GO:0060020</td>
      <td>0.944502</td>
      <td>0</td>
      <td>1</td>
      <td>Bergmann glial cell differentiation (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>10</th>
      <td>GO:0048743</td>
      <td>0.941837</td>
      <td>0</td>
      <td>11</td>
      <td>Positive regulation of skeletal muscle fiber development (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>11</th>
      <td>GO:0097345</td>
      <td>0.940950</td>
      <td>1</td>
      <td>5</td>
      <td>Mitochondrial outer membrane permeabilization (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>12</th>
      <td>GO:0006959</td>
      <td>0.940779</td>
      <td>3</td>
      <td>1</td>
      <td>Humoral immune response (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>13</th>
      <td>GO:1901863</td>
      <td>0.940703</td>
      <td>1</td>
      <td>3</td>
      <td>Positive regulation of muscle tissue development (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>14</th>
      <td>GO:0030890</td>
      <td>0.939581</td>
      <td>0</td>
      <td>9</td>
      <td>Positive regulation of b cell proliferation (1)</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="79">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> (<span class="dv">16</span>, <span class="dv">22</span>)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>rhos_top<span class="op">=</span>top15goterms_auc[<span class="st">"auc"</span>][<span class="dv">0</span>:<span class="dv">10</span>]</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>drugs_top<span class="op">=</span>top15goterms_auc[<span class="st">"GO_term"</span>][<span class="dv">0</span>:<span class="dv">10</span>]</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>colors <span class="op">=</span> [CB_color_cycle[<span class="dv">2</span>] <span class="cf">if</span> (x <span class="op">&lt;</span> <span class="fl">0.5</span>) <span class="cf">else</span> <span class="st">"#6492CA"</span> <span class="cf">for</span> x <span class="kw">in</span> rhos_top ]</span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>bars <span class="op">=</span> ax.bar(</span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>drugs_top,</span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a>    height<span class="op">=</span>rhos_top,</span>
<span id="cb94-10"><a href="#cb94-10" aria-hidden="true" tabindex="-1"></a>    edgecolor<span class="op">=</span><span class="st">"none"</span>,</span>
<span id="cb94-11"><a href="#cb94-11" aria-hidden="true" tabindex="-1"></a>    linewidth<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb94-12"><a href="#cb94-12" aria-hidden="true" tabindex="-1"></a>    color <span class="op">=</span> colors,</span>
<span id="cb94-13"><a href="#cb94-13" aria-hidden="true" tabindex="-1"></a>    width<span class="op">=</span><span class="fl">0.9</span></span>
<span id="cb94-14"><a href="#cb94-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb94-15"><a href="#cb94-15" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.yticks(fontsize=30)</span></span>
<span id="cb94-16"><a href="#cb94-16" aria-hidden="true" tabindex="-1"></a>plt.yticks([])</span>
<span id="cb94-17"><a href="#cb94-17" aria-hidden="true" tabindex="-1"></a>plt.xticks([])</span>
<span id="cb94-18"><a href="#cb94-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-19"><a href="#cb94-19" aria-hidden="true" tabindex="-1"></a><span class="co"># First, let's remove the top, right and left spines (figure borders)</span></span>
<span id="cb94-20"><a href="#cb94-20" aria-hidden="true" tabindex="-1"></a><span class="co"># which really aren't necessary for a bar chart.</span></span>
<span id="cb94-21"><a href="#cb94-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Also, make the bottom spine gray instead of black.</span></span>
<span id="cb94-22"><a href="#cb94-22" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb94-23"><a href="#cb94-23" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb94-24"><a href="#cb94-24" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'left'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb94-25"><a href="#cb94-25" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'bottom'</span>].set_color(<span class="st">'#DDDDDD'</span>)</span>
<span id="cb94-26"><a href="#cb94-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-27"><a href="#cb94-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-28"><a href="#cb94-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-29"><a href="#cb94-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Third, add a horizontal grid (but keep the vertical grid hidden).</span></span>
<span id="cb94-30"><a href="#cb94-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Color the lines a light gray as well.</span></span>
<span id="cb94-31"><a href="#cb94-31" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb94-32"><a href="#cb94-32" aria-hidden="true" tabindex="-1"></a><span class="co">#ax.yaxis.grid(False, color='#EEEEEE')</span></span>
<span id="cb94-33"><a href="#cb94-33" aria-hidden="true" tabindex="-1"></a>ax.xaxis.grid(<span class="va">False</span>)</span>
<span id="cb94-34"><a href="#cb94-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-35"><a href="#cb94-35" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.xticks(rotation=80,fontsize=40)</span></span>
<span id="cb94-36"><a href="#cb94-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-37"><a href="#cb94-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Add text annotations to the top of the bars.</span></span>
<span id="cb94-38"><a href="#cb94-38" aria-hidden="true" tabindex="-1"></a>bar_color <span class="op">=</span> bars[<span class="dv">0</span>].get_facecolor()</span>
<span id="cb94-39"><a href="#cb94-39" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bar <span class="kw">in</span> bars:</span>
<span id="cb94-40"><a href="#cb94-40" aria-hidden="true" tabindex="-1"></a>  ax.text(</span>
<span id="cb94-41"><a href="#cb94-41" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb94-42"><a href="#cb94-42" aria-hidden="true" tabindex="-1"></a>      bar.get_x() <span class="op">+</span> bar.get_width() <span class="op">/</span> <span class="dv">2</span>,</span>
<span id="cb94-43"><a href="#cb94-43" aria-hidden="true" tabindex="-1"></a>      bar.get_height() <span class="op">+</span> <span class="fl">0.03</span>,</span>
<span id="cb94-44"><a href="#cb94-44" aria-hidden="true" tabindex="-1"></a>      <span class="bu">round</span>(bar.get_height(), <span class="dv">3</span>),</span>
<span id="cb94-45"><a href="#cb94-45" aria-hidden="true" tabindex="-1"></a>      horizontalalignment<span class="op">=</span><span class="st">'center'</span>,</span>
<span id="cb94-46"><a href="#cb94-46" aria-hidden="true" tabindex="-1"></a>      <span class="co">#color=bar_color,</span></span>
<span id="cb94-47"><a href="#cb94-47" aria-hidden="true" tabindex="-1"></a>      color<span class="op">=</span><span class="st">'#000000'</span>,</span>
<span id="cb94-48"><a href="#cb94-48" aria-hidden="true" tabindex="-1"></a>      weight<span class="op">=</span><span class="st">'bold'</span>,</span>
<span id="cb94-49"><a href="#cb94-49" aria-hidden="true" tabindex="-1"></a>      fontsize<span class="op">=</span><span class="dv">80</span>,</span>
<span id="cb94-50"><a href="#cb94-50" aria-hidden="true" tabindex="-1"></a>      rotation<span class="op">=</span><span class="st">"vertical"</span></span>
<span id="cb94-51"><a href="#cb94-51" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb94-52"><a href="#cb94-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-53"><a href="#cb94-53" aria-hidden="true" tabindex="-1"></a>i<span class="op">=</span><span class="dv">0</span></span>
<span id="cb94-54"><a href="#cb94-54" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bar <span class="kw">in</span> bars:</span>
<span id="cb94-55"><a href="#cb94-55" aria-hidden="true" tabindex="-1"></a>    ax.text(</span>
<span id="cb94-56"><a href="#cb94-56" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb94-57"><a href="#cb94-57" aria-hidden="true" tabindex="-1"></a>      bar.get_x() <span class="op">+</span> bar.get_width() <span class="op">/</span> <span class="dv">2</span>,</span>
<span id="cb94-58"><a href="#cb94-58" aria-hidden="true" tabindex="-1"></a>      <span class="fl">0.45</span>,</span>
<span id="cb94-59"><a href="#cb94-59" aria-hidden="true" tabindex="-1"></a>      drugs_top[i],</span>
<span id="cb94-60"><a href="#cb94-60" aria-hidden="true" tabindex="-1"></a>      horizontalalignment<span class="op">=</span><span class="st">'center'</span>,</span>
<span id="cb94-61"><a href="#cb94-61" aria-hidden="true" tabindex="-1"></a>      <span class="co">#color=bar_color,</span></span>
<span id="cb94-62"><a href="#cb94-62" aria-hidden="true" tabindex="-1"></a>      color<span class="op">=</span><span class="st">'#000000'</span>,</span>
<span id="cb94-63"><a href="#cb94-63" aria-hidden="true" tabindex="-1"></a>      <span class="co">#weight='bold',</span></span>
<span id="cb94-64"><a href="#cb94-64" aria-hidden="true" tabindex="-1"></a>      fontsize<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb94-65"><a href="#cb94-65" aria-hidden="true" tabindex="-1"></a>      rotation<span class="op">=</span><span class="st">"vertical"</span></span>
<span id="cb94-66"><a href="#cb94-66" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb94-67"><a href="#cb94-67" aria-hidden="true" tabindex="-1"></a>    i<span class="op">=</span>i<span class="op">+</span><span class="dv">1</span></span>
<span id="cb94-68"><a href="#cb94-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-69"><a href="#cb94-69" aria-hidden="true" tabindex="-1"></a>ax.tick_params(bottom<span class="op">=</span><span class="va">True</span>, left<span class="op">=</span><span class="va">False</span>, axis<span class="op">=</span><span class="st">'x'</span>, which<span class="op">=</span><span class="st">'major'</span>, pad<span class="op">=-</span><span class="dv">1</span>)</span>
<span id="cb94-70"><a href="#cb94-70" aria-hidden="true" tabindex="-1"></a><span class="co"># Add labels and a title. Note the use of `labelpad` and `pad` to add some</span></span>
<span id="cb94-71"><a href="#cb94-71" aria-hidden="true" tabindex="-1"></a><span class="co"># extra space between the text and the tick labels.</span></span>
<span id="cb94-72"><a href="#cb94-72" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">''</span>, labelpad<span class="op">=</span><span class="dv">15</span>, color<span class="op">=</span><span class="st">'#333333'</span>)</span>
<span id="cb94-73"><a href="#cb94-73" aria-hidden="true" tabindex="-1"></a><span class="co">#ax.set_ylabel('r', labelpad=15, color='#333333',fontsize=30)</span></span>
<span id="cb94-74"><a href="#cb94-74" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="st">''</span>, color<span class="op">=</span><span class="st">'#333333'</span>,</span>
<span id="cb94-75"><a href="#cb94-75" aria-hidden="true" tabindex="-1"></a>             weight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb94-76"><a href="#cb94-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-77"><a href="#cb94-77" aria-hidden="true" tabindex="-1"></a><span class="co"># Make the chart fill out the figure better.</span></span>
<span id="cb94-78"><a href="#cb94-78" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb94-79"><a href="#cb94-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-80"><a href="#cb94-80" aria-hidden="true" tabindex="-1"></a>fig.savefig(resultsdir<span class="op">+</span><span class="st">'top10svm_models.png'</span>, transparent<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-80-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="worst-15-predicted-go-terms" class="level2">
<h2 class="anchored" data-anchor-id="worst-15-predicted-go-terms">WORST 15 PREDICTED GO TERMS</h2>
<div class="cell" data-execution_count="80">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>worst15goterms<span class="op">=</span> np.array(GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">True</span>)[<span class="dv">0</span>:<span class="dv">15</span>].index)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get Worst GO term names</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>worst15goterms_1 <span class="op">=</span> []</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> goterm <span class="kw">in</span> worst15goterms:</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>    worst15goterms_1.append(goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(<span class="dv">1</span>))</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>real_go_info_mod_worst <span class="op">=</span> real_go_info[real_go_info.GO_term.isin(worst15goterms_1)]</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>real_go_info_mod_worst.GO_term <span class="op">=</span> real_go_info_mod_worst.GO_term.<span class="bu">str</span>.replace(<span class="st">"_1"</span>,<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="85">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>worst15goterms_auc <span class="op">=</span> GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">True</span>)[<span class="dv">0</span>:<span class="dv">15</span>].reset_index()</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>worst15goterms_auc.columns<span class="op">=</span>[<span class="st">"GO_term"</span>,<span class="st">"auc"</span>,<span class="st">"levels"</span>,<span class="st">"parents"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="86">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>worst15goterms_auc.merge(real_go_info_mod_worst[real_go_info_mod_worst[<span class="st">"GO_term"</span>].isin(worst15goterms)], on<span class="op">=</span><span class="st">"GO_term"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="86">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>auc</th>
      <th>levels</th>
      <th>parents</th>
      <th>Name</th>
      <th>layer_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>GO:0032412</td>
      <td>0.351992</td>
      <td>3</td>
      <td>3</td>
      <td>Regulation of ion transmembrane transporter activity (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GO:0017157</td>
      <td>0.363878</td>
      <td>3</td>
      <td>5</td>
      <td>Regulation of exocytosis (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GO:0014013</td>
      <td>0.372925</td>
      <td>1</td>
      <td>2</td>
      <td>Regulation of gliogenesis (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>GO:0071356</td>
      <td>0.384013</td>
      <td>3</td>
      <td>2</td>
      <td>Cellular response to tumor necrosis factor (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>GO:0006163</td>
      <td>0.385537</td>
      <td>4</td>
      <td>2</td>
      <td>Purine nucleotide metabolic process (1)</td>
      <td>4.0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>GO:0048660</td>
      <td>0.389564</td>
      <td>2</td>
      <td>2</td>
      <td>Regulation of smooth muscle cell proliferation (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>GO:2000146</td>
      <td>0.390756</td>
      <td>5</td>
      <td>3</td>
      <td>Negative regulation of cell motility (1)</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>GO:0019827</td>
      <td>0.393045</td>
      <td>2</td>
      <td>1</td>
      <td>Stem cell population maintenance (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GO:0006475</td>
      <td>0.395102</td>
      <td>3</td>
      <td>1</td>
      <td>Internal protein amino acid acetylation (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GO:0015833</td>
      <td>0.395593</td>
      <td>5</td>
      <td>2</td>
      <td>Peptide transport (1)</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>10</th>
      <td>GO:0097305</td>
      <td>0.397744</td>
      <td>2</td>
      <td>1</td>
      <td>Response to alcohol (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>11</th>
      <td>GO:0060291</td>
      <td>0.402420</td>
      <td>2</td>
      <td>2</td>
      <td>Long-term synaptic potentiation (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>12</th>
      <td>GO:0032967</td>
      <td>0.406570</td>
      <td>0</td>
      <td>3</td>
      <td>Positive regulation of collagen biosynthetic process (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>13</th>
      <td>GO:0009743</td>
      <td>0.407096</td>
      <td>4</td>
      <td>1</td>
      <td>Response to carbohydrate (1)</td>
      <td>4.0</td>
    </tr>
    <tr>
      <th>14</th>
      <td>GO:0001659</td>
      <td>0.410349</td>
      <td>1</td>
      <td>2</td>
      <td>Temperature homeostasis (1)</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="aupr-histogram" class="level2">
<h2 class="anchored" data-anchor-id="aupr-histogram">AUPR histogram</h2>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="87">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(GO_terms_aupr_svm.items()),columns <span class="op">=</span> [<span class="st">'goterm'</span>,<span class="st">'aupr'</span>]).set_index(<span class="st">"goterm"</span>)</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df <span class="op">=</span> GO_terms_aupr_svm_df.dropna()</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"aupr"</span>], ascending<span class="op">=</span><span class="va">False</span>).head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="87">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>aupr</th>
    </tr>
    <tr>
      <th>goterm</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:0006468</th>
      <td>0.913629</td>
    </tr>
    <tr>
      <th>GO:0001782</th>
      <td>0.908636</td>
    </tr>
    <tr>
      <th>GO:0036211</th>
      <td>0.886811</td>
    </tr>
    <tr>
      <th>GO:0009058</th>
      <td>0.886773</td>
    </tr>
    <tr>
      <th>GO:0033033</th>
      <td>0.881536</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="88">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add number of parents</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>number_parents <span class="op">=</span> {}</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>levels <span class="op">=</span> {}</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(GO_terms_aupr_svm_df.index)):</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>    term <span class="op">=</span> GO_terms_aupr_svm_df.index[i]</span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>    number_parents[GO_terms_aupr_svm_df.index[i]]<span class="op">=</span><span class="bu">len</span>([source <span class="cf">for</span> source, _ <span class="kw">in</span>  dG.in_edges(term)])</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>    levels[GO_terms_aupr_svm_df.index[i]]<span class="op">=</span>level_number[term]<span class="op">-</span><span class="dv">1</span></span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>levels <span class="op">=</span> pd.DataFrame.from_dict(levels, orient<span class="op">=</span><span class="st">'index'</span>)</span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a>number_parents <span class="op">=</span> pd.DataFrame.from_dict(number_parents, orient<span class="op">=</span><span class="st">'index'</span>)</span>
<span id="cb100-10"><a href="#cb100-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-11"><a href="#cb100-11" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df <span class="op">=</span> pd.concat([GO_terms_aupr_svm_df, levels,number_parents], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb100-12"><a href="#cb100-12" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df.columns <span class="op">=</span> [<span class="st">"aupr"</span>,<span class="st">"levels"</span>,<span class="st">"parents"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="89">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df <span class="op">=</span> GO_terms_aupr_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"levels"</span>], ascending<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="91">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>c <span class="op">=</span> [<span class="st">'#E8384F'</span>, <span class="st">'#FD817D'</span>, <span class="st">'#FDAE33'</span>,<span class="st">'#EECC16'</span>, <span class="st">'#A4C61A'</span>, <span class="st">'#37A862'</span>,<span class="st">"#208EA3"</span>,<span class="st">"#3B6EAB"</span>]</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> px.data.tips()</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> px.box(GO_terms_aupr_svm_df, x<span class="op">=</span><span class="st">"levels"</span>, y<span class="op">=</span><span class="st">"aupr"</span>,</span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>             color<span class="op">=</span><span class="st">"levels"</span>,</span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>            color_discrete_sequence<span class="op">=</span>c,</span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>             width <span class="op">=</span><span class="dv">600</span>,</span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>             height<span class="op">=</span><span class="dv">400</span>,</span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a>              template<span class="op">=</span><span class="st">"simple_white"</span>,</span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a>              labels<span class="op">=</span><span class="bu">dict</span>(levels<span class="op">=</span><span class="st">"Level of GO hierarchy"</span>, aupr<span class="op">=</span><span class="st">"AUPR"</span>)</span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb102-13"><a href="#cb102-13" aria-hidden="true" tabindex="-1"></a>fig.update_traces(width<span class="op">=</span><span class="fl">0.9</span>)</span>
<span id="cb102-14"><a href="#cb102-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-15"><a href="#cb102-15" aria-hidden="true" tabindex="-1"></a>fig.add_shape( <span class="co"># add a horizontal "target" line</span></span>
<span id="cb102-16"><a href="#cb102-16" aria-hidden="true" tabindex="-1"></a>    <span class="bu">type</span><span class="op">=</span><span class="st">"line"</span>, line_color<span class="op">=</span><span class="st">"salmon"</span>, line_width<span class="op">=</span><span class="dv">3</span>, opacity<span class="op">=</span><span class="dv">1</span>, line_dash<span class="op">=</span><span class="st">"dot"</span>,</span>
<span id="cb102-17"><a href="#cb102-17" aria-hidden="true" tabindex="-1"></a>    x0<span class="op">=</span><span class="dv">0</span>, x1<span class="op">=</span><span class="dv">1</span>, xref<span class="op">=</span><span class="st">"paper"</span>, y0<span class="op">=</span><span class="fl">0.7</span>, y1<span class="op">=</span><span class="fl">0.7</span>, yref<span class="op">=</span><span class="st">"y"</span></span>
<span id="cb102-18"><a href="#cb102-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb102-19"><a href="#cb102-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-20"><a href="#cb102-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-21"><a href="#cb102-21" aria-hidden="true" tabindex="-1"></a>fig.update_layout(</span>
<span id="cb102-22"><a href="#cb102-22" aria-hidden="true" tabindex="-1"></a>   title<span class="op">=</span><span class="bu">dict</span>(text<span class="op">=</span><span class="st">"&lt;b&gt; AUPR value grouped by level of GO hierarchy &lt;b&gt;"</span>,</span>
<span id="cb102-23"><a href="#cb102-23" aria-hidden="true" tabindex="-1"></a>             x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb102-24"><a href="#cb102-24" aria-hidden="true" tabindex="-1"></a>             y<span class="op">=</span><span class="fl">0.9</span>,</span>
<span id="cb102-25"><a href="#cb102-25" aria-hidden="true" tabindex="-1"></a>              font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">20</span>),</span>
<span id="cb102-26"><a href="#cb102-26" aria-hidden="true" tabindex="-1"></a>              xanchor<span class="op">=</span><span class="st">'center'</span>,</span>
<span id="cb102-27"><a href="#cb102-27" aria-hidden="true" tabindex="-1"></a>              yanchor<span class="op">=</span><span class="st">'top'</span>),</span>
<span id="cb102-28"><a href="#cb102-28" aria-hidden="true" tabindex="-1"></a>    xaxis<span class="op">=</span><span class="bu">dict</span>(ticks<span class="op">=</span><span class="st">""</span>, showticklabels<span class="op">=</span><span class="va">False</span>, showgrid<span class="op">=</span><span class="va">False</span>, zeroline<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb102-29"><a href="#cb102-29" aria-hidden="true" tabindex="-1"></a>    yaxis<span class="op">=</span><span class="bu">dict</span>(ticks<span class="op">=</span><span class="st">""</span>, showticklabels<span class="op">=</span><span class="va">True</span>, showgrid<span class="op">=</span><span class="va">True</span>, zeroline<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb102-30"><a href="#cb102-30" aria-hidden="true" tabindex="-1"></a> <span class="co">#   yaxis_range=[min(yy.flatten()),max(yy.flatten())],</span></span>
<span id="cb102-31"><a href="#cb102-31" aria-hidden="true" tabindex="-1"></a>  <span class="co">#  xaxis_range=[min(xx.flatten()),max(xx.flatten())],</span></span>
<span id="cb102-32"><a href="#cb102-32" aria-hidden="true" tabindex="-1"></a>    legend<span class="op">=</span><span class="bu">dict</span>(x<span class="op">=</span><span class="fl">1.1</span>, y<span class="op">=</span><span class="dv">1</span>, orientation<span class="op">=</span><span class="st">"v"</span>,font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">16</span>)),</span>
<span id="cb102-33"><a href="#cb102-33" aria-hidden="true" tabindex="-1"></a>    paper_bgcolor<span class="op">=</span><span class="st">'rgba(0,0,0,0)'</span>,</span>
<span id="cb102-34"><a href="#cb102-34" aria-hidden="true" tabindex="-1"></a>    plot_bgcolor<span class="op">=</span><span class="st">'rgba(0,0,0,0)'</span>,</span>
<span id="cb102-35"><a href="#cb102-35" aria-hidden="true" tabindex="-1"></a>    font<span class="op">=</span><span class="bu">dict</span>(family<span class="op">=</span><span class="st">'Roboto'</span>,color<span class="op">=</span> <span class="st">"#36382E"</span>,size<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb102-36"><a href="#cb102-36" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb102-37"><a href="#cb102-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-38"><a href="#cb102-38" aria-hidden="true" tabindex="-1"></a>fig.show()</span>
<span id="cb102-39"><a href="#cb102-39" aria-hidden="true" tabindex="-1"></a><span class="co"># pio.write_image(fig, resultsdir+"AUPR_levels.png", width=600, height=400,scale=8)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Unable to display output for mime type(s): application/vnd.plotly.v1+json</code></pre>
</div>
</div>
</section>
<section id="example-prediction" class="level2">
<h2 class="anchored" data-anchor-id="example-prediction">Example prediction</h2>
<div class="cell" data-execution_count="92">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> f2(goterm):    </span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> goterm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="93">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>combobox_go <span class="op">=</span> interactive(f2, goterm<span class="op">=</span>widgets.Combobox(options<span class="op">=</span><span class="bu">list</span>(GO_terms_auc_svm_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">False</span>).index)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Choose drug to study…</strong></p>
<div class="cell" data-execution_count="94">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>display(combobox_go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"403eb665201541cf8fcc0765d10235ec","version_major":2,"version_minor":0}
</script>
</div>
</div>
<div class="cell" data-execution_count="95">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>selected_go <span class="op">=</span> combobox_go.result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="96">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co">#auc</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> (<span class="dv">4</span>, <span class="dv">2</span>)</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>fpr, tpr, _ <span class="op">=</span> metrics.roc_curve(slim_matrix_single_neuron.loc[selected_go],  platt_matrix.loc[selected_go])</span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>auc <span class="op">=</span> metrics.roc_auc_score(slim_matrix_single_neuron.loc[selected_go],  platt_matrix.loc[selected_go])</span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>plt.plot(fpr,tpr,label<span class="op">=</span><span class="st">"data 1, auc="</span><span class="op">+</span><span class="bu">str</span>(auc))</span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb108-8"><a href="#cb108-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-9"><a href="#cb108-9" aria-hidden="true" tabindex="-1"></a>plot <span class="op">=</span> pd.concat([pd.DataFrame(slim_matrix_single_neuron.loc[selected_go]),pd.DataFrame(platt_matrix.loc[selected_go])], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb108-10"><a href="#cb108-10" aria-hidden="true" tabindex="-1"></a>plot.columns <span class="op">=</span> [<span class="st">"slim"</span>,<span class="st">"probability"</span>]</span>
<span id="cb108-11"><a href="#cb108-11" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(x<span class="op">=</span><span class="st">"slim"</span>, y<span class="op">=</span><span class="st">"probability"</span>, data<span class="op">=</span>plot,showfliers<span class="op">=</span><span class="va">False</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-93-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-93-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="97">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co">#auc </span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>fpr, tpr, _ <span class="op">=</span> metrics.roc_curve(slim_matrix_single_neuron.loc[selected_go],  delta_logits_matrix.loc[selected_go])</span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>auc <span class="op">=</span> metrics.roc_auc_score(slim_matrix_single_neuron.loc[selected_go],  delta_logits_matrix.loc[selected_go])</span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>plt.plot(fpr,tpr,label<span class="op">=</span><span class="st">"data 1, auc="</span><span class="op">+</span><span class="bu">str</span>(auc))</span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb109-7"><a href="#cb109-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-8"><a href="#cb109-8" aria-hidden="true" tabindex="-1"></a>plot <span class="op">=</span> pd.concat([pd.DataFrame(slim_matrix_single_neuron.loc[selected_go]),pd.DataFrame(delta_logits_matrix.loc[selected_go])], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb109-9"><a href="#cb109-9" aria-hidden="true" tabindex="-1"></a>plot.columns <span class="op">=</span> [<span class="st">"slim"</span>,<span class="st">"probability"</span>]</span>
<span id="cb109-10"><a href="#cb109-10" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(x<span class="op">=</span><span class="st">"slim"</span>, y<span class="op">=</span><span class="st">"probability"</span>, data<span class="op">=</span>plot,showfliers<span class="op">=</span><span class="va">False</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-94-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-94-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="98">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> (<span class="dv">2</span>, <span class="dv">2</span>)</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>metrics.ConfusionMatrixDisplay.from_predictions(slim_matrix_single_neuron.loc[selected_go], preds_svm_matrix.loc[selected_go])</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>plt.grid(visible<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Accuracy:"</span>,metrics.accuracy_score(slim_matrix_single_neuron.loc[selected_go], preds_svm_matrix.loc[selected_go]))</span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Precision:"</span>,metrics.precision_score(slim_matrix_single_neuron.loc[selected_go], preds_svm_matrix.loc[selected_go]))</span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Recall:"</span>,metrics.recall_score(slim_matrix_single_neuron.loc[selected_go], preds_svm_matrix.loc[selected_go])) <span class="co">#TP / (TP+FN)</span></span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"AUC with score:"</span>,auc) <span class="co">#TP / (TP+FN)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Accuracy: 0.9735849056603774
Precision: 0.7894736842105263
Recall: 0.8333333333333334
AUC with score: 0.9565901934322987</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-95-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>TN - FP</p>
<p>FN - TP</p>
<div class="cell" data-execution_count="99">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> (<span class="dv">4</span>, <span class="dv">2</span>)</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>precision, recall, thresholds <span class="op">=</span> metrics.precision_recall_curve(slim_matrix_single_neuron.loc[selected_go],  preds_svm_matrix.loc[selected_go])</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>auc_precision_recall <span class="op">=</span> metrics.auc(recall, precision)</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>plt.plot(recall, precision,label<span class="op">=</span><span class="bu">str</span>(auc_precision_recall))</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-96-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="metrics-drugs" class="level2">
<h2 class="anchored" data-anchor-id="metrics-drugs">METRICS drugs</h2>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="100">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>auc_drugs <span class="op">=</span> {}</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>aupr_drugs <span class="op">=</span> {}</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>precision_drugs <span class="op">=</span> {}</span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> drug <span class="kw">in</span> <span class="bu">list</span>(slim_matrix_single_neuron.columns):</span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> slim_matrix_single_neuron.loc[:,drug].<span class="bu">sum</span>() <span class="op">==</span><span class="dv">0</span>:</span>
<span id="cb113-6"><a href="#cb113-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb113-7"><a href="#cb113-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">#fpr, tpr, _ = metrics.roc_curve(slim_matrix_single_neuron.loc[:,drug], logits_matrix.loc[:,drug])</span></span>
<span id="cb113-8"><a href="#cb113-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">#auc_drugs[drug]  = metrics.auc(fpr, tpr)</span></span>
<span id="cb113-9"><a href="#cb113-9" aria-hidden="true" tabindex="-1"></a>    auc_drugs[drug] <span class="op">=</span> metrics.roc_auc_score(slim_matrix_single_neuron.loc[:,drug],  platt_matrix.loc[:,drug])</span>
<span id="cb113-10"><a href="#cb113-10" aria-hidden="true" tabindex="-1"></a>    precision, recall, thresholds <span class="op">=</span> metrics.precision_recall_curve(slim_matrix_single_neuron.loc[:,drug],  platt_matrix.loc[:,drug])</span>
<span id="cb113-11"><a href="#cb113-11" aria-hidden="true" tabindex="-1"></a>    aupr_drugs[drug] <span class="op">=</span> metrics.auc(recall, precision)</span>
<span id="cb113-12"><a href="#cb113-12" aria-hidden="true" tabindex="-1"></a>    precision_drugs[drug] <span class="op">=</span> metrics.precision_score(slim_matrix_single_neuron.loc[:,drug],  preds_svm_matrix.loc[:,drug])</span>
<span id="cb113-13"><a href="#cb113-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-14"><a href="#cb113-14" aria-hidden="true" tabindex="-1"></a>auc_drugs_df <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(auc_drugs.items()),columns <span class="op">=</span> [<span class="st">'goterm'</span>,<span class="st">'auc'</span>]).set_index(<span class="st">"goterm"</span>)</span>
<span id="cb113-15"><a href="#cb113-15" aria-hidden="true" tabindex="-1"></a>auc_drugs_df <span class="op">=</span> auc_drugs_df.dropna()</span>
<span id="cb113-16"><a href="#cb113-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-17"><a href="#cb113-17" aria-hidden="true" tabindex="-1"></a>aupr_drugs_df <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(aupr_drugs.items()),columns <span class="op">=</span> [<span class="st">'goterm'</span>,<span class="st">'aupr'</span>]).set_index(<span class="st">"goterm"</span>)</span>
<span id="cb113-18"><a href="#cb113-18" aria-hidden="true" tabindex="-1"></a>aupr_drugs_df <span class="op">=</span> aupr_drugs_df.dropna()</span>
<span id="cb113-19"><a href="#cb113-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-20"><a href="#cb113-20" aria-hidden="true" tabindex="-1"></a>precision_drugs_df <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(precision_drugs.items()),columns <span class="op">=</span> [<span class="st">'goterm'</span>,<span class="st">'precision'</span>]).set_index(<span class="st">"goterm"</span>)</span>
<span id="cb113-21"><a href="#cb113-21" aria-hidden="true" tabindex="-1"></a>precision_drugs_df <span class="op">=</span> precision_drugs_df.dropna()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="auc-histogram-drugs" class="level3">
<h3 class="anchored" data-anchor-id="auc-histogram-drugs">AUC histogram drugs</h3>
<div class="cell" data-execution_count="101">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">'figure.figsize'</span>:(<span class="dv">10</span>,<span class="dv">6</span>)})</span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>perc <span class="op">=</span> <span class="bu">str</span>(<span class="bu">round</span>((<span class="dv">100</span><span class="op">*</span><span class="bu">len</span>(auc_drugs_df[auc_drugs_df[<span class="st">"auc"</span>]<span class="op">&gt;</span><span class="fl">0.7</span>])<span class="op">/</span><span class="bu">len</span>(auc_drugs_df)),<span class="dv">2</span>))<span class="op">+</span><span class="st">"%"</span></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a>N, bins, patches <span class="op">=</span> plt.hist(auc_drugs_df, color<span class="op">=</span>CB_color_cycle[<span class="dv">6</span>],bins<span class="op">=</span><span class="dv">50</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(bins)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb114-7"><a href="#cb114-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> bins[i]<span class="op">&gt;</span><span class="fl">0.7</span>:</span>
<span id="cb114-8"><a href="#cb114-8" aria-hidden="true" tabindex="-1"></a>        patches[i].set_facecolor(CB_color_cycle[<span class="dv">5</span>])</span>
<span id="cb114-9"><a href="#cb114-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-10"><a href="#cb114-10" aria-hidden="true" tabindex="-1"></a>plt.yticks(fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb114-11"><a href="#cb114-11" aria-hidden="true" tabindex="-1"></a>plt.xticks(fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb114-12"><a href="#cb114-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-13"><a href="#cb114-13" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb114-14"><a href="#cb114-14" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb114-15"><a href="#cb114-15" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'left'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb114-16"><a href="#cb114-16" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'bottom'</span>].set_color(<span class="st">'#DDDDDD'</span>)</span>
<span id="cb114-17"><a href="#cb114-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-18"><a href="#cb114-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Second, remove the ticks as well.</span></span>
<span id="cb114-19"><a href="#cb114-19" aria-hidden="true" tabindex="-1"></a>ax.tick_params(bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb114-20"><a href="#cb114-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-21"><a href="#cb114-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Third, add a horizontal grid (but keep the vertical grid hidden).</span></span>
<span id="cb114-22"><a href="#cb114-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Color the lines a light gray as well.</span></span>
<span id="cb114-23"><a href="#cb114-23" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb114-24"><a href="#cb114-24" aria-hidden="true" tabindex="-1"></a>ax.yaxis.grid(<span class="va">True</span>, color<span class="op">=</span><span class="st">'#EEEEEE'</span>)</span>
<span id="cb114-25"><a href="#cb114-25" aria-hidden="true" tabindex="-1"></a>ax.xaxis.grid(<span class="va">False</span>)</span>
<span id="cb114-26"><a href="#cb114-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-27"><a href="#cb114-27" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"AUC value"</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb114-28"><a href="#cb114-28" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Number of drugs"</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb114-29"><a href="#cb114-29" aria-hidden="true" tabindex="-1"></a>colors2 <span class="op">=</span> {<span class="st">'Drugs with AUC&gt;=0.7'</span>:CB_color_cycle[<span class="dv">5</span>]}  </span>
<span id="cb114-30"><a href="#cb114-30" aria-hidden="true" tabindex="-1"></a>labels <span class="op">=</span> <span class="bu">list</span>(colors2.keys())</span>
<span id="cb114-31"><a href="#cb114-31" aria-hidden="true" tabindex="-1"></a>handles <span class="op">=</span> [plt.Rectangle((<span class="dv">0</span>,<span class="dv">0</span>),<span class="dv">1</span>,<span class="dv">1</span>, color<span class="op">=</span>colors2[label]) <span class="cf">for</span> label <span class="kw">in</span> labels]</span>
<span id="cb114-32"><a href="#cb114-32" aria-hidden="true" tabindex="-1"></a>plt.legend(handles, labels,fontsize<span class="op">=</span><span class="dv">20</span>, loc<span class="op">=</span><span class="st">"lower left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">0.35</span>,<span class="op">-</span><span class="fl">0.35</span>))</span>
<span id="cb114-33"><a href="#cb114-33" aria-hidden="true" tabindex="-1"></a>plt.text(<span class="fl">0.79</span>, <span class="dv">6</span>, <span class="bu">str</span>(perc), fontsize<span class="op">=</span><span class="dv">20</span>,color<span class="op">=</span><span class="st">'#333333'</span>)</span>
<span id="cb114-34"><a href="#cb114-34" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Overall performance by drugs using mutations"</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span>
<span id="cb114-35"><a href="#cb114-35" aria-hidden="true" tabindex="-1"></a><span class="co"># con el que mejor funciona es con la suma normal del attribution </span></span>
<span id="cb114-36"><a href="#cb114-36" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb114-37"><a href="#cb114-37" aria-hidden="true" tabindex="-1"></a>fig.savefig(resultsdir<span class="op">+</span><span class="st">'drugsAUC.png'</span>, transparent<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-98-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="auc-waterfall-plot-drugs" class="level3">
<h3 class="anchored" data-anchor-id="auc-waterfall-plot-drugs">AUC waterfall plot drugs</h3>
<div class="cell" data-execution_count="102">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>auc_drugs_df <span class="op">=</span>auc_drugs_df.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="103">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> (<span class="dv">12</span>, <span class="dv">9</span>)</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>drugs <span class="op">=</span> auc_drugs_df.index</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>rhos <span class="op">=</span> auc_drugs_df[<span class="st">"auc"</span>]</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>percentage <span class="op">=</span> <span class="bu">round</span>((<span class="bu">sum</span>(rhos<span class="op">&gt;</span><span class="fl">0.69</span>)<span class="op">/</span><span class="bu">len</span>(rhos))<span class="op">*</span><span class="dv">100</span>,<span class="dv">1</span>)</span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a><span class="co">#colors = ['#208EA3' if (x &lt; 0.5) else '#A4C61A' for x in rhos ]</span></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>colors <span class="op">=</span> [<span class="st">'#C9C9C9'</span> <span class="cf">if</span> (x <span class="op">&lt;</span> <span class="fl">0.69</span>) <span class="cf">else</span> <span class="st">"#B678BE"</span> <span class="cf">for</span> x <span class="kw">in</span> rhos ]</span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>ax.bar(</span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>drugs,</span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>    height<span class="op">=</span>rhos,</span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a>    edgecolor<span class="op">=</span>colors,</span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>    linewidth<span class="op">=</span><span class="dv">3</span></span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a>plt.xticks([])</span>
<span id="cb116-17"><a href="#cb116-17" aria-hidden="true" tabindex="-1"></a>plt.yticks(fontsize<span class="op">=</span><span class="dv">28</span>)</span>
<span id="cb116-18"><a href="#cb116-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-19"><a href="#cb116-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-20"><a href="#cb116-20" aria-hidden="true" tabindex="-1"></a><span class="co"># First, let's remove the top, right and left spines (figure borders)</span></span>
<span id="cb116-21"><a href="#cb116-21" aria-hidden="true" tabindex="-1"></a><span class="co"># which really aren't necessary for a bar chart.</span></span>
<span id="cb116-22"><a href="#cb116-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Also, make the bottom spine gray instead of black.</span></span>
<span id="cb116-23"><a href="#cb116-23" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb116-24"><a href="#cb116-24" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb116-25"><a href="#cb116-25" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'left'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb116-26"><a href="#cb116-26" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'bottom'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb116-27"><a href="#cb116-27" aria-hidden="true" tabindex="-1"></a><span class="co">#ax.spines['bottom'].set_color('#DDDDDD')</span></span>
<span id="cb116-28"><a href="#cb116-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-29"><a href="#cb116-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Second, remove the ticks as well.</span></span>
<span id="cb116-30"><a href="#cb116-30" aria-hidden="true" tabindex="-1"></a>ax.tick_params(bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb116-31"><a href="#cb116-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-32"><a href="#cb116-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Third, add a horizontal grid (but keep the vertical grid hidden).</span></span>
<span id="cb116-33"><a href="#cb116-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Color the lines a light gray as well.</span></span>
<span id="cb116-34"><a href="#cb116-34" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">False</span>)</span>
<span id="cb116-35"><a href="#cb116-35" aria-hidden="true" tabindex="-1"></a>ax.yaxis.grid(<span class="va">False</span>)</span>
<span id="cb116-36"><a href="#cb116-36" aria-hidden="true" tabindex="-1"></a><span class="co">#ax.yaxis.grid(True, color='#EEEEEE')</span></span>
<span id="cb116-37"><a href="#cb116-37" aria-hidden="true" tabindex="-1"></a>ax.xaxis.grid(<span class="va">False</span>)</span>
<span id="cb116-38"><a href="#cb116-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-39"><a href="#cb116-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-40"><a href="#cb116-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Add labels and a title. Note the use of `labelpad` and `pad` to add some</span></span>
<span id="cb116-41"><a href="#cb116-41" aria-hidden="true" tabindex="-1"></a><span class="co"># extra space between the text and the tick labels.</span></span>
<span id="cb116-42"><a href="#cb116-42" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'Drugs'</span>, labelpad<span class="op">=-</span><span class="dv">30</span>, color<span class="op">=</span><span class="st">'#333333'</span>,fontsize<span class="op">=</span><span class="dv">50</span>)</span>
<span id="cb116-43"><a href="#cb116-43" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'AUC-ROC value'</span>, labelpad<span class="op">=</span><span class="dv">15</span>, color<span class="op">=</span><span class="st">'#333333'</span>,fontsize<span class="op">=</span><span class="dv">50</span>)</span>
<span id="cb116-44"><a href="#cb116-44" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="st">''</span>, color<span class="op">=</span><span class="st">'#333333'</span>,</span>
<span id="cb116-45"><a href="#cb116-45" aria-hidden="true" tabindex="-1"></a>             weight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb116-46"><a href="#cb116-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-47"><a href="#cb116-47" aria-hidden="true" tabindex="-1"></a>colors2 <span class="op">=</span> {<span class="st">'High confidence drugs (r&gt;0.5)'</span>:<span class="st">'#A4C61A'</span>}  </span>
<span id="cb116-48"><a href="#cb116-48" aria-hidden="true" tabindex="-1"></a>labels <span class="op">=</span> <span class="bu">list</span>(colors2.keys())</span>
<span id="cb116-49"><a href="#cb116-49" aria-hidden="true" tabindex="-1"></a>handles <span class="op">=</span> [plt.Rectangle((<span class="dv">0</span>,<span class="dv">0</span>),<span class="dv">1</span>,<span class="dv">1</span>, color<span class="op">=</span>colors2[label]) <span class="cf">for</span> label <span class="kw">in</span> labels]</span>
<span id="cb116-50"><a href="#cb116-50" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.legend(handles, labels,fontsize=40, loc="lower left",bbox_to_anchor=(0, -0.215))</span></span>
<span id="cb116-51"><a href="#cb116-51" aria-hidden="true" tabindex="-1"></a>plt.text(<span class="dv">77</span>, <span class="fl">0.32</span>, <span class="bu">str</span>(percentage)<span class="op">+</span><span class="st">"%"</span>, fontsize<span class="op">=</span><span class="dv">60</span>,color<span class="op">=</span><span class="st">'#000000'</span>)</span>
<span id="cb116-52"><a href="#cb116-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-53"><a href="#cb116-53" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="fl">0.1</span>,<span class="fl">1.1</span>))</span>
<span id="cb116-54"><a href="#cb116-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Make the chart fill out the figure better.</span></span>
<span id="cb116-55"><a href="#cb116-55" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb116-56"><a href="#cb116-56" aria-hidden="true" tabindex="-1"></a>fig.savefig(resultsdir<span class="op">+</span><span class="st">'WaterfallModelsSVM_drugs.png'</span>, transparent<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-100-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="aupr-histogram-drugs" class="level3">
<h3 class="anchored" data-anchor-id="aupr-histogram-drugs">AUPR histogram drugs</h3>
<div class="cell" data-execution_count="104">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">'figure.figsize'</span>:(<span class="dv">5</span>,<span class="dv">3</span>)})</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>perc <span class="op">=</span> <span class="bu">str</span>(<span class="bu">round</span>((<span class="dv">100</span><span class="op">*</span><span class="bu">len</span>(aupr_drugs_df[aupr_drugs_df[<span class="st">"aupr"</span>]<span class="op">&gt;</span><span class="fl">0.69</span>])<span class="op">/</span><span class="bu">len</span>(aupr_drugs_df)),<span class="dv">2</span>))<span class="op">+</span><span class="st">"%"</span></span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>N, bins, patches <span class="op">=</span> plt.hist(aupr_drugs_df, color<span class="op">=</span>CB_color_cycle[<span class="dv">6</span>],bins<span class="op">=</span><span class="dv">50</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(bins)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb117-5"><a href="#cb117-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> bins[i]<span class="op">&gt;</span><span class="fl">0.69</span>:</span>
<span id="cb117-6"><a href="#cb117-6" aria-hidden="true" tabindex="-1"></a>        patches[i].set_facecolor(CB_color_cycle[<span class="dv">3</span>])</span>
<span id="cb117-7"><a href="#cb117-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-8"><a href="#cb117-8" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"AUPR drugs"</span>, fontsize<span class="op">=</span><span class="dv">16</span>)  </span>
<span id="cb117-9"><a href="#cb117-9" aria-hidden="true" tabindex="-1"></a>plt.title(perc, fontsize<span class="op">=</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="104">
<pre><code>Text(0.5, 1.0, '33.58%')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-101-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="example-drug-prediction" class="level3">
<h3 class="anchored" data-anchor-id="example-drug-prediction">Example drug prediction</h3>
<div class="cell" data-execution_count="105">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> f(drug):    </span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> drug</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="106">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>predictions_nodes <span class="op">=</span> []</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> goterm <span class="kw">in</span> <span class="bu">list</span>(platt_matrix.index):</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>    predictions_nodes.append(goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add names to go terms</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>real_go_info_svm<span class="op">=</span> real_go_info[real_go_info.GO_term.isin(predictions_nodes)]</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>real_go_info_svm.GO_term <span class="op">=</span> real_go_info_svm.GO_term.<span class="bu">str</span>.replace(<span class="st">"_1"</span>,<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>combobox <span class="op">=</span> interactive(f, drug<span class="op">=</span>widgets.Combobox(options<span class="op">=</span><span class="bu">list</span>(precision_drugs_df.sort_values(by<span class="op">=</span>[<span class="st">"precision"</span>], ascending<span class="op">=</span><span class="va">False</span>).index)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Choose drug to study…</strong></p>
<div class="cell" data-execution_count="198">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>display(combobox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"6de302386d11478db9c9cc2bb80cf0f5","version_major":2,"version_minor":0}
</script>
</div>
</div>
<div class="cell" data-execution_count="202">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>selected_drug_name <span class="op">=</span> combobox.result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="206">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">'figure.figsize'</span>:(<span class="dv">4</span>,<span class="dv">2</span>)})</span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a><span class="co">#auc</span></span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>fpr, tpr, _ <span class="op">=</span> metrics.roc_curve(slim_matrix_single_neuron.loc[:,selected_drug_name], platt_matrix.loc[:,selected_drug_name] )</span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>auc <span class="op">=</span> metrics.roc_auc_score(slim_matrix_single_neuron.loc[:,selected_drug_name],  platt_matrix.loc[:,selected_drug_name])</span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>plt.plot(fpr,tpr,label<span class="op">=</span><span class="st">"data 1, auc="</span><span class="op">+</span><span class="bu">str</span>(auc))</span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb125-7"><a href="#cb125-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb125-8"><a href="#cb125-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb125-9"><a href="#cb125-9" aria-hidden="true" tabindex="-1"></a>plot <span class="op">=</span> pd.concat([pd.DataFrame(slim_matrix_single_neuron.loc[:,selected_drug_name]),pd.DataFrame(platt_matrix.loc[:,selected_drug_name])], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb125-10"><a href="#cb125-10" aria-hidden="true" tabindex="-1"></a>plot.columns <span class="op">=</span> [<span class="st">"slim"</span>,<span class="st">"svm score"</span>]</span>
<span id="cb125-11"><a href="#cb125-11" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(x<span class="op">=</span><span class="st">"slim"</span>, y<span class="op">=</span><span class="st">"svm score"</span>, data<span class="op">=</span>plot,showfliers<span class="op">=</span><span class="va">False</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-108-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-108-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="207">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>plot <span class="op">=</span> pd.concat([pd.DataFrame(slim_matrix.loc[:,selected_drug_name]),pd.DataFrame(attribution_data_annotated.loc[:,selected_drug_name]<span class="op">*</span><span class="fl">1e4</span>)], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>plot.columns <span class="op">=</span> [<span class="st">"slim"</span>,<span class="st">"attribution"</span>]</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(x<span class="op">=</span><span class="st">"slim"</span>, y<span class="op">=</span><span class="st">"attribution"</span>, data<span class="op">=</span>plot,showfliers<span class="op">=</span><span class="va">True</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-109-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="208">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>metrics.ConfusionMatrixDisplay.from_predictions(slim_matrix_single_neuron.loc[:,selected_drug_name].<span class="bu">round</span>(), preds_svm_matrix.loc[:,selected_drug_name])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="208">
<pre><code>&lt;sklearn.metrics._plot.confusion_matrix.ConfusionMatrixDisplay at 0x2ce2b82e0&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-110-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="209">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Accuracy:"</span>,metrics.accuracy_score(slim_matrix_single_neuron.loc[:,selected_drug_name], preds_svm_matrix.loc[:,selected_drug_name]))</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Precision:"</span>,metrics.precision_score(slim_matrix_single_neuron.loc[:,selected_drug_name], preds_svm_matrix.loc[:,selected_drug_name]))</span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Recall:"</span>,metrics.recall_score(slim_matrix_single_neuron.loc[:,selected_drug_name], preds_svm_matrix.loc[:,selected_drug_name])) <span class="co">#TP / (TP+FN)</span></span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"AUC with score:"</span>,auc) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Accuracy: 0.7197508896797153
Precision: 0.9100917431192661
Recall: 0.6509186351706037
AUC with score: 0.8904018213192963</code></pre>
</div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="210">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># LOS LOGITS DE </span><span class="al">TEST</span><span class="co">!!</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>train_drug_logs <span class="op">=</span> pd.DataFrame(delta_logits_matrix.loc[:,selected_drug_name]).reset_index()</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>train_drug_logs.columns  <span class="op">=</span> [<span class="st">"GO_term"</span>,<span class="st">"probability"</span>]</span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>train_drug_logs <span class="op">=</span> train_drug_logs.merge(real_go_info_svm, on<span class="op">=</span><span class="st">"GO_term"</span>)</span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a>train_drug_logs.sort_values(by<span class="op">=</span>[<span class="st">"probability"</span>], ascending<span class="op">=</span><span class="va">False</span>).head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="210">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>probability</th>
      <th>Name</th>
      <th>layer_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>329</th>
      <td>GO:0006939</td>
      <td>2.871769</td>
      <td>Smooth muscle contraction (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>381</th>
      <td>GO:0048041</td>
      <td>2.568268</td>
      <td>Focal adhesion assembly (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>144</th>
      <td>GO:0002318</td>
      <td>2.533049</td>
      <td>Myeloid progenitor cell differentiation (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>408</th>
      <td>GO:0048008</td>
      <td>2.529048</td>
      <td>Platelet-derived growth factor receptor signaling pathway (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>595</th>
      <td>GO:0032526</td>
      <td>2.437785</td>
      <td>Response to retinoic acid (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>822</th>
      <td>GO:0046651</td>
      <td>2.300822</td>
      <td>Lymphocyte proliferation (1)</td>
      <td>4.0</td>
    </tr>
    <tr>
      <th>183</th>
      <td>GO:0048608</td>
      <td>2.222701</td>
      <td>Reproductive structure development (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>458</th>
      <td>GO:0048839</td>
      <td>2.066365</td>
      <td>Inner ear development (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>793</th>
      <td>GO:1901888</td>
      <td>2.058388</td>
      <td>Regulation of cell junction assembly (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>956</th>
      <td>GO:0002327</td>
      <td>1.921062</td>
      <td>Immature b cell differentiation (1)</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
</section>
<section id="final-model-svm" class="level2">
<h2 class="anchored" data-anchor-id="final-model-svm">Final model SVM</h2>
<p>Once the models have been cross-validated we create the final models using all samples…</p>
<div class="cell" data-execution_count="116">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_final <span class="op">=</span> {}</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_final <span class="op">=</span> {}</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>GO_terms_precision_svm_final <span class="op">=</span> {}</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>models_svm <span class="op">=</span> {}</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform logistics</span></span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> goterm <span class="kw">in</span> sparseGO_terms:</span>
<span id="cb132-8"><a href="#cb132-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">#print(goterm)</span></span>
<span id="cb132-9"><a href="#cb132-9" aria-hidden="true" tabindex="-1"></a>    goterm_drugs <span class="op">=</span> slim_matrix.loc[[goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(<span class="dv">1</span>)]].values.flatten()</span>
<span id="cb132-10"><a href="#cb132-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb132-11"><a href="#cb132-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">sum</span>(goterm_drugs) <span class="op">&lt;=</span> <span class="dv">16</span>:</span>
<span id="cb132-12"><a href="#cb132-12" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb132-13"><a href="#cb132-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-14"><a href="#cb132-14" aria-hidden="true" tabindex="-1"></a>    list_nodes <span class="op">=</span> []</span>
<span id="cb132-15"><a href="#cb132-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">7</span>):</span>
<span id="cb132-16"><a href="#cb132-16" aria-hidden="true" tabindex="-1"></a>        list_nodes.append(goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(i))</span>
<span id="cb132-17"><a href="#cb132-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-18"><a href="#cb132-18" aria-hidden="true" tabindex="-1"></a>    score <span class="op">=</span> attribution_data_annotated.loc[list_nodes].T</span>
<span id="cb132-19"><a href="#cb132-19" aria-hidden="true" tabindex="-1"></a>    score_mod <span class="op">=</span> score.divide(score.std()).fillna(<span class="dv">0</span>)</span>
<span id="cb132-20"><a href="#cb132-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb132-21"><a href="#cb132-21" aria-hidden="true" tabindex="-1"></a>    X_train <span class="op">=</span> score_mod</span>
<span id="cb132-22"><a href="#cb132-22" aria-hidden="true" tabindex="-1"></a>    X_test <span class="op">=</span> score_mod</span>
<span id="cb132-23"><a href="#cb132-23" aria-hidden="true" tabindex="-1"></a>    y_train <span class="op">=</span> goterm_drugs</span>
<span id="cb132-24"><a href="#cb132-24" aria-hidden="true" tabindex="-1"></a>    y_test <span class="op">=</span> goterm_drugs</span>
<span id="cb132-25"><a href="#cb132-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb132-26"><a href="#cb132-26" aria-hidden="true" tabindex="-1"></a>    <span class="co">#gamma = 1/(X_train.shape[1]*X_train.to_numpy().var())</span></span>
<span id="cb132-27"><a href="#cb132-27" aria-hidden="true" tabindex="-1"></a>    gamma<span class="op">=</span><span class="st">"scale"</span></span>
<span id="cb132-28"><a href="#cb132-28" aria-hidden="true" tabindex="-1"></a>    C<span class="op">=</span><span class="dv">1</span></span>
<span id="cb132-29"><a href="#cb132-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb132-30"><a href="#cb132-30" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb132-31"><a href="#cb132-31" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb132-32"><a href="#cb132-32" aria-hidden="true" tabindex="-1"></a>    svm_model <span class="op">=</span> svm.SVC(C<span class="op">=</span>C,gamma<span class="op">=</span>gamma, kernel<span class="op">=</span><span class="st">'rbf'</span>,</span>
<span id="cb132-33"><a href="#cb132-33" aria-hidden="true" tabindex="-1"></a>                           class_weight<span class="op">=</span><span class="st">"balanced"</span>,</span>
<span id="cb132-34"><a href="#cb132-34" aria-hidden="true" tabindex="-1"></a>                            tol<span class="op">=</span><span class="fl">0.001</span>,</span>
<span id="cb132-35"><a href="#cb132-35" aria-hidden="true" tabindex="-1"></a>                            probability<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb132-36"><a href="#cb132-36" aria-hidden="true" tabindex="-1"></a>                           random_state<span class="op">=</span><span class="dv">1234</span>)</span>
<span id="cb132-37"><a href="#cb132-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># fit the model with data</span></span>
<span id="cb132-38"><a href="#cb132-38" aria-hidden="true" tabindex="-1"></a>    svm_model.fit(X_train,y_train)</span>
<span id="cb132-39"><a href="#cb132-39" aria-hidden="true" tabindex="-1"></a>    y_pred<span class="op">=</span>svm_model.predict(X_test)</span>
<span id="cb132-40"><a href="#cb132-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-41"><a href="#cb132-41" aria-hidden="true" tabindex="-1"></a>    <span class="co">#auc</span></span>
<span id="cb132-42"><a href="#cb132-42" aria-hidden="true" tabindex="-1"></a>    y_pred_proba <span class="op">=</span> svm_model.predict_proba(X_test)[::,<span class="dv">1</span>]  <span class="co"># platt values</span></span>
<span id="cb132-43"><a href="#cb132-43" aria-hidden="true" tabindex="-1"></a>    <span class="co">#y_pred_proba = svm_model.decision_function(X_test)</span></span>
<span id="cb132-44"><a href="#cb132-44" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb132-45"><a href="#cb132-45" aria-hidden="true" tabindex="-1"></a>    GO_terms_auc_svm_final[goterm] <span class="op">=</span> metrics.roc_auc_score(y_test, y_pred_proba)</span>
<span id="cb132-46"><a href="#cb132-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-47"><a href="#cb132-47" aria-hidden="true" tabindex="-1"></a>    precision, recall, thresholds <span class="op">=</span> metrics.precision_recall_curve(y_test, y_pred_proba)</span>
<span id="cb132-48"><a href="#cb132-48" aria-hidden="true" tabindex="-1"></a>    GO_terms_aupr_svm_final[goterm] <span class="op">=</span> metrics.auc(recall, precision)</span>
<span id="cb132-49"><a href="#cb132-49" aria-hidden="true" tabindex="-1"></a>    GO_terms_precision_svm_final[goterm] <span class="op">=</span> metrics.precision_score(y_test, y_pred)</span>
<span id="cb132-50"><a href="#cb132-50" aria-hidden="true" tabindex="-1"></a>    models_svm[goterm]<span class="op">=</span>svm_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="117">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(models_svm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="117">
<pre><code>1124</code></pre>
</div>
</div>
<section id="final-model-auc" class="level3">
<h3 class="anchored" data-anchor-id="final-model-auc">Final model AUC</h3>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="118">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df_final <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(GO_terms_auc_svm_final.items()),columns <span class="op">=</span> [<span class="st">'goterm'</span>,<span class="st">'auc'</span>]).set_index(<span class="st">"goterm"</span>)</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df_final <span class="op">=</span> GO_terms_auc_svm_df_final.dropna()</span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>GO_terms_auc_svm_df_final.sort_values(by<span class="op">=</span>[<span class="st">"auc"</span>], ascending<span class="op">=</span><span class="va">False</span>).head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="118">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>auc</th>
    </tr>
    <tr>
      <th>goterm</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:0030890</th>
      <td>0.995511</td>
    </tr>
    <tr>
      <th>GO:1901029</th>
      <td>0.994437</td>
    </tr>
    <tr>
      <th>GO:0051452</th>
      <td>0.993673</td>
    </tr>
    <tr>
      <th>GO:0030279</th>
      <td>0.993469</td>
    </tr>
    <tr>
      <th>GO:1901863</th>
      <td>0.991582</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="119">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">'figure.figsize'</span>:(<span class="dv">6</span>,<span class="dv">4</span>)})</span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>perc <span class="op">=</span> <span class="bu">str</span>(<span class="bu">round</span>((<span class="dv">100</span><span class="op">*</span><span class="bu">len</span>(GO_terms_auc_svm_df_final[GO_terms_auc_svm_df_final[<span class="st">"auc"</span>]<span class="op">&gt;</span><span class="fl">0.7</span>])<span class="op">/</span><span class="bu">len</span>(GO_terms_auc_svm_df_final)),<span class="dv">2</span>))<span class="op">+</span><span class="st">"%"</span></span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>N, bins, patches <span class="op">=</span> plt.hist(GO_terms_auc_svm_df_final, color<span class="op">=</span>CB_color_cycle[<span class="dv">6</span>],bins<span class="op">=</span><span class="dv">50</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(bins)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> bins[i]<span class="op">&gt;</span><span class="fl">0.7</span>:</span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a>        patches[i].set_facecolor(CB_color_cycle[<span class="dv">2</span>])</span>
<span id="cb136-8"><a href="#cb136-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-9"><a href="#cb136-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"AUC (logistic 1)"</span>, fontsize<span class="op">=</span><span class="dv">16</span>)  </span>
<span id="cb136-10"><a href="#cb136-10" aria-hidden="true" tabindex="-1"></a>plt.title(perc, fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb136-11"><a href="#cb136-11" aria-hidden="true" tabindex="-1"></a><span class="co"># con el que mejor funciona es con la suma normal del attribution </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="119">
<pre><code>Text(0.5, 1.0, '93.42%')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-116-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="final-model-aupr" class="level3">
<h3 class="anchored" data-anchor-id="final-model-aupr">Final model AUPR</h3>
<div class="cell" data-execution_count="120">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df_final <span class="op">=</span> pd.DataFrame(<span class="bu">list</span>(GO_terms_aupr_svm_final.items()),columns <span class="op">=</span> [<span class="st">'goterm'</span>,<span class="st">'aupr'</span>]).set_index(<span class="st">"goterm"</span>)</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df_final <span class="op">=</span> GO_terms_aupr_svm_df_final.dropna()</span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>GO_terms_aupr_svm_df_final.sort_values(by<span class="op">=</span>[<span class="st">"aupr"</span>], ascending<span class="op">=</span><span class="va">False</span>).head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="120">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>aupr</th>
    </tr>
    <tr>
      <th>goterm</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GO:0009058</th>
      <td>0.974362</td>
    </tr>
    <tr>
      <th>GO:0030890</th>
      <td>0.965743</td>
    </tr>
    <tr>
      <th>GO:0033033</th>
      <td>0.965598</td>
    </tr>
    <tr>
      <th>GO:0097345</th>
      <td>0.960797</td>
    </tr>
    <tr>
      <th>GO:0010467</th>
      <td>0.959505</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="121">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co"># TENGO PROBLEMA CON EL RECALL </span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">'figure.figsize'</span>:(<span class="dv">5</span>,<span class="dv">3</span>)})</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>perc <span class="op">=</span> <span class="bu">str</span>(<span class="bu">round</span>((<span class="dv">100</span><span class="op">*</span><span class="bu">len</span>(GO_terms_aupr_svm_df_final[GO_terms_aupr_svm_df_final[<span class="st">"aupr"</span>]<span class="op">&gt;</span><span class="fl">0.7</span>])<span class="op">/</span><span class="bu">len</span>(GO_terms_aupr_svm_df_final)),<span class="dv">2</span>))<span class="op">+</span><span class="st">"%"</span></span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>N, bins, patches <span class="op">=</span> plt.hist(GO_terms_aupr_svm_df_final, color<span class="op">=</span>CB_color_cycle[<span class="dv">6</span>],bins<span class="op">=</span><span class="dv">50</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(bins)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> bins[i]<span class="op">&gt;</span><span class="fl">0.7</span>:</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>        patches[i].set_facecolor(CB_color_cycle[<span class="dv">3</span>])</span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"AUPR"</span>, fontsize<span class="op">=</span><span class="dv">16</span>)  </span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a>plt.title(perc, fontsize<span class="op">=</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="121">
<pre><code>Text(0.5, 1.0, '23.31%')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-118-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="predict-for-a-new-drug" class="level2">
<h2 class="anchored" data-anchor-id="predict-for-a-new-drug">Predict for a new drug</h2>
<section id="make-predictions" class="level3">
<h3 class="anchored" data-anchor-id="make-predictions">Make predictions</h3>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="122">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>unknown <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(attribution_data_all.columns)<span class="op">-</span><span class="bu">set</span>(attribution_data_annotated.columns))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get the probabilities for all unknown drugs</p>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>predictions <span class="op">=</span> {}</span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>distances <span class="op">=</span> {}</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>probabilities_unknown <span class="op">=</span> pd.DataFrame()</span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>preds_unknown <span class="op">=</span> pd.DataFrame()</span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> drug <span class="kw">in</span> unknown:</span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a>    probabilities <span class="op">=</span> {}</span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> goterm <span class="kw">in</span> models_svm.keys():</span>
<span id="cb142-9"><a href="#cb142-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-10"><a href="#cb142-10" aria-hidden="true" tabindex="-1"></a>        list_nodes <span class="op">=</span> <span class="bu">list</span>(models_svm[goterm].feature_names_in_) <span class="co"># Extract the feature names from the model (those are the attributions we need)</span></span>
<span id="cb142-11"><a href="#cb142-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-12"><a href="#cb142-12" aria-hidden="true" tabindex="-1"></a>        score <span class="op">=</span> attribution_data_all.loc[list_nodes][drug].to_frame().T </span>
<span id="cb142-13"><a href="#cb142-13" aria-hidden="true" tabindex="-1"></a>        score_mod <span class="op">=</span> score.divide(attribution_data_annotated.loc[list_nodes].T.std()).fillna(<span class="dv">0</span>) <span class="co">#divide by std of each neuron, only use drugs that trained the models</span></span>
<span id="cb142-14"><a href="#cb142-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-15"><a href="#cb142-15" aria-hidden="true" tabindex="-1"></a>        predictions[goterm]<span class="op">=</span>models_svm[goterm].predict(score_mod)</span>
<span id="cb142-16"><a href="#cb142-16" aria-hidden="true" tabindex="-1"></a>        probabilities[goterm] <span class="op">=</span> models_svm[goterm].predict_proba(score_mod)[::,<span class="dv">1</span>]  <span class="co"># platt values</span></span>
<span id="cb142-17"><a href="#cb142-17" aria-hidden="true" tabindex="-1"></a>        <span class="co"># distances[goterm] = models_svm[goterm].decision_function(score_mod)</span></span>
<span id="cb142-18"><a href="#cb142-18" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb142-19"><a href="#cb142-19" aria-hidden="true" tabindex="-1"></a>    drug_probs <span class="op">=</span> pd.DataFrame.from_dict(probabilities).T</span>
<span id="cb142-20"><a href="#cb142-20" aria-hidden="true" tabindex="-1"></a>    drug_probs.columns <span class="op">=</span> [drug]</span>
<span id="cb142-21"><a href="#cb142-21" aria-hidden="true" tabindex="-1"></a>    drug_preds <span class="op">=</span> pd.DataFrame.from_dict(predictions).T</span>
<span id="cb142-22"><a href="#cb142-22" aria-hidden="true" tabindex="-1"></a>    drug_preds.columns <span class="op">=</span> [drug]    </span>
<span id="cb142-23"><a href="#cb142-23" aria-hidden="true" tabindex="-1"></a>    probabilities_unknown <span class="op">=</span> pd.concat([probabilities_unknown,drug_probs], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb142-24"><a href="#cb142-24" aria-hidden="true" tabindex="-1"></a>    preds_unknown <span class="op">=</span> pd.concat([preds_unknown,drug_preds], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb142-25"><a href="#cb142-25" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(drug)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="123">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To import dataframe created before</span></span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(resultsdir<span class="op">+</span><span class="st">'probabilities_unknown_SVM.pkl'</span>, <span class="st">'rb'</span>) <span class="im">as</span> dictionary_file:</span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a>    probabilities_unknown <span class="op">=</span> pickle.load(dictionary_file)  </span>
<span id="cb143-4"><a href="#cb143-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb143-5"><a href="#cb143-5" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(resultsdir<span class="op">+</span><span class="st">'preds_unknown_SVM.pkl'</span>, <span class="st">'rb'</span>) <span class="im">as</span> dictionary_file:</span>
<span id="cb143-6"><a href="#cb143-6" aria-hidden="true" tabindex="-1"></a>    preds_unknown <span class="op">=</span> pickle.load(dictionary_file) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="study-drug-with-unknown-moa" class="level3">
<h3 class="anchored" data-anchor-id="study-drug-with-unknown-moa">Study drug with unknown MOA</h3>
<p><strong>Choose drug with unknown MOA…</strong></p>
<div class="cell" data-execution_count="124">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>combobox_u <span class="op">=</span> interactive(f, drug<span class="op">=</span>widgets.Combobox(options<span class="op">=</span>unknown))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="125">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>predictions_nodes <span class="op">=</span> []</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> goterm <span class="kw">in</span> <span class="bu">list</span>(platt_matrix.index):</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>    predictions_nodes.append(goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add names to go terms</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>real_go_info_svm<span class="op">=</span> real_go_info[real_go_info.GO_term.isin(predictions_nodes)]</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a>real_go_info_svm.GO_term <span class="op">=</span> real_go_info_svm.GO_term.<span class="bu">str</span>.replace(<span class="st">"_1"</span>,<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="127">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>display(combobox_u)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"9790bb6e37a6483abc17a2bdbe3767ed","version_major":2,"version_minor":0}
</script>
</div>
</div>
<div class="cell" data-execution_count="128">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>selected_drug_u_name <span class="op">=</span> combobox_u.result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="129">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>predictions_df <span class="op">=</span> pd.DataFrame.from_dict(preds_unknown.loc[:,selected_drug_u_name]).reset_index()</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>predictions_df.columns  <span class="op">=</span> [<span class="st">"GO_term"</span>,<span class="st">"predictions"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="130">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>probabilities_df <span class="op">=</span> pd.DataFrame.from_dict(probabilities_unknown.loc[:,selected_drug_u_name]).reset_index()</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>probabilities_df.columns  <span class="op">=</span> [<span class="st">"GO_term"</span>,<span class="st">"probability"</span>]</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>probabilities_df <span class="op">=</span> probabilities_df.merge(real_go_info_svm, on<span class="op">=</span><span class="st">"GO_term"</span>)</span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>probabilities_df <span class="op">=</span> probabilities_df.merge(predictions_df, on<span class="op">=</span><span class="st">"GO_term"</span>)</span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>probabilities_df.loc[probabilities_df[<span class="st">"layer_number"</span>] <span class="op">&lt;=</span><span class="dv">3</span>].sort_values(by<span class="op">=</span>[<span class="st">"probability"</span>], ascending<span class="op">=</span><span class="va">False</span>).head(<span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="130">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>probability</th>
      <th>Name</th>
      <th>layer_number</th>
      <th>predictions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>15</th>
      <td>GO:0007051</td>
      <td>0.775651</td>
      <td>Spindle organization (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>25</th>
      <td>GO:0045930</td>
      <td>0.728483</td>
      <td>Negative regulation of mitotic cell cycle (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>312</th>
      <td>GO:0033033</td>
      <td>0.668570</td>
      <td>Negative regulation of myeloid cell apoptotic process (1)</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>85</th>
      <td>GO:0001783</td>
      <td>0.648156</td>
      <td>B cell apoptotic process (1)</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>241</th>
      <td>GO:0018105</td>
      <td>0.645704</td>
      <td>Peptidyl-serine phosphorylation (1)</td>
      <td>2.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>579</th>
      <td>GO:0031328</td>
      <td>0.645487</td>
      <td>Positive regulation of cellular biosynthetic process (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>365</th>
      <td>GO:0007098</td>
      <td>0.597286</td>
      <td>Centrosome cycle (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>789</th>
      <td>GO:1900118</td>
      <td>0.562231</td>
      <td>Negative regulation of execution phase of apoptosis (1)</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>444</th>
      <td>GO:1901990</td>
      <td>0.539374</td>
      <td>Regulation of mitotic cell cycle phase transition (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>636</th>
      <td>GO:0042633</td>
      <td>0.538572</td>
      <td>Hair cycle (1)</td>
      <td>2.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>73</th>
      <td>GO:0071456</td>
      <td>0.526585</td>
      <td>Cellular response to hypoxia (1)</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>888</th>
      <td>GO:1901988</td>
      <td>0.526572</td>
      <td>Negative regulation of cell cycle phase transition (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>242</th>
      <td>GO:0018107</td>
      <td>0.515176</td>
      <td>Peptidyl-threonine phosphorylation (1)</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>362</th>
      <td>GO:0007030</td>
      <td>0.511032</td>
      <td>Golgi organization (1)</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>465</th>
      <td>GO:0007517</td>
      <td>0.510204</td>
      <td>Muscle organ development (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>532</th>
      <td>GO:0010557</td>
      <td>0.509768</td>
      <td>Positive regulation of macromolecule biosynthetic process (1)</td>
      <td>3.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>502</th>
      <td>GO:0008284</td>
      <td>0.500000</td>
      <td>Positive regulation of cell population proliferation (1)</td>
      <td>3.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>741</th>
      <td>GO:0055001</td>
      <td>0.500000</td>
      <td>Muscle cell development (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>795</th>
      <td>GO:1902236</td>
      <td>0.500000</td>
      <td>Negative regulation of endoplasmic reticulum stress-induced intrinsic apoptotic signaling pathway (1)</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>233</th>
      <td>GO:0045944</td>
      <td>0.481732</td>
      <td>Positive regulation of transcription by rna polymerase ii (1)</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>321</th>
      <td>GO:2000811</td>
      <td>0.478198</td>
      <td>Negative regulation of anoikis (1)</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>227</th>
      <td>GO:0006366</td>
      <td>0.474951</td>
      <td>Transcription by rna polymerase ii (1)</td>
      <td>3.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>511</th>
      <td>GO:0009410</td>
      <td>0.466373</td>
      <td>Response to xenobiotic stimulus (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>776</th>
      <td>GO:0072593</td>
      <td>0.458202</td>
      <td>Reactive oxygen species metabolic process (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>78</th>
      <td>GO:0001701</td>
      <td>0.448882</td>
      <td>In utero embryonic development (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>177</th>
      <td>GO:0007283</td>
      <td>0.445234</td>
      <td>Spermatogenesis (1)</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>319</th>
      <td>GO:0097194</td>
      <td>0.419988</td>
      <td>Execution phase of apoptosis (1)</td>
      <td>2.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>86</th>
      <td>GO:0002903</td>
      <td>0.417616</td>
      <td>Negative regulation of b cell apoptotic process (1)</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>614</th>
      <td>GO:0035094</td>
      <td>0.412038</td>
      <td>Response to nicotine (1)</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>1117</th>
      <td>GO:2001243</td>
      <td>0.409103</td>
      <td>Negative regulation of intrinsic apoptotic signaling pathway (1)</td>
      <td>2.0</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="131">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="bu">sum</span>(probabilities_df[<span class="st">"predictions"</span>] <span class="op">==</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="131">
<pre><code>303</code></pre>
</div>
</div>
<div class="cell" data-execution_count="132">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="bu">sum</span>(probabilities_df[<span class="st">"predictions"</span>] <span class="op">==</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="132">
<pre><code>821</code></pre>
</div>
</div>
<p>Probability &lt; 0.5 doesn’t mean it does not belong to the class, a probability of for example 0.2 can represent a 1 (annotated to MoA)</p>
</section>
</section>
<section id="modify-probabilities" class="level2">
<h2 class="anchored" data-anchor-id="modify-probabilities">Modify probabilities</h2>
<p><strong>Take into account the annotations each GO term has</strong> (general GO terms are easier to predict as they have more annotations)</p>
<section id="for-drug-with-unknown-moa" class="level3">
<h3 class="anchored" data-anchor-id="for-drug-with-unknown-moa">For drug with unknown MOA…</h3>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="133">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>sum_annotations <span class="op">=</span> slim_matrix_single_neuron.T.<span class="bu">sum</span>()<span class="op">/</span>slim_matrix_single_neuron.shape[<span class="dv">1</span>]</span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a>logits_apriori <span class="op">=</span> np.log(sum_annotations<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>sum_annotations))</span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb155-4"><a href="#cb155-4" aria-hidden="true" tabindex="-1"></a>logits_apost<span class="op">=</span> np.log(probabilities_df[<span class="st">"probability"</span>]<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>probabilities_df[<span class="st">"probability"</span>]))</span>
<span id="cb155-5"><a href="#cb155-5" aria-hidden="true" tabindex="-1"></a>delta_logits <span class="op">=</span>logits_apost.to_numpy()<span class="op">-</span> logits_apriori.to_numpy()</span>
<span id="cb155-6"><a href="#cb155-6" aria-hidden="true" tabindex="-1"></a>delta_logits_df <span class="op">=</span> pd.DataFrame(delta_logits)</span>
<span id="cb155-7"><a href="#cb155-7" aria-hidden="true" tabindex="-1"></a>delta_logits_df.columns <span class="op">=</span> [<span class="st">"delta_logits"</span>]</span>
<span id="cb155-8"><a href="#cb155-8" aria-hidden="true" tabindex="-1"></a>probabilities_mod <span class="op">=</span> probabilities_df.merge(delta_logits_df, left_index<span class="op">=</span><span class="va">True</span>,right_index<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="134">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>probabilities_mod.loc[probabilities_mod[<span class="st">"predictions"</span>] <span class="op">==</span><span class="dv">1</span>].loc[probabilities_mod[<span class="st">"layer_number"</span>] <span class="op">&lt;=</span> <span class="dv">7</span>].sort_values(by<span class="op">=</span>[<span class="st">"delta_logits"</span>], ascending<span class="op">=</span><span class="va">False</span>).head(<span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="134">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>probability</th>
      <th>Name</th>
      <th>layer_number</th>
      <th>predictions</th>
      <th>delta_logits</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>15</th>
      <td>GO:0007051</td>
      <td>0.775651</td>
      <td>Spindle organization (1)</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>3.801391</td>
    </tr>
    <tr>
      <th>312</th>
      <td>GO:0033033</td>
      <td>0.668570</td>
      <td>Negative regulation of myeloid cell apoptotic process (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>3.207249</td>
    </tr>
    <tr>
      <th>85</th>
      <td>GO:0001783</td>
      <td>0.648156</td>
      <td>B cell apoptotic process (1)</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>2.917688</td>
    </tr>
    <tr>
      <th>789</th>
      <td>GO:1900118</td>
      <td>0.562231</td>
      <td>Negative regulation of execution phase of apoptosis (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>2.811113</td>
    </tr>
    <tr>
      <th>364</th>
      <td>GO:0007059</td>
      <td>0.651932</td>
      <td>Chromosome segregation (1)</td>
      <td>5.0</td>
      <td>1.0</td>
      <td>2.648874</td>
    </tr>
    <tr>
      <th>25</th>
      <td>GO:0045930</td>
      <td>0.728483</td>
      <td>Negative regulation of mitotic cell cycle (1)</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>2.600911</td>
    </tr>
    <tr>
      <th>795</th>
      <td>GO:1902236</td>
      <td>0.500000</td>
      <td>Negative regulation of endoplasmic reticulum stress-induced intrinsic apoptotic signaling pathway (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>2.560893</td>
    </tr>
    <tr>
      <th>365</th>
      <td>GO:0007098</td>
      <td>0.597286</td>
      <td>Centrosome cycle (1)</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>2.530027</td>
    </tr>
    <tr>
      <th>362</th>
      <td>GO:0007030</td>
      <td>0.511032</td>
      <td>Golgi organization (1)</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>2.350877</td>
    </tr>
    <tr>
      <th>376</th>
      <td>GO:0098813</td>
      <td>0.474213</td>
      <td>Nuclear chromosome segregation (1)</td>
      <td>4.0</td>
      <td>1.0</td>
      <td>2.073193</td>
    </tr>
    <tr>
      <th>762</th>
      <td>GO:0070925</td>
      <td>0.703152</td>
      <td>Organelle assembly (1)</td>
      <td>4.0</td>
      <td>1.0</td>
      <td>2.069600</td>
    </tr>
    <tr>
      <th>903</th>
      <td>GO:0072384</td>
      <td>0.336307</td>
      <td>Organelle transport along microtubule (1)</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>2.000421</td>
    </tr>
    <tr>
      <th>86</th>
      <td>GO:0002903</td>
      <td>0.417616</td>
      <td>Negative regulation of b cell apoptotic process (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.974177</td>
    </tr>
    <tr>
      <th>370</th>
      <td>GO:0045786</td>
      <td>0.624233</td>
      <td>Negative regulation of cell cycle (1)</td>
      <td>5.0</td>
      <td>1.0</td>
      <td>1.917603</td>
    </tr>
    <tr>
      <th>367</th>
      <td>GO:0010948</td>
      <td>0.603975</td>
      <td>Negative regulation of cell cycle process (1)</td>
      <td>4.0</td>
      <td>1.0</td>
      <td>1.905513</td>
    </tr>
    <tr>
      <th>1058</th>
      <td>GO:2000669</td>
      <td>0.341573</td>
      <td>Negative regulation of dendritic cell apoptotic process (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.796355</td>
    </tr>
    <tr>
      <th>1056</th>
      <td>GO:0070233</td>
      <td>0.356845</td>
      <td>Negative regulation of t cell apoptotic process (1)</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.764358</td>
    </tr>
    <tr>
      <th>839</th>
      <td>GO:1902166</td>
      <td>0.329441</td>
      <td>Negative regulation of intrinsic apoptotic signaling pathway in response to dna damage by p53 class mediator (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.741929</td>
    </tr>
    <tr>
      <th>368</th>
      <td>GO:0032465</td>
      <td>0.382860</td>
      <td>Regulation of cytokinesis (1)</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>1.740942</td>
    </tr>
    <tr>
      <th>321</th>
      <td>GO:2000811</td>
      <td>0.478198</td>
      <td>Negative regulation of anoikis (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.731163</td>
    </tr>
    <tr>
      <th>319</th>
      <td>GO:0097194</td>
      <td>0.419988</td>
      <td>Execution phase of apoptosis (1)</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>1.698511</td>
    </tr>
    <tr>
      <th>16</th>
      <td>GO:0031023</td>
      <td>0.312643</td>
      <td>Microtubule organizing center organization (1)</td>
      <td>4.0</td>
      <td>1.0</td>
      <td>1.664852</td>
    </tr>
    <tr>
      <th>888</th>
      <td>GO:1901988</td>
      <td>0.526572</td>
      <td>Negative regulation of cell cycle phase transition (1)</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>1.640736</td>
    </tr>
    <tr>
      <th>636</th>
      <td>GO:0042633</td>
      <td>0.538572</td>
      <td>Hair cycle (1)</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>1.613210</td>
    </tr>
    <tr>
      <th>994</th>
      <td>GO:1900182</td>
      <td>0.296416</td>
      <td>Positive regulation of protein localization to nucleus (1)</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.537597</td>
    </tr>
    <tr>
      <th>965</th>
      <td>GO:0030705</td>
      <td>0.274210</td>
      <td>Cytoskeleton-dependent intracellular transport (1)</td>
      <td>4.0</td>
      <td>1.0</td>
      <td>1.532160</td>
    </tr>
    <tr>
      <th>977</th>
      <td>GO:0031648</td>
      <td>0.261530</td>
      <td>Protein destabilization (1)</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.522862</td>
    </tr>
    <tr>
      <th>289</th>
      <td>GO:0032469</td>
      <td>0.301677</td>
      <td>Endoplasmic reticulum calcium ion homeostasis (1)</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.514119</td>
    </tr>
    <tr>
      <th>465</th>
      <td>GO:0007517</td>
      <td>0.510204</td>
      <td>Muscle organ development (1)</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>1.499437</td>
    </tr>
    <tr>
      <th>614</th>
      <td>GO:0035094</td>
      <td>0.412038</td>
      <td>Response to nicotine (1)</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.494658</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="135">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>names2 <span class="op">=</span> <span class="bu">list</span>(probabilities_mod.loc[probabilities_mod[<span class="st">"predictions"</span>] <span class="op">==</span><span class="dv">1</span>].loc[probabilities_mod[<span class="st">"layer_number"</span>] <span class="op">&lt;=</span><span class="dv">7</span>].sort_values(by<span class="op">=</span>[<span class="st">"delta_logits"</span>], ascending<span class="op">=</span><span class="va">False</span>)[<span class="st">"Name"</span>].head(<span class="dv">30</span>))</span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>terms2 <span class="op">=</span> <span class="bu">list</span>(probabilities_mod.loc[probabilities_mod[<span class="st">"predictions"</span>] <span class="op">==</span><span class="dv">1</span>].loc[probabilities_mod[<span class="st">"layer_number"</span>] <span class="op">&lt;=</span><span class="dv">7</span>].sort_values(by<span class="op">=</span>[<span class="st">"delta_logits"</span>], ascending<span class="op">=</span><span class="va">False</span>)[<span class="st">"GO_term"</span>].head(<span class="dv">30</span>))</span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a>logits2 <span class="op">=</span> <span class="bu">list</span>(probabilities_mod.loc[probabilities_mod[<span class="st">"predictions"</span>] <span class="op">==</span><span class="dv">1</span>].loc[probabilities_mod[<span class="st">"layer_number"</span>] <span class="op">&lt;=</span><span class="dv">7</span>].sort_values(by<span class="op">=</span>[<span class="st">"delta_logits"</span>], ascending<span class="op">=</span><span class="va">False</span>)[<span class="st">"delta_logits"</span>].head(<span class="dv">30</span>))</span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a>names2 <span class="op">=</span> [x[:<span class="op">-</span><span class="dv">4</span>] <span class="cf">for</span> x <span class="kw">in</span> names2] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="136">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(names2)):</span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(terms2[i],names2[i],logits2[i])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GO:0007051 Spindle organization 3.801390702868236
GO:0033033 Negative regulation of myeloid cell apoptotic process 3.2072492134319956
GO:0001783 B cell apoptotic process 2.9176875271685683
GO:1900118 Negative regulation of execution phase of apoptosis 2.8111130136057216
GO:0007059 Chromosome segregation 2.64887443203506
GO:0045930 Negative regulation of mitotic cell cycle 2.6009108234102625
GO:1902236 Negative regulation of endoplasmic reticulum stress-induced intrinsic apoptotic signaling pathway 2.5608925567659226
GO:0007098 Centrosome cycle 2.530026738008422
GO:0007030 Golgi organization 2.3508774173095186
GO:0098813 Nuclear chromosome segregation 2.0731926893576196
GO:0070925 Organelle assembly 2.069600192282952
GO:0072384 Organelle transport along microtubule 2.0004213788686807
GO:0002903 Negative regulation of b cell apoptotic process 1.9741768883148958
GO:0045786 Negative regulation of cell cycle 1.917603030826096
GO:0010948 Negative regulation of cell cycle process 1.9055131265535517
GO:2000669 Negative regulation of dendritic cell apoptotic process 1.7963554850648258
GO:0070233 Negative regulation of t cell apoptotic process 1.7643576360799753
GO:1902166 Negative regulation of intrinsic apoptotic signaling pathway in response to dna damage by p53 class mediator 1.7419293617961054
GO:0032465 Regulation of cytokinesis 1.7409418248944983
GO:2000811 Negative regulation of anoikis 1.7311634175406592
GO:0097194 Execution phase of apoptosis 1.6985112888037164
GO:0031023 Microtubule organizing center organization 1.6648522832794037
GO:1901988 Negative regulation of cell cycle phase transition 1.640735739239383
GO:0042633 Hair cycle 1.6132096412865522
GO:1900182 Positive regulation of protein localization to nucleus 1.5375973554160984
GO:0030705 Cytoskeleton-dependent intracellular transport 1.5321597954461827
GO:0031648 Protein destabilization 1.5228619022250793
GO:0032469 Endoplasmic reticulum calcium ion homeostasis 1.5141189889858477
GO:0007517 Muscle organ development 1.499437003178439
GO:0035094 Response to nicotine 1.494657812504115</code></pre>
</div>
</div>
<div class="cell" data-execution_count="138">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="co"># import libraries</span></span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb160-3"><a href="#cb160-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb160-4"><a href="#cb160-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb160-5"><a href="#cb160-5" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb160-6"><a href="#cb160-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-7"><a href="#cb160-7" aria-hidden="true" tabindex="-1"></a><span class="co"># set font</span></span>
<span id="cb160-8"><a href="#cb160-8" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'font.family'</span>] <span class="op">=</span> <span class="st">'sans-serif'</span></span>
<span id="cb160-9"><a href="#cb160-9" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'font.sans-serif'</span>] <span class="op">=</span> <span class="st">'Roboto'</span></span>
<span id="cb160-10"><a href="#cb160-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-11"><a href="#cb160-11" aria-hidden="true" tabindex="-1"></a><span class="co"># set the style of the axes and the text color</span></span>
<span id="cb160-12"><a href="#cb160-12" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'axes.edgecolor'</span>]<span class="op">=</span><span class="st">'#333F4B'</span></span>
<span id="cb160-13"><a href="#cb160-13" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'axes.linewidth'</span>]<span class="op">=</span><span class="fl">0.8</span></span>
<span id="cb160-14"><a href="#cb160-14" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'xtick.color'</span>]<span class="op">=</span><span class="st">'#333F4B'</span></span>
<span id="cb160-15"><a href="#cb160-15" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'ytick.color'</span>]<span class="op">=</span><span class="st">'#333F4B'</span></span>
<span id="cb160-16"><a href="#cb160-16" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'text.color'</span>]<span class="op">=</span><span class="st">'#333F4B'</span></span>
<span id="cb160-17"><a href="#cb160-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-18"><a href="#cb160-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-19"><a href="#cb160-19" aria-hidden="true" tabindex="-1"></a><span class="co"># create some fake data</span></span>
<span id="cb160-20"><a href="#cb160-20" aria-hidden="true" tabindex="-1"></a>percentages <span class="op">=</span> pd.Series(logits2, </span>
<span id="cb160-21"><a href="#cb160-21" aria-hidden="true" tabindex="-1"></a>                        index<span class="op">=</span>[s.title() <span class="cf">for</span> s <span class="kw">in</span> names2])</span>
<span id="cb160-22"><a href="#cb160-22" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame({<span class="st">'percentage'</span> : percentages})</span>
<span id="cb160-23"><a href="#cb160-23" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.sort_values(by<span class="op">=</span><span class="st">'percentage'</span>)</span>
<span id="cb160-24"><a href="#cb160-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-25"><a href="#cb160-25" aria-hidden="true" tabindex="-1"></a><span class="co"># we first need a numeric placeholder for the y axis</span></span>
<span id="cb160-26"><a href="#cb160-26" aria-hidden="true" tabindex="-1"></a>my_range<span class="op">=</span><span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="bu">len</span>(df.index)<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb160-27"><a href="#cb160-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-28"><a href="#cb160-28" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>,<span class="dv">17</span>))</span>
<span id="cb160-29"><a href="#cb160-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-30"><a href="#cb160-30" aria-hidden="true" tabindex="-1"></a><span class="co"># create for each expense type an horizontal line that starts at x = 0 with the length </span></span>
<span id="cb160-31"><a href="#cb160-31" aria-hidden="true" tabindex="-1"></a><span class="co"># represented by the specific value.</span></span>
<span id="cb160-32"><a href="#cb160-32" aria-hidden="true" tabindex="-1"></a>plt.hlines(y<span class="op">=</span>my_range, xmin<span class="op">=</span><span class="dv">0</span>, xmax<span class="op">=</span>df[<span class="st">'percentage'</span>], color<span class="op">=</span><span class="st">'#208EA3'</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, linewidth<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb160-33"><a href="#cb160-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-34"><a href="#cb160-34" aria-hidden="true" tabindex="-1"></a><span class="co"># create for each value type a dot at the level of the value</span></span>
<span id="cb160-35"><a href="#cb160-35" aria-hidden="true" tabindex="-1"></a>plt.plot(df[<span class="st">'percentage'</span>], my_range, <span class="st">"o"</span>, markersize<span class="op">=</span><span class="dv">14</span>, color<span class="op">=</span><span class="st">'#208EA3'</span>, alpha<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb160-36"><a href="#cb160-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-37"><a href="#cb160-37" aria-hidden="true" tabindex="-1"></a><span class="co"># set labels</span></span>
<span id="cb160-38"><a href="#cb160-38" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">' Δlogit'</span>, fontsize<span class="op">=</span><span class="dv">25</span>, fontweight<span class="op">=</span><span class="st">'black'</span>, color <span class="op">=</span> <span class="st">'#36382E'</span>)</span>
<span id="cb160-39"><a href="#cb160-39" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">''</span>)</span>
<span id="cb160-40"><a href="#cb160-40" aria-hidden="true" tabindex="-1"></a>ax.set_facecolor(color<span class="op">=</span><span class="st">"white"</span>)</span>
<span id="cb160-41"><a href="#cb160-41" aria-hidden="true" tabindex="-1"></a>ax.set_alpha(<span class="dv">1</span>)</span>
<span id="cb160-42"><a href="#cb160-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-43"><a href="#cb160-43" aria-hidden="true" tabindex="-1"></a><span class="co"># set axis</span></span>
<span id="cb160-44"><a href="#cb160-44" aria-hidden="true" tabindex="-1"></a>ax.tick_params(axis<span class="op">=</span><span class="st">'both'</span>, which<span class="op">=</span><span class="st">'major'</span>, labelsize<span class="op">=</span><span class="dv">32</span>)</span>
<span id="cb160-45"><a href="#cb160-45" aria-hidden="true" tabindex="-1"></a>plt.yticks(my_range, df.index)</span>
<span id="cb160-46"><a href="#cb160-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-47"><a href="#cb160-47" aria-hidden="true" tabindex="-1"></a><span class="co"># add an horizonal label for the y axis </span></span>
<span id="cb160-48"><a href="#cb160-48" aria-hidden="true" tabindex="-1"></a>fig.text(<span class="op">-</span><span class="fl">0.58</span>, <span class="fl">0.862</span>, <span class="st">'MoA (GO terms)'</span>, fontsize<span class="op">=</span><span class="dv">27</span>, fontweight<span class="op">=</span><span class="st">'black'</span>, color <span class="op">=</span> <span class="st">'#36382E'</span>)</span>
<span id="cb160-49"><a href="#cb160-49" aria-hidden="true" tabindex="-1"></a>fig.text(<span class="fl">0.2</span>, <span class="fl">0.9</span>, selected_drug_u_name.capitalize(), fontsize<span class="op">=</span><span class="dv">32</span>, fontweight<span class="op">=</span><span class="st">'black'</span>, color <span class="op">=</span> <span class="st">'#36382E'</span>)</span>
<span id="cb160-50"><a href="#cb160-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-51"><a href="#cb160-51" aria-hidden="true" tabindex="-1"></a><span class="co"># .upper() capitalize</span></span>
<span id="cb160-52"><a href="#cb160-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-53"><a href="#cb160-53" aria-hidden="true" tabindex="-1"></a><span class="co"># change the style of the axis spines</span></span>
<span id="cb160-54"><a href="#cb160-54" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb160-55"><a href="#cb160-55" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb160-56"><a href="#cb160-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-57"><a href="#cb160-57" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'left'</span>].set_bounds((<span class="dv">1</span>, <span class="bu">len</span>(my_range)))</span>
<span id="cb160-58"><a href="#cb160-58" aria-hidden="true" tabindex="-1"></a>ax.set_xlim(<span class="dv">0</span>,<span class="bu">max</span>(logits2)<span class="op">+</span><span class="fl">0.1</span>)</span>
<span id="cb160-59"><a href="#cb160-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-60"><a href="#cb160-60" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'left'</span>].set_position((<span class="st">'outward'</span>, <span class="dv">8</span>))</span>
<span id="cb160-61"><a href="#cb160-61" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'bottom'</span>].set_position((<span class="st">'outward'</span>, <span class="dv">5</span>))</span>
<span id="cb160-62"><a href="#cb160-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-63"><a href="#cb160-63" aria-hidden="true" tabindex="-1"></a>plt.savefig(resultsdir<span class="op">+</span>selected_drug_u_name<span class="op">+</span><span class="st">'_top_terms.png'</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-135-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="for-known-drug" class="level3">
<h3 class="anchored" data-anchor-id="for-known-drug">For known drug…</h3>
<div class="cell" data-execution_count="139">
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a>display(combobox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"e2b7434a9229422789185cdf3d99e899","version_major":2,"version_minor":0}
</script>
</div>
</div>
<div class="cell" data-execution_count="140">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>selected_drug_name <span class="op">=</span> combobox.result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="141">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># LOS LOGITS DE </span><span class="al">TEST</span><span class="co">!!</span></span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>train_drug_logs <span class="op">=</span> pd.DataFrame(platt_matrix.loc[:,selected_drug_name]).reset_index()</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>train_drug_logs.columns  <span class="op">=</span> [<span class="st">"GO_term"</span>,<span class="st">"probability"</span>]</span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>train_drug_logs <span class="op">=</span> train_drug_logs.merge(real_go_info_svm, on<span class="op">=</span><span class="st">"GO_term"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="142">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>train_drug_logs.loc[train_drug_logs[<span class="st">"layer_number"</span>] <span class="op">&lt;=</span><span class="dv">3</span>].sort_values(by<span class="op">=</span>[<span class="st">"probability"</span>], ascending<span class="op">=</span><span class="va">False</span>).head(<span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="142">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>probability</th>
      <th>Name</th>
      <th>layer_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>183</th>
      <td>GO:0048608</td>
      <td>0.742225</td>
      <td>Reproductive structure development (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>381</th>
      <td>GO:0048041</td>
      <td>0.737675</td>
      <td>Focal adhesion assembly (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>535</th>
      <td>GO:0010628</td>
      <td>0.646147</td>
      <td>Positive regulation of gene expression (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>181</th>
      <td>GO:0045137</td>
      <td>0.645515</td>
      <td>Development of primary sexual characteristics (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>329</th>
      <td>GO:0006939</td>
      <td>0.637615</td>
      <td>Smooth muscle contraction (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>532</th>
      <td>GO:0010557</td>
      <td>0.628009</td>
      <td>Positive regulation of macromolecule biosynthetic process (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>579</th>
      <td>GO:0031328</td>
      <td>0.620431</td>
      <td>Positive regulation of cellular biosynthetic process (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>408</th>
      <td>GO:0048008</td>
      <td>0.615540</td>
      <td>Platelet-derived growth factor receptor signaling pathway (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>502</th>
      <td>GO:0008284</td>
      <td>0.594571</td>
      <td>Positive regulation of cell population proliferation (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>776</th>
      <td>GO:0072593</td>
      <td>0.591717</td>
      <td>Reactive oxygen species metabolic process (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>558</th>
      <td>GO:0021700</td>
      <td>0.587295</td>
      <td>Developmental maturation (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>192</th>
      <td>GO:0003254</td>
      <td>0.571930</td>
      <td>Regulation of membrane depolarization (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>869</th>
      <td>GO:1902533</td>
      <td>0.567760</td>
      <td>Positive regulation of intracellular signal transduction (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>144</th>
      <td>GO:0002318</td>
      <td>0.544781</td>
      <td>Myeloid progenitor cell differentiation (1)</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>120</th>
      <td>GO:0045860</td>
      <td>0.539516</td>
      <td>Positive regulation of protein kinase activity (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>741</th>
      <td>GO:0055001</td>
      <td>0.538092</td>
      <td>Muscle cell development (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>10</th>
      <td>GO:0043410</td>
      <td>0.531527</td>
      <td>Positive regulation of mapk cascade (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>133</th>
      <td>GO:0061138</td>
      <td>0.520278</td>
      <td>Morphogenesis of a branching epithelium (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>527</th>
      <td>GO:0010035</td>
      <td>0.514162</td>
      <td>Response to inorganic substance (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>1040</th>
      <td>GO:0042113</td>
      <td>0.509569</td>
      <td>B cell activation (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>595</th>
      <td>GO:0032526</td>
      <td>0.508940</td>
      <td>Response to retinoic acid (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>164</th>
      <td>GO:0030183</td>
      <td>0.500000</td>
      <td>B cell differentiation (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>793</th>
      <td>GO:1901888</td>
      <td>0.500000</td>
      <td>Regulation of cell junction assembly (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>1045</th>
      <td>GO:0051899</td>
      <td>0.500000</td>
      <td>Membrane depolarization (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>406</th>
      <td>GO:0038084</td>
      <td>0.491700</td>
      <td>Vascular endothelial growth factor signaling pathway (1)</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>227</th>
      <td>GO:0006366</td>
      <td>0.489305</td>
      <td>Transcription by rna polymerase ii (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>42</th>
      <td>GO:0001503</td>
      <td>0.478874</td>
      <td>Ossification (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>12</th>
      <td>GO:0070371</td>
      <td>0.477843</td>
      <td>Erk1 and erk2 cascade (1)</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>64</th>
      <td>GO:0001656</td>
      <td>0.476283</td>
      <td>Metanephros development (1)</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>1022</th>
      <td>GO:0048015</td>
      <td>0.460363</td>
      <td>Phosphatidylinositol-mediated signaling (1)</td>
      <td>3.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="144">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(x<span class="op">=</span>slim_matrix_single_neuron.loc[train_drug_logs[<span class="st">"GO_term"</span>],selected_drug_name], y<span class="op">=</span>train_drug_logs.set_index(<span class="st">"GO_term"</span>)[<span class="st">"probability"</span>], data<span class="op">=</span>plot,showfliers<span class="op">=</span><span class="va">True</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-140-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="145">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same as before</span></span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>sum_annotations <span class="op">=</span> slim_matrix_single_neuron.T.<span class="bu">sum</span>()<span class="op">/</span>slim_matrix_single_neuron.shape[<span class="dv">1</span>]</span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>logits_apriori<span class="op">=</span> np.log(sum_annotations<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>sum_annotations))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="146">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a>logits_apost<span class="op">=</span> np.log(train_drug_logs[<span class="st">"probability"</span>]<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>train_drug_logs[<span class="st">"probability"</span>]))</span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>delta_logits <span class="op">=</span> logits_apost.to_numpy()<span class="op">-</span>logits_apriori.to_numpy()</span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a>delta_logits_df <span class="op">=</span> pd.DataFrame(delta_logits)</span>
<span id="cb167-4"><a href="#cb167-4" aria-hidden="true" tabindex="-1"></a>delta_logits_df.columns <span class="op">=</span> [<span class="st">"delta_logits"</span>]</span>
<span id="cb167-5"><a href="#cb167-5" aria-hidden="true" tabindex="-1"></a>train_drug_mod <span class="op">=</span> train_drug_logs.merge(delta_logits_df, left_index<span class="op">=</span><span class="va">True</span>,right_index<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="147">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a>train_drug_mod.loc[train_drug_mod[<span class="st">"layer_number"</span>] <span class="op">&lt;=</span><span class="dv">3</span>].sort_values(by<span class="op">=</span>[<span class="st">"delta_logits"</span>], ascending<span class="op">=</span><span class="va">False</span>).head(<span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="147">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>probability</th>
      <th>Name</th>
      <th>layer_number</th>
      <th>delta_logits</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>329</th>
      <td>GO:0006939</td>
      <td>0.637615</td>
      <td>Smooth muscle contraction (1)</td>
      <td>3.0</td>
      <td>2.871769</td>
    </tr>
    <tr>
      <th>381</th>
      <td>GO:0048041</td>
      <td>0.737675</td>
      <td>Focal adhesion assembly (1)</td>
      <td>2.0</td>
      <td>2.568268</td>
    </tr>
    <tr>
      <th>144</th>
      <td>GO:0002318</td>
      <td>0.544781</td>
      <td>Myeloid progenitor cell differentiation (1)</td>
      <td>0.0</td>
      <td>2.533049</td>
    </tr>
    <tr>
      <th>408</th>
      <td>GO:0048008</td>
      <td>0.615540</td>
      <td>Platelet-derived growth factor receptor signaling pathway (1)</td>
      <td>1.0</td>
      <td>2.529048</td>
    </tr>
    <tr>
      <th>595</th>
      <td>GO:0032526</td>
      <td>0.508940</td>
      <td>Response to retinoic acid (1)</td>
      <td>1.0</td>
      <td>2.437785</td>
    </tr>
    <tr>
      <th>183</th>
      <td>GO:0048608</td>
      <td>0.742225</td>
      <td>Reproductive structure development (1)</td>
      <td>3.0</td>
      <td>2.222701</td>
    </tr>
    <tr>
      <th>458</th>
      <td>GO:0048839</td>
      <td>0.391941</td>
      <td>Inner ear development (1)</td>
      <td>3.0</td>
      <td>2.066365</td>
    </tr>
    <tr>
      <th>793</th>
      <td>GO:1901888</td>
      <td>0.500000</td>
      <td>Regulation of cell junction assembly (1)</td>
      <td>2.0</td>
      <td>2.058388</td>
    </tr>
    <tr>
      <th>956</th>
      <td>GO:0002327</td>
      <td>0.382026</td>
      <td>Immature b cell differentiation (1)</td>
      <td>0.0</td>
      <td>1.921062</td>
    </tr>
    <tr>
      <th>192</th>
      <td>GO:0003254</td>
      <td>0.571930</td>
      <td>Regulation of membrane depolarization (1)</td>
      <td>1.0</td>
      <td>1.903704</td>
    </tr>
    <tr>
      <th>181</th>
      <td>GO:0045137</td>
      <td>0.645515</td>
      <td>Development of primary sexual characteristics (1)</td>
      <td>3.0</td>
      <td>1.806630</td>
    </tr>
    <tr>
      <th>1029</th>
      <td>GO:0035909</td>
      <td>0.293492</td>
      <td>Aorta morphogenesis (1)</td>
      <td>1.0</td>
      <td>1.801732</td>
    </tr>
    <tr>
      <th>1118</th>
      <td>GO:0097530</td>
      <td>0.339315</td>
      <td>Granulocyte migration (1)</td>
      <td>3.0</td>
      <td>1.786297</td>
    </tr>
    <tr>
      <th>24</th>
      <td>GO:0045840</td>
      <td>0.365727</td>
      <td>Positive regulation of mitotic nuclear division (1)</td>
      <td>1.0</td>
      <td>1.756150</td>
    </tr>
    <tr>
      <th>65</th>
      <td>GO:0072075</td>
      <td>0.273943</td>
      <td>Metanephric mesenchyme development (1)</td>
      <td>0.0</td>
      <td>1.705507</td>
    </tr>
    <tr>
      <th>958</th>
      <td>GO:0045580</td>
      <td>0.389135</td>
      <td>Regulation of t cell differentiation (1)</td>
      <td>3.0</td>
      <td>1.684905</td>
    </tr>
    <tr>
      <th>406</th>
      <td>GO:0038084</td>
      <td>0.491700</td>
      <td>Vascular endothelial growth factor signaling pathway (1)</td>
      <td>1.0</td>
      <td>1.636301</td>
    </tr>
    <tr>
      <th>474</th>
      <td>GO:0007585</td>
      <td>0.259741</td>
      <td>Respiratory gaseous exchange by respiratory system (1)</td>
      <td>2.0</td>
      <td>1.632898</td>
    </tr>
    <tr>
      <th>936</th>
      <td>GO:0035584</td>
      <td>0.282146</td>
      <td>Calcium-mediated signaling using intracellular calcium source (1)</td>
      <td>0.0</td>
      <td>1.627051</td>
    </tr>
    <tr>
      <th>776</th>
      <td>GO:0072593</td>
      <td>0.591717</td>
      <td>Reactive oxygen species metabolic process (1)</td>
      <td>3.0</td>
      <td>1.621405</td>
    </tr>
    <tr>
      <th>479</th>
      <td>GO:0050910</td>
      <td>0.256674</td>
      <td>Detection of mechanical stimulus involved in sensory perception of sound (1)</td>
      <td>0.0</td>
      <td>1.616889</td>
    </tr>
    <tr>
      <th>961</th>
      <td>GO:0055003</td>
      <td>0.261761</td>
      <td>Cardiac myofibril assembly (1)</td>
      <td>0.0</td>
      <td>1.582183</td>
    </tr>
    <tr>
      <th>741</th>
      <td>GO:0055001</td>
      <td>0.538092</td>
      <td>Muscle cell development (1)</td>
      <td>3.0</td>
      <td>1.562711</td>
    </tr>
    <tr>
      <th>558</th>
      <td>GO:0021700</td>
      <td>0.587295</td>
      <td>Developmental maturation (1)</td>
      <td>3.0</td>
      <td>1.560039</td>
    </tr>
    <tr>
      <th>609</th>
      <td>GO:0034103</td>
      <td>0.287738</td>
      <td>Regulation of tissue remodeling (1)</td>
      <td>1.0</td>
      <td>1.546250</td>
    </tr>
    <tr>
      <th>1045</th>
      <td>GO:0051899</td>
      <td>0.500000</td>
      <td>Membrane depolarization (1)</td>
      <td>2.0</td>
      <td>1.534347</td>
    </tr>
    <tr>
      <th>590</th>
      <td>GO:0032355</td>
      <td>0.380108</td>
      <td>Response to estradiol (1)</td>
      <td>1.0</td>
      <td>1.532244</td>
    </tr>
    <tr>
      <th>129</th>
      <td>GO:0051894</td>
      <td>0.341633</td>
      <td>Positive regulation of focal adhesion assembly (1)</td>
      <td>0.0</td>
      <td>1.520410</td>
    </tr>
    <tr>
      <th>64</th>
      <td>GO:0001656</td>
      <td>0.476283</td>
      <td>Metanephros development (1)</td>
      <td>3.0</td>
      <td>1.492026</td>
    </tr>
    <tr>
      <th>428</th>
      <td>GO:0007266</td>
      <td>0.373714</td>
      <td>Rho protein signal transduction (1)</td>
      <td>2.0</td>
      <td>1.468985</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="148">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(x<span class="op">=</span>slim_matrix_single_neuron.loc[train_drug_mod[<span class="st">"GO_term"</span>],selected_drug_name], y<span class="op">=</span>train_drug_mod.set_index(<span class="st">"GO_term"</span>)[<span class="st">"delta_logits"</span>], data<span class="op">=</span>plot,showfliers<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="DeepMoA_files/figure-html/cell-144-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="svm-go-term-2d-representation" class="level2">
<h2 class="anchored" data-anchor-id="svm-go-term-2d-representation">SVM GO TERM 2D representation</h2>
<div class="cell" data-execution_count="149">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.manifold <span class="im">import</span> TSNE</span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.express <span class="im">as</span> px</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Choose go to study…</strong></p>
<div class="cell" data-execution_count="150">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a>display(combobox_go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"403eb665201541cf8fcc0765d10235ec","version_major":2,"version_minor":0}
</script>
</div>
</div>
<div class="cell" data-execution_count="151">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a>selected_goterm <span class="op">=</span> combobox_go.result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="152">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>real_go_info[real_go_info[<span class="st">"GO_term"</span>]<span class="op">==</span>selected_goterm<span class="op">+</span><span class="st">"_1"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="152">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>GO_term</th>
      <th>Name</th>
      <th>layer_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>24</th>
      <td>GO:0007051_1</td>
      <td>Spindle organization (1)</td>
      <td>3.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="153">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>list_nodes <span class="op">=</span> []</span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">7</span>):</span>
<span id="cb174-3"><a href="#cb174-3" aria-hidden="true" tabindex="-1"></a>    list_nodes.append(selected_goterm<span class="op">+</span><span class="st">"_"</span><span class="op">+</span><span class="bu">str</span>(i))</span>
<span id="cb174-4"><a href="#cb174-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb174-5"><a href="#cb174-5" aria-hidden="true" tabindex="-1"></a>score <span class="op">=</span> attribution_data_annotated.loc[list_nodes].T</span>
<span id="cb174-6"><a href="#cb174-6" aria-hidden="true" tabindex="-1"></a>score_mod <span class="op">=</span> score.divide(score.std()).fillna(<span class="dv">0</span>) </span>
<span id="cb174-7"><a href="#cb174-7" aria-hidden="true" tabindex="-1"></a>annotations  <span class="op">=</span>slim_matrix_single_neuron.loc[selected_goterm,]</span>
<span id="cb174-8"><a href="#cb174-8" aria-hidden="true" tabindex="-1"></a>y_predicted <span class="op">=</span> models_svm[selected_goterm].predict(score_mod.astype(<span class="bu">float</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="plot-svm" class="level3">
<h3 class="anchored" data-anchor-id="plot-svm">Plot SVM</h3>
<section id="view-statistics-of-goterm" class="level4">
<h4 class="anchored" data-anchor-id="view-statistics-of-goterm">View statistics of GOterm</h4>
<p><em>“Perfect” model (with train data)</em></p>
<div class="cell" data-execution_count="154">
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a>auc <span class="op">=</span> metrics.roc_auc_score(annotations, models_svm[selected_goterm].decision_function(score_mod.astype(<span class="bu">float</span>)))</span>
<span id="cb175-2"><a href="#cb175-2" aria-hidden="true" tabindex="-1"></a>cnf_matrix <span class="op">=</span> metrics.confusion_matrix(annotations,y_predicted)</span>
<span id="cb175-3"><a href="#cb175-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cnf_matrix)</span>
<span id="cb175-4"><a href="#cb175-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb175-5"><a href="#cb175-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Accuracy:"</span>,metrics.accuracy_score(annotations, y_predicted))</span>
<span id="cb175-6"><a href="#cb175-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Precision:"</span>,metrics.precision_score(annotations,y_predicted)) <span class="co"># TP / (TP+FP)</span></span>
<span id="cb175-7"><a href="#cb175-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Recall:"</span>,metrics.recall_score(annotations, y_predicted)) <span class="co">#TP / (TP+FN)</span></span>
<span id="cb175-8"><a href="#cb175-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"AUC with score:"</span>,auc) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[227  19]
 [  3  16]]
Accuracy: 0.9169811320754717
Precision: 0.45714285714285713
Recall: 0.8421052631578947
AUC with score: 0.9531450577663672</code></pre>
</div>
</div>
<p>TN - FP</p>
<p>FN - TP</p>
<p>En mi opinion interesa mucho el precision, prefiero que haya menos FP no??</p>
<p><em>Test statistics…</em></p>
<div class="cell" data-execution_count="155">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>auc <span class="op">=</span> metrics.roc_auc_score(slim_matrix_single_neuron.loc[selected_goterm],  platt_matrix.loc[selected_goterm])</span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>cnf_matrix <span class="op">=</span> metrics.confusion_matrix(slim_matrix_single_neuron.loc[selected_goterm], preds_svm_matrix.loc[selected_goterm])</span>
<span id="cb177-3"><a href="#cb177-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cnf_matrix)</span>
<span id="cb177-4"><a href="#cb177-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-5"><a href="#cb177-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Accuracy:"</span>,metrics.accuracy_score(slim_matrix_single_neuron.loc[selected_goterm], preds_svm_matrix.loc[selected_goterm]))</span>
<span id="cb177-6"><a href="#cb177-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Precision:"</span>,metrics.precision_score(slim_matrix_single_neuron.loc[selected_goterm], preds_svm_matrix.loc[selected_goterm]))</span>
<span id="cb177-7"><a href="#cb177-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Recall:"</span>,metrics.recall_score(slim_matrix_single_neuron.loc[selected_goterm], preds_svm_matrix.loc[selected_goterm])) <span class="co">#TP / (TP+FN)</span></span>
<span id="cb177-8"><a href="#cb177-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"AUC with score:"</span>,auc) <span class="co">#TP / (TP+FN)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[225  21]
 [  6  13]]
Accuracy: 0.8981132075471698
Precision: 0.38235294117647056
Recall: 0.6842105263157895
AUC with score: 0.891313649978605</code></pre>
</div>
</div>
<div class="cell" data-execution_count="157">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> colorlover <span class="im">as</span> cl</span>
<span id="cb179-2"><a href="#cb179-2" aria-hidden="true" tabindex="-1"></a>matrix <span class="op">=</span> metrics.confusion_matrix(annotations,y_predicted)</span>
<span id="cb179-3"><a href="#cb179-3" aria-hidden="true" tabindex="-1"></a>tn, fp, fn, tp <span class="op">=</span> matrix.ravel()</span>
<span id="cb179-4"><a href="#cb179-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb179-5"><a href="#cb179-5" aria-hidden="true" tabindex="-1"></a>values <span class="op">=</span> [tp, fn, fp, tn]</span>
<span id="cb179-6"><a href="#cb179-6" aria-hidden="true" tabindex="-1"></a>label_text <span class="op">=</span> [<span class="st">"True Positive"</span>, <span class="st">"False Negative"</span>, <span class="st">"False Positive"</span>, <span class="st">"True Negative"</span>]</span>
<span id="cb179-7"><a href="#cb179-7" aria-hidden="true" tabindex="-1"></a>labels <span class="op">=</span> [<span class="st">"&lt;b&gt;TP&lt;/b&gt;"</span>, <span class="st">"&lt;b&gt;FN&lt;/b&gt;"</span>, <span class="st">"&lt;b&gt;FP&lt;/b&gt;"</span>, <span class="st">"&lt;b&gt;TN&lt;/b&gt;"</span>]</span>
<span id="cb179-8"><a href="#cb179-8" aria-hidden="true" tabindex="-1"></a>blue <span class="op">=</span> cl.flipper()[<span class="st">"seq"</span>][<span class="st">"9"</span>][<span class="st">"Blues"</span>]</span>
<span id="cb179-9"><a href="#cb179-9" aria-hidden="true" tabindex="-1"></a>red <span class="op">=</span> cl.flipper()[<span class="st">"seq"</span>][<span class="st">"9"</span>][<span class="st">"Reds"</span>]</span>
<span id="cb179-10"><a href="#cb179-10" aria-hidden="true" tabindex="-1"></a>colors <span class="op">=</span> [<span class="st">"#ff3700"</span>,<span class="st">"#FFA0A0"</span>, <span class="st">"#CCE9FF"</span>,  <span class="st">"#0b8bff"</span>]</span>
<span id="cb179-11"><a href="#cb179-11" aria-hidden="true" tabindex="-1"></a>trace0 <span class="op">=</span> go.Pie(</span>
<span id="cb179-12"><a href="#cb179-12" aria-hidden="true" tabindex="-1"></a>    labels<span class="op">=</span>label_text,</span>
<span id="cb179-13"><a href="#cb179-13" aria-hidden="true" tabindex="-1"></a>    values<span class="op">=</span>values,</span>
<span id="cb179-14"><a href="#cb179-14" aria-hidden="true" tabindex="-1"></a>    hoverinfo<span class="op">=</span><span class="st">"label+value+percent"</span>,</span>
<span id="cb179-15"><a href="#cb179-15" aria-hidden="true" tabindex="-1"></a>    textinfo<span class="op">=</span><span class="st">"text+value"</span>,</span>
<span id="cb179-16"><a href="#cb179-16" aria-hidden="true" tabindex="-1"></a>    text<span class="op">=</span>labels,</span>
<span id="cb179-17"><a href="#cb179-17" aria-hidden="true" tabindex="-1"></a>    sort<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb179-18"><a href="#cb179-18" aria-hidden="true" tabindex="-1"></a>    marker<span class="op">=</span><span class="bu">dict</span>(colors<span class="op">=</span>colors),</span>
<span id="cb179-19"><a href="#cb179-19" aria-hidden="true" tabindex="-1"></a>    insidetextfont<span class="op">=</span>{<span class="st">"color"</span>: <span class="st">"#36382E"</span>},</span>
<span id="cb179-20"><a href="#cb179-20" aria-hidden="true" tabindex="-1"></a>    rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb179-21"><a href="#cb179-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb179-22"><a href="#cb179-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb179-23"><a href="#cb179-23" aria-hidden="true" tabindex="-1"></a>layout <span class="op">=</span> go.Layout(</span>
<span id="cb179-24"><a href="#cb179-24" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="bu">dict</span>(text<span class="op">=</span><span class="st">"Confusion Matrix"</span>,</span>
<span id="cb179-25"><a href="#cb179-25" aria-hidden="true" tabindex="-1"></a>              x<span class="op">=</span><span class="fl">0.3</span>,</span>
<span id="cb179-26"><a href="#cb179-26" aria-hidden="true" tabindex="-1"></a>              y<span class="op">=</span><span class="fl">0.8</span>,</span>
<span id="cb179-27"><a href="#cb179-27" aria-hidden="true" tabindex="-1"></a>              font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">14</span>),</span>
<span id="cb179-28"><a href="#cb179-28" aria-hidden="true" tabindex="-1"></a>              xanchor<span class="op">=</span><span class="st">'center'</span>,</span>
<span id="cb179-29"><a href="#cb179-29" aria-hidden="true" tabindex="-1"></a>              yanchor<span class="op">=</span><span class="st">'top'</span>),</span>
<span id="cb179-30"><a href="#cb179-30" aria-hidden="true" tabindex="-1"></a>    <span class="co">#margin=dict(l=50, r=50, t=100, b=10),</span></span>
<span id="cb179-31"><a href="#cb179-31" aria-hidden="true" tabindex="-1"></a>    legend<span class="op">=</span><span class="bu">dict</span>(font<span class="op">=</span>{<span class="st">"color"</span>: <span class="st">"#36382E"</span>}, orientation<span class="op">=</span><span class="st">"h"</span>,x<span class="op">=</span><span class="fl">0.1</span>, y<span class="op">=-</span><span class="fl">0.03</span>),</span>
<span id="cb179-32"><a href="#cb179-32" aria-hidden="true" tabindex="-1"></a><span class="co">#    plot_bgcolor="#282b38",</span></span>
<span id="cb179-33"><a href="#cb179-33" aria-hidden="true" tabindex="-1"></a><span class="co">#    paper_bgcolor="#282b38",</span></span>
<span id="cb179-34"><a href="#cb179-34" aria-hidden="true" tabindex="-1"></a>    font<span class="op">=</span><span class="bu">dict</span>(family<span class="op">=</span><span class="st">'Roboto'</span>,color<span class="op">=</span> <span class="st">"#36382E"</span>),</span>
<span id="cb179-35"><a href="#cb179-35" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb179-36"><a href="#cb179-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb179-37"><a href="#cb179-37" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> [trace0]</span>
<span id="cb179-38"><a href="#cb179-38" aria-hidden="true" tabindex="-1"></a>figure <span class="op">=</span> go.Figure(data<span class="op">=</span>data, layout<span class="op">=</span>layout)</span>
<span id="cb179-39"><a href="#cb179-39" aria-hidden="true" tabindex="-1"></a>figure</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Unable to display output for mime type(s): application/vnd.plotly.v1+json</code></pre>
</div>
</div>
<div class="cell" data-execution_count="158">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a>y_test<span class="op">=</span>annotations</span>
<span id="cb181-2"><a href="#cb181-2" aria-hidden="true" tabindex="-1"></a>decision_test<span class="op">=</span>y_predicted</span>
<span id="cb181-3"><a href="#cb181-3" aria-hidden="true" tabindex="-1"></a>fpr, tpr, threshold <span class="op">=</span> metrics.roc_curve(y_test, decision_test)</span>
<span id="cb181-4"><a href="#cb181-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-5"><a href="#cb181-5" aria-hidden="true" tabindex="-1"></a><span class="co"># AUC Score</span></span>
<span id="cb181-6"><a href="#cb181-6" aria-hidden="true" tabindex="-1"></a>auc_score <span class="op">=</span> metrics.roc_auc_score(y_true<span class="op">=</span>y_test, y_score<span class="op">=</span>decision_test)</span>
<span id="cb181-7"><a href="#cb181-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-8"><a href="#cb181-8" aria-hidden="true" tabindex="-1"></a>trace0 <span class="op">=</span> go.Scatter(</span>
<span id="cb181-9"><a href="#cb181-9" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>fpr, y<span class="op">=</span>tpr, mode<span class="op">=</span><span class="st">"lines"</span>, name<span class="op">=</span><span class="st">"Test Data"</span>, marker<span class="op">=</span>{<span class="st">"color"</span>: <span class="st">"#ff3700"</span>}</span>
<span id="cb181-10"><a href="#cb181-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb181-11"><a href="#cb181-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-12"><a href="#cb181-12" aria-hidden="true" tabindex="-1"></a>layout <span class="op">=</span> go.Layout(</span>
<span id="cb181-13"><a href="#cb181-13" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="bu">dict</span>(text<span class="op">=</span><span class="ss">f"ROC Curve (AUC = </span><span class="sc">{</span>auc_score<span class="sc">:.3f}</span><span class="ss">)"</span>,</span>
<span id="cb181-14"><a href="#cb181-14" aria-hidden="true" tabindex="-1"></a>            x<span class="op">=</span><span class="fl">0.6</span>,</span>
<span id="cb181-15"><a href="#cb181-15" aria-hidden="true" tabindex="-1"></a>            y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb181-16"><a href="#cb181-16" aria-hidden="true" tabindex="-1"></a>            font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb181-17"><a href="#cb181-17" aria-hidden="true" tabindex="-1"></a>              ),</span>
<span id="cb181-18"><a href="#cb181-18" aria-hidden="true" tabindex="-1"></a>    xaxis<span class="op">=</span><span class="bu">dict</span>(title<span class="op">=</span><span class="st">"False Positive Rate"</span>, gridcolor<span class="op">=</span><span class="st">"white"</span>),</span>
<span id="cb181-19"><a href="#cb181-19" aria-hidden="true" tabindex="-1"></a>    yaxis<span class="op">=</span><span class="bu">dict</span>(title<span class="op">=</span><span class="st">"True Positive Rate"</span>, gridcolor<span class="op">=</span><span class="st">"white"</span>),</span>
<span id="cb181-20"><a href="#cb181-20" aria-hidden="true" tabindex="-1"></a>    legend<span class="op">=</span><span class="bu">dict</span>(x<span class="op">=</span><span class="dv">0</span>, y<span class="op">=</span><span class="fl">1.05</span>, orientation<span class="op">=</span><span class="st">"h"</span>),</span>
<span id="cb181-21"><a href="#cb181-21" aria-hidden="true" tabindex="-1"></a>    margin<span class="op">=</span><span class="bu">dict</span>(l<span class="op">=</span><span class="dv">100</span>, r<span class="op">=</span><span class="dv">10</span>, t<span class="op">=</span><span class="dv">25</span>, b<span class="op">=</span><span class="dv">40</span>),</span>
<span id="cb181-22"><a href="#cb181-22" aria-hidden="true" tabindex="-1"></a><span class="co">#    plot_bgcolor="#282b38",</span></span>
<span id="cb181-23"><a href="#cb181-23" aria-hidden="true" tabindex="-1"></a><span class="co">#    paper_bgcolor="#282b38",</span></span>
<span id="cb181-24"><a href="#cb181-24" aria-hidden="true" tabindex="-1"></a>    font<span class="op">=</span><span class="bu">dict</span>(family<span class="op">=</span><span class="st">'Roboto'</span>,color<span class="op">=</span> <span class="st">"#36382E"</span>),</span>
<span id="cb181-25"><a href="#cb181-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb181-26"><a href="#cb181-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-27"><a href="#cb181-27" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> [trace0]</span>
<span id="cb181-28"><a href="#cb181-28" aria-hidden="true" tabindex="-1"></a>figure <span class="op">=</span> go.Figure(data<span class="op">=</span>data, layout<span class="op">=</span>layout)</span>
<span id="cb181-29"><a href="#cb181-29" aria-hidden="true" tabindex="-1"></a>figure</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Unable to display output for mime type(s): application/vnd.plotly.v1+json</code></pre>
</div>
</div>
</section>
<section id="plot-svm-with-unknown-labels" class="level4">
<h4 class="anchored" data-anchor-id="plot-svm-with-unknown-labels">Plot SVM with unknown labels</h4>
<section id="voronoi-tessellation" class="level5">
<h5 class="anchored" data-anchor-id="voronoi-tessellation">Voronoi Tessellation</h5>
<p>What is a Voronoi Tessellation? Given a set P := {p1, …, pn} of sites, a Voronoi Tessellation is a subdivision of the space into n cells, one for each site in P, with the property that a point q lies in the cell corresponding to a site pi iff d(pi, q) &lt; d(pj, q) for i distinct from j. The segments in a Voronoi Tessellation correspond to all points in the plane equidistant to the two nearest sites. Voronoi Tessellations have applications in computer science.</p>
<p>https://stackoverflow.com/questions/61225052/svm-plot-decision-surface-when-working-with-more-than-2-features</p>
<div class="cell" data-execution_count="159">
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a>tsne <span class="op">=</span> TSNE(n_components<span class="op">=</span><span class="dv">2</span>, verbose<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb183-2"><a href="#cb183-2" aria-hidden="true" tabindex="-1"></a>           init<span class="op">=</span><span class="st">"pca"</span>,</span>
<span id="cb183-3"><a href="#cb183-3" aria-hidden="true" tabindex="-1"></a>            perplexity<span class="op">=</span><span class="dv">30</span>,</span>
<span id="cb183-4"><a href="#cb183-4" aria-hidden="true" tabindex="-1"></a>            random_state<span class="op">=</span><span class="dv">123</span></span>
<span id="cb183-5"><a href="#cb183-5" aria-hidden="true" tabindex="-1"></a>           )</span>
<span id="cb183-6"><a href="#cb183-6" aria-hidden="true" tabindex="-1"></a>z <span class="op">=</span> tsne.fit_transform(score_mod.astype(<span class="bu">float</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="160">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a>list_nodes <span class="op">=</span> <span class="bu">list</span>(models_svm[selected_goterm].feature_names_in_) <span class="co"># Extract the feature names from the model (those are the attributions we need)</span></span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a>score_unknown <span class="op">=</span> attribution_data_all.loc[list_nodes,unknown].T</span>
<span id="cb184-3"><a href="#cb184-3" aria-hidden="true" tabindex="-1"></a>score_unknown_mod <span class="op">=</span> score_unknown.divide(score.std()).fillna(<span class="dv">0</span>) <span class="co"># normalize</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="161">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>y_unknown <span class="op">=</span> np.full(score_unknown_mod.shape[<span class="dv">0</span>],<span class="dv">2</span>) <span class="co"># 2=unknown MOA</span></span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>y_pred_unknown <span class="op">=</span> models_svm[selected_goterm].predict(score_unknown_mod.astype(<span class="bu">float</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="162">
<div class="sourceCode cell-code" id="cb186"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb186-1"><a href="#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join scores and annotations from known and unknown drugs</span></span>
<span id="cb186-2"><a href="#cb186-2" aria-hidden="true" tabindex="-1"></a>all_score <span class="op">=</span> pd.concat([score_mod,score_unknown_mod])</span>
<span id="cb186-3"><a href="#cb186-3" aria-hidden="true" tabindex="-1"></a>all_y <span class="op">=</span> np.concatenate((annotations,y_unknown))  <span class="co"># 2=unknown MOA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot T-SNE SVM</p>
<div class="cell" data-execution_count="163">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.neighbors._classification <span class="im">import</span> KNeighborsClassifier</span>
<span id="cb187-2"><a href="#cb187-2" aria-hidden="true" tabindex="-1"></a><span class="co"># https://github.com/plotly/dash-sample-apps/blob/main/apps/dash-svm/utils/dash_reusable_components.py</span></span>
<span id="cb187-3"><a href="#cb187-3" aria-hidden="true" tabindex="-1"></a>z <span class="op">=</span> tsne.fit_transform(all_score.astype(<span class="bu">float</span>)) </span>
<span id="cb187-4"><a href="#cb187-4" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame()</span>
<span id="cb187-5"><a href="#cb187-5" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"y"</span>] <span class="op">=</span> all_y</span>
<span id="cb187-6"><a href="#cb187-6" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"comp-1"</span>] <span class="op">=</span> z[:,<span class="dv">0</span>]</span>
<span id="cb187-7"><a href="#cb187-7" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"comp-2"</span>] <span class="op">=</span> z[:,<span class="dv">1</span>]</span>
<span id="cb187-8"><a href="#cb187-8" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"name"</span>] <span class="op">=</span><span class="bu">list</span>(all_score.index)</span>
<span id="cb187-9"><a href="#cb187-9" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.sort_values(by<span class="op">=</span>[<span class="st">'y'</span>])</span>
<span id="cb187-10"><a href="#cb187-10" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"y"</span>] <span class="op">=</span> df[<span class="st">"y"</span>].astype(<span class="bu">str</span>)</span>
<span id="cb187-11"><a href="#cb187-11" aria-hidden="true" tabindex="-1"></a>X,y <span class="op">=</span> all_score.astype(<span class="bu">float</span>), all_y</span>
<span id="cb187-12"><a href="#cb187-12" aria-hidden="true" tabindex="-1"></a>y_predicted <span class="op">=</span> models_svm[selected_goterm].predict(X)</span>
<span id="cb187-13"><a href="#cb187-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-14"><a href="#cb187-14" aria-hidden="true" tabindex="-1"></a>resolution <span class="op">=</span> <span class="dv">300</span> <span class="co"># 100x100 background pixels</span></span>
<span id="cb187-15"><a href="#cb187-15" aria-hidden="true" tabindex="-1"></a>X2d_xmin, X2d_xmax <span class="op">=</span> np.<span class="bu">min</span>(z[:,<span class="dv">0</span>])<span class="op">-</span><span class="dv">1</span>, np.<span class="bu">max</span>(z[:,<span class="dv">0</span>])<span class="op">+</span><span class="dv">1</span></span>
<span id="cb187-16"><a href="#cb187-16" aria-hidden="true" tabindex="-1"></a>X2d_ymin, X2d_ymax <span class="op">=</span> np.<span class="bu">min</span>(z[:,<span class="dv">1</span>])<span class="op">-</span><span class="dv">1</span>, np.<span class="bu">max</span>(z[:,<span class="dv">1</span>])<span class="op">+</span><span class="dv">1</span></span>
<span id="cb187-17"><a href="#cb187-17" aria-hidden="true" tabindex="-1"></a>xx, yy <span class="op">=</span> np.meshgrid(np.linspace(X2d_xmin, X2d_xmax, resolution), np.linspace(X2d_ymin, X2d_ymax, resolution))</span>
<span id="cb187-18"><a href="#cb187-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-19"><a href="#cb187-19" aria-hidden="true" tabindex="-1"></a><span class="co"># approximate Voronoi tesselation on resolution x resolution grid using 1-NN</span></span>
<span id="cb187-20"><a href="#cb187-20" aria-hidden="true" tabindex="-1"></a>background_model <span class="op">=</span> KNeighborsClassifier(n_neighbors<span class="op">=</span><span class="dv">1</span>).fit(z, y_predicted) </span>
<span id="cb187-21"><a href="#cb187-21" aria-hidden="true" tabindex="-1"></a>voronoiBackground <span class="op">=</span> background_model.predict(np.c_[xx.ravel(), yy.ravel()])</span>
<span id="cb187-22"><a href="#cb187-22" aria-hidden="true" tabindex="-1"></a>voronoiBackground <span class="op">=</span> voronoiBackground.reshape((resolution, resolution))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="164">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a>go_name<span class="op">=</span>real_go_info[real_go_info[<span class="st">"GO_term"</span>]<span class="op">==</span>selected_goterm<span class="op">+</span><span class="st">"_1"</span>][<span class="st">"Name"</span>].values[<span class="dv">0</span>][:<span class="op">-</span><span class="dv">4</span>]</span>
<span id="cb188-2"><a href="#cb188-2" aria-hidden="true" tabindex="-1"></a>go_name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="164">
<pre><code>'Spindle organization'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="165">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a>bright_cscale <span class="op">=</span> [[<span class="dv">0</span>, <span class="st">"#0b8bff"</span>], [<span class="fl">0.5</span>, <span class="st">"#ff3700"</span>],[<span class="dv">1</span>, <span class="st">"#36382E"</span>]]</span>
<span id="cb190-2"><a href="#cb190-2" aria-hidden="true" tabindex="-1"></a>new_cscale <span class="op">=</span> [[<span class="dv">0</span>, <span class="st">"#CCE9FF"</span>], [<span class="dv">1</span>, <span class="st">"#FFA0A0"</span>]]</span>
<span id="cb190-3"><a href="#cb190-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-4"><a href="#cb190-4" aria-hidden="true" tabindex="-1"></a>trace0 <span class="op">=</span> go.Contour(</span>
<span id="cb190-5"><a href="#cb190-5" aria-hidden="true" tabindex="-1"></a>        x<span class="op">=</span>xx.flatten(),</span>
<span id="cb190-6"><a href="#cb190-6" aria-hidden="true" tabindex="-1"></a>        y<span class="op">=</span>yy.flatten(),</span>
<span id="cb190-7"><a href="#cb190-7" aria-hidden="true" tabindex="-1"></a>        z<span class="op">=</span>voronoiBackground.flatten(),</span>
<span id="cb190-8"><a href="#cb190-8" aria-hidden="true" tabindex="-1"></a>        hoverinfo<span class="op">=</span><span class="st">"none"</span>,</span>
<span id="cb190-9"><a href="#cb190-9" aria-hidden="true" tabindex="-1"></a>        showscale<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb190-10"><a href="#cb190-10" aria-hidden="true" tabindex="-1"></a>        contours<span class="op">=</span><span class="bu">dict</span>(showlines<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb190-11"><a href="#cb190-11" aria-hidden="true" tabindex="-1"></a>        colorscale<span class="op">=</span>new_cscale,</span>
<span id="cb190-12"><a href="#cb190-12" aria-hidden="true" tabindex="-1"></a>        opacity<span class="op">=</span><span class="fl">0.9</span>,</span>
<span id="cb190-13"><a href="#cb190-13" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb190-14"><a href="#cb190-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb190-15"><a href="#cb190-15" aria-hidden="true" tabindex="-1"></a>trace1 <span class="op">=</span> go.Contour(</span>
<span id="cb190-16"><a href="#cb190-16" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>xx.flatten(),</span>
<span id="cb190-17"><a href="#cb190-17" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>yy.flatten(),</span>
<span id="cb190-18"><a href="#cb190-18" aria-hidden="true" tabindex="-1"></a>    z<span class="op">=</span>voronoiBackground.flatten(),</span>
<span id="cb190-19"><a href="#cb190-19" aria-hidden="true" tabindex="-1"></a>    showscale<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb190-20"><a href="#cb190-20" aria-hidden="true" tabindex="-1"></a>    hoverinfo<span class="op">=</span><span class="st">"none"</span>,</span>
<span id="cb190-21"><a href="#cb190-21" aria-hidden="true" tabindex="-1"></a>    colorscale<span class="op">=</span>new_cscale,</span>
<span id="cb190-22"><a href="#cb190-22" aria-hidden="true" tabindex="-1"></a>    line<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">"#ff3700"</span>),</span>
<span id="cb190-23"><a href="#cb190-23" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb190-24"><a href="#cb190-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-25"><a href="#cb190-25" aria-hidden="true" tabindex="-1"></a>trace2 <span class="op">=</span> go.Scatter(</span>
<span id="cb190-26"><a href="#cb190-26" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>df[<span class="st">"comp-1"</span>],</span>
<span id="cb190-27"><a href="#cb190-27" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>df[<span class="st">"comp-2"</span>],</span>
<span id="cb190-28"><a href="#cb190-28" aria-hidden="true" tabindex="-1"></a>    mode<span class="op">=</span><span class="st">"markers"</span>,</span>
<span id="cb190-29"><a href="#cb190-29" aria-hidden="true" tabindex="-1"></a>    text<span class="op">=</span>df[<span class="st">"name"</span>].to_numpy(),</span>
<span id="cb190-30"><a href="#cb190-30" aria-hidden="true" tabindex="-1"></a>    marker<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">7</span>, color<span class="op">=</span>df[<span class="st">"y"</span>].to_numpy(<span class="bu">int</span>),colorscale<span class="op">=</span>bright_cscale),</span>
<span id="cb190-31"><a href="#cb190-31" aria-hidden="true" tabindex="-1"></a>    showlegend<span class="op">=</span><span class="va">False</span></span>
<span id="cb190-32"><a href="#cb190-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb190-33"><a href="#cb190-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-34"><a href="#cb190-34" aria-hidden="true" tabindex="-1"></a>legend1 <span class="op">=</span> go.Scatter(</span>
<span id="cb190-35"><a href="#cb190-35" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>[<span class="va">None</span>],</span>
<span id="cb190-36"><a href="#cb190-36" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>[<span class="va">None</span>],</span>
<span id="cb190-37"><a href="#cb190-37" aria-hidden="true" tabindex="-1"></a>    mode<span class="op">=</span><span class="st">"markers"</span>,</span>
<span id="cb190-38"><a href="#cb190-38" aria-hidden="true" tabindex="-1"></a>    name<span class="op">=</span><span class="st">"Not annotated&lt;br&gt;to "</span><span class="op">+</span>selected_goterm,</span>
<span id="cb190-39"><a href="#cb190-39" aria-hidden="true" tabindex="-1"></a>    marker<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">7</span>, color<span class="op">=</span><span class="st">"#0b8bff"</span>,symbol<span class="op">=</span><span class="st">'circle'</span>),</span>
<span id="cb190-40"><a href="#cb190-40" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb190-41"><a href="#cb190-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-42"><a href="#cb190-42" aria-hidden="true" tabindex="-1"></a>legend2 <span class="op">=</span> go.Scatter(</span>
<span id="cb190-43"><a href="#cb190-43" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>[<span class="va">None</span>],</span>
<span id="cb190-44"><a href="#cb190-44" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>[<span class="va">None</span>],</span>
<span id="cb190-45"><a href="#cb190-45" aria-hidden="true" tabindex="-1"></a>    mode<span class="op">=</span><span class="st">"markers"</span>,</span>
<span id="cb190-46"><a href="#cb190-46" aria-hidden="true" tabindex="-1"></a>    name<span class="op">=</span><span class="st">"Drug annotated&lt;br&gt;to "</span><span class="op">+</span>selected_goterm,</span>
<span id="cb190-47"><a href="#cb190-47" aria-hidden="true" tabindex="-1"></a>    marker<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">7</span>, color<span class="op">=</span><span class="st">"#ff3700"</span>,symbol<span class="op">=</span><span class="st">'circle'</span>),</span>
<span id="cb190-48"><a href="#cb190-48" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb190-49"><a href="#cb190-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-50"><a href="#cb190-50" aria-hidden="true" tabindex="-1"></a>legend3 <span class="op">=</span> go.Scatter(</span>
<span id="cb190-51"><a href="#cb190-51" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>[<span class="va">None</span>],</span>
<span id="cb190-52"><a href="#cb190-52" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>[<span class="va">None</span>],</span>
<span id="cb190-53"><a href="#cb190-53" aria-hidden="true" tabindex="-1"></a>    mode<span class="op">=</span><span class="st">"markers"</span>,</span>
<span id="cb190-54"><a href="#cb190-54" aria-hidden="true" tabindex="-1"></a>    name<span class="op">=</span><span class="st">"Unknown MOA&lt;br&gt;annotations"</span>,</span>
<span id="cb190-55"><a href="#cb190-55" aria-hidden="true" tabindex="-1"></a>    marker<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">7</span>, color<span class="op">=</span><span class="st">"#36382E"</span>,symbol<span class="op">=</span><span class="st">'circle'</span>),</span>
<span id="cb190-56"><a href="#cb190-56" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb190-57"><a href="#cb190-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-58"><a href="#cb190-58" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb190-59"><a href="#cb190-59" aria-hidden="true" tabindex="-1"></a>layout <span class="op">=</span> go.Layout(</span>
<span id="cb190-60"><a href="#cb190-60" aria-hidden="true" tabindex="-1"></a>   title<span class="op">=</span><span class="bu">dict</span>(text<span class="op">=</span><span class="st">"&lt;b&gt;"</span><span class="op">+</span>selected_goterm<span class="op">+</span><span class="st">"&lt;/b&gt; "</span><span class="op">+</span>go_name,</span>
<span id="cb190-61"><a href="#cb190-61" aria-hidden="true" tabindex="-1"></a>              x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb190-62"><a href="#cb190-62" aria-hidden="true" tabindex="-1"></a>              y<span class="op">=</span><span class="fl">0.92</span>,</span>
<span id="cb190-63"><a href="#cb190-63" aria-hidden="true" tabindex="-1"></a>              font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">23</span>),</span>
<span id="cb190-64"><a href="#cb190-64" aria-hidden="true" tabindex="-1"></a>              xanchor<span class="op">=</span><span class="st">'center'</span>,</span>
<span id="cb190-65"><a href="#cb190-65" aria-hidden="true" tabindex="-1"></a>              yanchor<span class="op">=</span><span class="st">'top'</span>),</span>
<span id="cb190-66"><a href="#cb190-66" aria-hidden="true" tabindex="-1"></a>    xaxis<span class="op">=</span><span class="bu">dict</span>(ticks<span class="op">=</span><span class="st">""</span>, showticklabels<span class="op">=</span><span class="va">False</span>, showgrid<span class="op">=</span><span class="va">False</span>, zeroline<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb190-67"><a href="#cb190-67" aria-hidden="true" tabindex="-1"></a>    yaxis<span class="op">=</span><span class="bu">dict</span>(ticks<span class="op">=</span><span class="st">""</span>, showticklabels<span class="op">=</span><span class="va">False</span>, showgrid<span class="op">=</span><span class="va">False</span>, zeroline<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb190-68"><a href="#cb190-68" aria-hidden="true" tabindex="-1"></a>    yaxis_range<span class="op">=</span>[<span class="bu">min</span>(yy.flatten()),<span class="bu">max</span>(yy.flatten())],</span>
<span id="cb190-69"><a href="#cb190-69" aria-hidden="true" tabindex="-1"></a>    xaxis_range<span class="op">=</span>[<span class="bu">min</span>(xx.flatten()),<span class="bu">max</span>(xx.flatten())],</span>
<span id="cb190-70"><a href="#cb190-70" aria-hidden="true" tabindex="-1"></a>    legend<span class="op">=</span><span class="bu">dict</span>(x<span class="op">=</span><span class="dv">0</span>, y<span class="op">=</span><span class="dv">0</span>, orientation<span class="op">=</span><span class="st">"h"</span>,font<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">15</span>)),</span>
<span id="cb190-71"><a href="#cb190-71" aria-hidden="true" tabindex="-1"></a>    paper_bgcolor<span class="op">=</span><span class="st">'rgba(0,0,0,0)'</span>,</span>
<span id="cb190-72"><a href="#cb190-72" aria-hidden="true" tabindex="-1"></a>    width<span class="op">=</span><span class="dv">600</span>, height<span class="op">=</span><span class="dv">800</span>,</span>
<span id="cb190-73"><a href="#cb190-73" aria-hidden="true" tabindex="-1"></a>    font<span class="op">=</span><span class="bu">dict</span>(family<span class="op">=</span><span class="st">'Roboto'</span>,color<span class="op">=</span> <span class="st">"#36382E"</span>,size<span class="op">=</span><span class="dv">15</span>)</span>
<span id="cb190-74"><a href="#cb190-74" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb190-75"><a href="#cb190-75" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> [trace0,trace1,trace2,legend2,legend1,legend3]</span>
<span id="cb190-76"><a href="#cb190-76" aria-hidden="true" tabindex="-1"></a>figure <span class="op">=</span> go.Figure(data<span class="op">=</span>data,layout<span class="op">=</span>layout)</span>
<span id="cb190-77"><a href="#cb190-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-78"><a href="#cb190-78" aria-hidden="true" tabindex="-1"></a>figure</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Unable to display output for mime type(s): application/vnd.plotly.v1+json</code></pre>
</div>
</div>
<p>Add drugs</p>
<div class="cell" data-execution_count="166">
<div class="sourceCode cell-code" id="cb192"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb192-1"><a href="#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Spindle organization</span></span>
<span id="cb192-2"><a href="#cb192-2" aria-hidden="true" tabindex="-1"></a>figure <span class="op">=</span> go.Figure(data<span class="op">=</span>data,layout<span class="op">=</span>layout)</span>
<span id="cb192-3"><a href="#cb192-3" aria-hidden="true" tabindex="-1"></a>figure.add_annotation(</span>
<span id="cb192-4"><a href="#cb192-4" aria-hidden="true" tabindex="-1"></a>         x<span class="op">=</span>df[<span class="st">"comp-1"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"sb-743921"</span>].to_numpy()[<span class="dv">0</span>], y<span class="op">=</span>df[<span class="st">"comp-2"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"sb-743921"</span>].to_numpy()[<span class="dv">0</span>],</span>
<span id="cb192-5"><a href="#cb192-5" aria-hidden="true" tabindex="-1"></a>         xref<span class="op">=</span><span class="st">"x"</span>, yref<span class="op">=</span><span class="st">"y"</span>,</span>
<span id="cb192-6"><a href="#cb192-6" aria-hidden="true" tabindex="-1"></a>         text<span class="op">=</span><span class="st">"&lt;b&gt;"</span><span class="op">+</span><span class="st">'SB-743921'"&lt;b&gt;"</span>,</span>
<span id="cb192-7"><a href="#cb192-7" aria-hidden="true" tabindex="-1"></a>         showarrow<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb192-8"><a href="#cb192-8" aria-hidden="true" tabindex="-1"></a>         font<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb192-9"><a href="#cb192-9" aria-hidden="true" tabindex="-1"></a>            <span class="co"># family="Courier New, monospace",</span></span>
<span id="cb192-10"><a href="#cb192-10" aria-hidden="true" tabindex="-1"></a>             size<span class="op">=</span><span class="dv">19</span>,</span>
<span id="cb192-11"><a href="#cb192-11" aria-hidden="true" tabindex="-1"></a>             color<span class="op">=</span><span class="st">"#000000"</span></span>
<span id="cb192-12"><a href="#cb192-12" aria-hidden="true" tabindex="-1"></a>             ),</span>
<span id="cb192-13"><a href="#cb192-13" aria-hidden="true" tabindex="-1"></a>         align<span class="op">=</span><span class="st">"center"</span>,</span>
<span id="cb192-14"><a href="#cb192-14" aria-hidden="true" tabindex="-1"></a>         arrowhead<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-15"><a href="#cb192-15" aria-hidden="true" tabindex="-1"></a>         arrowsize<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-16"><a href="#cb192-16" aria-hidden="true" tabindex="-1"></a>         arrowwidth<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-17"><a href="#cb192-17" aria-hidden="true" tabindex="-1"></a>         arrowcolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-18"><a href="#cb192-18" aria-hidden="true" tabindex="-1"></a>         ax<span class="op">=-</span><span class="dv">250</span>,</span>
<span id="cb192-19"><a href="#cb192-19" aria-hidden="true" tabindex="-1"></a>         ay<span class="op">=-</span><span class="dv">10</span>,</span>
<span id="cb192-20"><a href="#cb192-20" aria-hidden="true" tabindex="-1"></a>         bordercolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-21"><a href="#cb192-21" aria-hidden="true" tabindex="-1"></a>         borderwidth<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-22"><a href="#cb192-22" aria-hidden="true" tabindex="-1"></a>         borderpad<span class="op">=</span><span class="dv">4</span>,</span>
<span id="cb192-23"><a href="#cb192-23" aria-hidden="true" tabindex="-1"></a>         bgcolor<span class="op">=</span><span class="st">"#ffffff"</span>,</span>
<span id="cb192-24"><a href="#cb192-24" aria-hidden="true" tabindex="-1"></a>         opacity<span class="op">=</span><span class="fl">0.8</span></span>
<span id="cb192-25"><a href="#cb192-25" aria-hidden="true" tabindex="-1"></a>         )</span>
<span id="cb192-26"><a href="#cb192-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-27"><a href="#cb192-27" aria-hidden="true" tabindex="-1"></a>figure.add_annotation(</span>
<span id="cb192-28"><a href="#cb192-28" aria-hidden="true" tabindex="-1"></a>         x<span class="op">=</span>df[<span class="st">"comp-1"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"thapsigargin"</span>].to_numpy()[<span class="dv">0</span>], y<span class="op">=</span>df[<span class="st">"comp-2"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"thapsigargin"</span>].to_numpy()[<span class="dv">0</span>],</span>
<span id="cb192-29"><a href="#cb192-29" aria-hidden="true" tabindex="-1"></a>         xref<span class="op">=</span><span class="st">"x"</span>, yref<span class="op">=</span><span class="st">"y"</span>,</span>
<span id="cb192-30"><a href="#cb192-30" aria-hidden="true" tabindex="-1"></a>         text<span class="op">=</span><span class="st">"&lt;b&gt;"</span><span class="op">+</span><span class="st">'Thapsigargin'"&lt;b&gt;"</span>,</span>
<span id="cb192-31"><a href="#cb192-31" aria-hidden="true" tabindex="-1"></a>         showarrow<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb192-32"><a href="#cb192-32" aria-hidden="true" tabindex="-1"></a>         font<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb192-33"><a href="#cb192-33" aria-hidden="true" tabindex="-1"></a>            <span class="co"># family="Courier New, monospace",</span></span>
<span id="cb192-34"><a href="#cb192-34" aria-hidden="true" tabindex="-1"></a>             size<span class="op">=</span><span class="dv">19</span>,</span>
<span id="cb192-35"><a href="#cb192-35" aria-hidden="true" tabindex="-1"></a>             color<span class="op">=</span><span class="st">"#000000"</span></span>
<span id="cb192-36"><a href="#cb192-36" aria-hidden="true" tabindex="-1"></a>             ),</span>
<span id="cb192-37"><a href="#cb192-37" aria-hidden="true" tabindex="-1"></a>         align<span class="op">=</span><span class="st">"center"</span>,</span>
<span id="cb192-38"><a href="#cb192-38" aria-hidden="true" tabindex="-1"></a>         arrowhead<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-39"><a href="#cb192-39" aria-hidden="true" tabindex="-1"></a>         arrowsize<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-40"><a href="#cb192-40" aria-hidden="true" tabindex="-1"></a>         arrowwidth<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-41"><a href="#cb192-41" aria-hidden="true" tabindex="-1"></a>         arrowcolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-42"><a href="#cb192-42" aria-hidden="true" tabindex="-1"></a>         ax<span class="op">=-</span><span class="dv">60</span>,</span>
<span id="cb192-43"><a href="#cb192-43" aria-hidden="true" tabindex="-1"></a>         ay<span class="op">=-</span><span class="dv">90</span>,</span>
<span id="cb192-44"><a href="#cb192-44" aria-hidden="true" tabindex="-1"></a>         bordercolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-45"><a href="#cb192-45" aria-hidden="true" tabindex="-1"></a>         borderwidth<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-46"><a href="#cb192-46" aria-hidden="true" tabindex="-1"></a>         borderpad<span class="op">=</span><span class="dv">4</span>,</span>
<span id="cb192-47"><a href="#cb192-47" aria-hidden="true" tabindex="-1"></a>         bgcolor<span class="op">=</span><span class="st">"#ffffff"</span>,</span>
<span id="cb192-48"><a href="#cb192-48" aria-hidden="true" tabindex="-1"></a>         opacity<span class="op">=</span><span class="fl">0.8</span></span>
<span id="cb192-49"><a href="#cb192-49" aria-hidden="true" tabindex="-1"></a>         )</span>
<span id="cb192-50"><a href="#cb192-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-51"><a href="#cb192-51" aria-hidden="true" tabindex="-1"></a>figure.add_annotation(</span>
<span id="cb192-52"><a href="#cb192-52" aria-hidden="true" tabindex="-1"></a>         x<span class="op">=</span>df[<span class="st">"comp-1"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"triptolide"</span>].to_numpy()[<span class="dv">0</span>], y<span class="op">=</span>df[<span class="st">"comp-2"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"triptolide"</span>].to_numpy()[<span class="dv">0</span>],</span>
<span id="cb192-53"><a href="#cb192-53" aria-hidden="true" tabindex="-1"></a>         xref<span class="op">=</span><span class="st">"x"</span>, yref<span class="op">=</span><span class="st">"y"</span>,</span>
<span id="cb192-54"><a href="#cb192-54" aria-hidden="true" tabindex="-1"></a>         text<span class="op">=</span><span class="st">"&lt;b&gt;"</span><span class="op">+</span><span class="st">'Triptolide'"&lt;b&gt;"</span>,</span>
<span id="cb192-55"><a href="#cb192-55" aria-hidden="true" tabindex="-1"></a>         showarrow<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb192-56"><a href="#cb192-56" aria-hidden="true" tabindex="-1"></a>         font<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb192-57"><a href="#cb192-57" aria-hidden="true" tabindex="-1"></a>            <span class="co"># family="Courier New, monospace",</span></span>
<span id="cb192-58"><a href="#cb192-58" aria-hidden="true" tabindex="-1"></a>             size<span class="op">=</span><span class="dv">19</span>,</span>
<span id="cb192-59"><a href="#cb192-59" aria-hidden="true" tabindex="-1"></a>             color<span class="op">=</span><span class="st">"#000000"</span></span>
<span id="cb192-60"><a href="#cb192-60" aria-hidden="true" tabindex="-1"></a>             ),</span>
<span id="cb192-61"><a href="#cb192-61" aria-hidden="true" tabindex="-1"></a>         align<span class="op">=</span><span class="st">"center"</span>,</span>
<span id="cb192-62"><a href="#cb192-62" aria-hidden="true" tabindex="-1"></a>         arrowhead<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-63"><a href="#cb192-63" aria-hidden="true" tabindex="-1"></a>         arrowsize<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-64"><a href="#cb192-64" aria-hidden="true" tabindex="-1"></a>         arrowwidth<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-65"><a href="#cb192-65" aria-hidden="true" tabindex="-1"></a>         arrowcolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-66"><a href="#cb192-66" aria-hidden="true" tabindex="-1"></a>         ax<span class="op">=-</span><span class="dv">130</span>,</span>
<span id="cb192-67"><a href="#cb192-67" aria-hidden="true" tabindex="-1"></a>         ay<span class="op">=-</span><span class="dv">80</span>,</span>
<span id="cb192-68"><a href="#cb192-68" aria-hidden="true" tabindex="-1"></a>         bordercolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-69"><a href="#cb192-69" aria-hidden="true" tabindex="-1"></a>         borderwidth<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-70"><a href="#cb192-70" aria-hidden="true" tabindex="-1"></a>         borderpad<span class="op">=</span><span class="dv">4</span>,</span>
<span id="cb192-71"><a href="#cb192-71" aria-hidden="true" tabindex="-1"></a>         bgcolor<span class="op">=</span><span class="st">"#ffffff"</span>,</span>
<span id="cb192-72"><a href="#cb192-72" aria-hidden="true" tabindex="-1"></a>         opacity<span class="op">=</span><span class="fl">0.8</span></span>
<span id="cb192-73"><a href="#cb192-73" aria-hidden="true" tabindex="-1"></a>         )</span>
<span id="cb192-74"><a href="#cb192-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-75"><a href="#cb192-75" aria-hidden="true" tabindex="-1"></a>figure.add_annotation(</span>
<span id="cb192-76"><a href="#cb192-76" aria-hidden="true" tabindex="-1"></a>         x<span class="op">=</span>df[<span class="st">"comp-1"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"gw843682x"</span>].to_numpy()[<span class="dv">0</span>], y<span class="op">=</span>df[<span class="st">"comp-2"</span>].loc[df[<span class="st">"name"</span>]<span class="op">==</span><span class="st">"gw843682x"</span>].to_numpy()[<span class="dv">0</span>],</span>
<span id="cb192-77"><a href="#cb192-77" aria-hidden="true" tabindex="-1"></a>         xref<span class="op">=</span><span class="st">"x"</span>, yref<span class="op">=</span><span class="st">"y"</span>,</span>
<span id="cb192-78"><a href="#cb192-78" aria-hidden="true" tabindex="-1"></a>         text<span class="op">=</span><span class="st">"&lt;b&gt;"</span><span class="op">+</span><span class="st">'GW843682X'"&lt;b&gt;"</span>,</span>
<span id="cb192-79"><a href="#cb192-79" aria-hidden="true" tabindex="-1"></a>         showarrow<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb192-80"><a href="#cb192-80" aria-hidden="true" tabindex="-1"></a>         font<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb192-81"><a href="#cb192-81" aria-hidden="true" tabindex="-1"></a>            <span class="co"># family="Courier New, monospace",</span></span>
<span id="cb192-82"><a href="#cb192-82" aria-hidden="true" tabindex="-1"></a>             size<span class="op">=</span><span class="dv">19</span>,</span>
<span id="cb192-83"><a href="#cb192-83" aria-hidden="true" tabindex="-1"></a>             color<span class="op">=</span><span class="st">"#000000"</span></span>
<span id="cb192-84"><a href="#cb192-84" aria-hidden="true" tabindex="-1"></a>             ),</span>
<span id="cb192-85"><a href="#cb192-85" aria-hidden="true" tabindex="-1"></a>         align<span class="op">=</span><span class="st">"center"</span>,</span>
<span id="cb192-86"><a href="#cb192-86" aria-hidden="true" tabindex="-1"></a>         arrowhead<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-87"><a href="#cb192-87" aria-hidden="true" tabindex="-1"></a>         arrowsize<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-88"><a href="#cb192-88" aria-hidden="true" tabindex="-1"></a>         arrowwidth<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb192-89"><a href="#cb192-89" aria-hidden="true" tabindex="-1"></a>         arrowcolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-90"><a href="#cb192-90" aria-hidden="true" tabindex="-1"></a>         ax<span class="op">=-</span><span class="dv">110</span>,</span>
<span id="cb192-91"><a href="#cb192-91" aria-hidden="true" tabindex="-1"></a>         ay<span class="op">=</span><span class="dv">70</span>,</span>
<span id="cb192-92"><a href="#cb192-92" aria-hidden="true" tabindex="-1"></a>         bordercolor<span class="op">=</span><span class="st">"#636363"</span>,</span>
<span id="cb192-93"><a href="#cb192-93" aria-hidden="true" tabindex="-1"></a>         borderwidth<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb192-94"><a href="#cb192-94" aria-hidden="true" tabindex="-1"></a>         borderpad<span class="op">=</span><span class="dv">4</span>,</span>
<span id="cb192-95"><a href="#cb192-95" aria-hidden="true" tabindex="-1"></a>         bgcolor<span class="op">=</span><span class="st">"#ffffff"</span>,</span>
<span id="cb192-96"><a href="#cb192-96" aria-hidden="true" tabindex="-1"></a>         opacity<span class="op">=</span><span class="fl">0.8</span></span>
<span id="cb192-97"><a href="#cb192-97" aria-hidden="true" tabindex="-1"></a>         )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Unable to display output for mime type(s): application/vnd.plotly.v1+json</code></pre>
</div>
</div>
</section>
</section>
</section>
</section>
</section>
<section id="top-genes" class="level1">
<h1>Top Genes</h1>
<div class="cell" data-execution_count="167">
<div class="sourceCode cell-code" id="cb194"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb194-1"><a href="#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> captum.attr <span class="im">import</span> LayerDeepLift, DeepLift</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="168">
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_compound_names(file_name):</span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a>    compounds <span class="op">=</span> []</span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb195-4"><a href="#cb195-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(file_name, <span class="st">'r'</span>) <span class="im">as</span> fi:</span>
<span id="cb195-5"><a href="#cb195-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> fi:</span>
<span id="cb195-6"><a href="#cb195-6" aria-hidden="true" tabindex="-1"></a>            tokens <span class="op">=</span> line.strip().split(<span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb195-7"><a href="#cb195-7" aria-hidden="true" tabindex="-1"></a>            compounds.append([tokens[<span class="dv">1</span>],tokens[<span class="dv">2</span>]])</span>
<span id="cb195-8"><a href="#cb195-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> compounds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="169">
<div class="sourceCode cell-code" id="cb196"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb196-1"><a href="#cb196-1" aria-hidden="true" tabindex="-1"></a>drugs <span class="op">=</span> get_compound_names(inputdir<span class="op">+</span><span class="st">"compound_names.txt"</span>)</span>
<span id="cb196-2"><a href="#cb196-2" aria-hidden="true" tabindex="-1"></a>drugs.pop(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="169">
<pre><code>['SMILE', 'Name']</code></pre>
</div>
</div>
<div class="cell" data-execution_count="170">
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> f(drug):    </span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> drug</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="171">
<div class="sourceCode cell-code" id="cb199"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb199-1"><a href="#cb199-1" aria-hidden="true" tabindex="-1"></a>num_neurons_per_GO <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb199-2"><a href="#cb199-2" aria-hidden="true" tabindex="-1"></a>median_cell_features <span class="op">=</span> np.median(cell_features,axis<span class="op">=</span><span class="dv">0</span>) <span class="co"># to use as a reference</span></span>
<span id="cb199-3"><a href="#cb199-3" aria-hidden="true" tabindex="-1"></a>median_drug_features <span class="op">=</span> np.genfromtxt(computer<span class="op">+</span><span class="st">"SparseGO_code/data/glucose_fingerprint.txt"</span>, delimiter<span class="op">=</span><span class="st">','</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="172">
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_layer_attribution_genes(input_data,baseline,selected_drug_data):</span>
<span id="cb200-2"><a href="#cb200-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">#dl = DeepLift(model,  multiply_by_inputs=True)  # CHOOSE LAYER TO STUDY</span></span>
<span id="cb200-3"><a href="#cb200-3" aria-hidden="true" tabindex="-1"></a>    dl <span class="op">=</span> DeepLift(model) </span>
<span id="cb200-4"><a href="#cb200-4" aria-hidden="true" tabindex="-1"></a>    dl_attr_test <span class="op">=</span> dl.attribute(input_data,baseline)</span>
<span id="cb200-5"><a href="#cb200-5" aria-hidden="true" tabindex="-1"></a>    dl_attr_test_sum <span class="op">=</span> dl_attr_test.cpu().detach().numpy().<span class="bu">sum</span>(<span class="dv">0</span>) <span class="co"># se suman las attributions para cada sample</span></span>
<span id="cb200-6"><a href="#cb200-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># dl_attr_test_norm_sum = dl_attr_test_sum / np.linalg.norm(dl_attr_test_sum, ord=1) # se normaliza </span></span>
<span id="cb200-7"><a href="#cb200-7" aria-hidden="true" tabindex="-1"></a>    dl_attr_test_norm_sum <span class="op">=</span> dl_attr_test_sum </span>
<span id="cb200-8"><a href="#cb200-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb200-9"><a href="#cb200-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># gene2id_mapping WAS USED TO CREATE THE FIRST LAYER OF THE NETWORK, SO THAT INPUT DATA MATCHES THE FIRST LAYER </span></span>
<span id="cb200-10"><a href="#cb200-10" aria-hidden="true" tabindex="-1"></a>    attribution_data <span class="op">=</span> pd.DataFrame(np.column_stack((np.array(<span class="bu">list</span>(gene2id_mapping.keys())),dl_attr_test_norm_sum[<span class="dv">0</span>:<span class="bu">len</span>(gene2id_mapping.keys())])), columns<span class="op">=</span>[<span class="st">"Gene"</span>,selected_drug_data[<span class="dv">1</span>]])</span>
<span id="cb200-11"><a href="#cb200-11" aria-hidden="true" tabindex="-1"></a>    attribution_data[[selected_drug_data[<span class="dv">1</span>]]] <span class="op">=</span> attribution_data[[selected_drug_data[<span class="dv">1</span>]]].<span class="bu">apply</span>(pd.to_numeric).<span class="bu">round</span>(<span class="dv">8</span>)</span>
<span id="cb200-12"><a href="#cb200-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> attribution_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb201"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb201-1"><a href="#cb201-1" aria-hidden="true" tabindex="-1"></a>first <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb201-2"><a href="#cb201-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain the top GO terms on all layers for each drug</span></span>
<span id="cb201-3"><a href="#cb201-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> selected_drug_data <span class="kw">in</span> drugs:</span>
<span id="cb201-4"><a href="#cb201-4" aria-hidden="true" tabindex="-1"></a>    selected_drug <span class="op">=</span>selected_drug_data[<span class="dv">0</span>] <span class="co"># DRUG smile</span></span>
<span id="cb201-5"><a href="#cb201-5" aria-hidden="true" tabindex="-1"></a>    selected_drug_features <span class="op">=</span> []</span>
<span id="cb201-6"><a href="#cb201-6" aria-hidden="true" tabindex="-1"></a>    drug_specific_features<span class="op">=</span>drug_features[drug2id_mapping[selected_drug]] <span class="co"># features of drug</span></span>
<span id="cb201-7"><a href="#cb201-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(cell2id_mapping)): <span class="co"># make all combinations of selected_drug and cell types </span></span>
<span id="cb201-8"><a href="#cb201-8" aria-hidden="true" tabindex="-1"></a>        selected_drug_features.append(np.concatenate((cell_features[i], drug_specific_features), axis<span class="op">=</span><span class="va">None</span>))</span>
<span id="cb201-9"><a href="#cb201-9" aria-hidden="true" tabindex="-1"></a>    selected_drug_features <span class="op">=</span> torch.FloatTensor(np.array(selected_drug_features))</span>
<span id="cb201-10"><a href="#cb201-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb201-11"><a href="#cb201-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Data for deeplift</span></span>
<span id="cb201-12"><a href="#cb201-12" aria-hidden="true" tabindex="-1"></a>    input_data <span class="op">=</span> torch.autograd.Variable(selected_drug_features.cuda(<span class="dv">0</span>))</span>
<span id="cb201-13"><a href="#cb201-13" aria-hidden="true" tabindex="-1"></a>    baseline <span class="op">=</span> torch.FloatTensor(np.concatenate((median_cell_features, median_drug_features), axis<span class="op">=</span><span class="va">None</span>))</span>
<span id="cb201-14"><a href="#cb201-14" aria-hidden="true" tabindex="-1"></a>    baseline <span class="op">=</span> torch.reshape(baseline, (<span class="dv">1</span>, baseline.size()[<span class="dv">0</span>]))</span>
<span id="cb201-15"><a href="#cb201-15" aria-hidden="true" tabindex="-1"></a>    baseline <span class="op">=</span> torch.autograd.Variable(baseline.cuda(<span class="dv">0</span>))</span>
<span id="cb201-16"><a href="#cb201-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb201-17"><a href="#cb201-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb201-18"><a href="#cb201-18" aria-hidden="true" tabindex="-1"></a>    attribution_data_drug <span class="op">=</span> get_layer_attribution_genes(input_data,baseline,selected_drug_data)</span>
<span id="cb201-19"><a href="#cb201-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># top_scores = attribution_data_drug.nlargest(5, selected_drug_data[1]) # PROBAR TMB CON ABS VALUE</span></span>
<span id="cb201-20"><a href="#cb201-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># top_genes.loc[selected_drug_data[1]] = top_scores['Gene'].values # PERO IGUAL HABRIA QUE ELIMINAR ANTES LOS QUE TIENEN ALTO SCORE EN TODOS</span></span>
<span id="cb201-21"><a href="#cb201-21" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb201-22"><a href="#cb201-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> first<span class="op">==</span><span class="dv">1</span>:</span>
<span id="cb201-23"><a href="#cb201-23" aria-hidden="true" tabindex="-1"></a>        attribution_data_all_genes <span class="op">=</span> attribution_data_drug</span>
<span id="cb201-24"><a href="#cb201-24" aria-hidden="true" tabindex="-1"></a>        first<span class="op">=</span><span class="dv">0</span></span>
<span id="cb201-25"><a href="#cb201-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb201-26"><a href="#cb201-26" aria-hidden="true" tabindex="-1"></a>        attribution_data_all_genes <span class="op">=</span> pd.concat([attribution_data_all_genes,attribution_data_drug.iloc[:,<span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb201-27"><a href="#cb201-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb201-28"><a href="#cb201-28" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(selected_drug_data[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb202"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb202-1"><a href="#cb202-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all_genes <span class="op">=</span> attribution_data_all_genes.set_index(<span class="st">"Gene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="176">
<div class="sourceCode cell-code" id="cb203"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb203-1"><a href="#cb203-1" aria-hidden="true" tabindex="-1"></a>top_var <span class="op">=</span> attribution_data_all_genes.var(axis<span class="op">=</span><span class="dv">1</span>).sort_values(axis<span class="op">=</span><span class="dv">0</span>, ascending<span class="op">=</span><span class="va">False</span>)[<span class="dv">0</span>:<span class="dv">1000000</span>] <span class="co"># Almost no change</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="177">
<div class="sourceCode cell-code" id="cb204"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb204-1"><a href="#cb204-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all_genes_var <span class="op">=</span> attribution_data_all_genes.loc[top_var.index.values]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="178">
<div class="sourceCode cell-code" id="cb205"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb205-1"><a href="#cb205-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all_genes_var.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="178">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>(-)-Parthenolide</th>
      <th>(5Z)-7-Oxozeaenol</th>
      <th>5-Fluorouracil</th>
      <th>681640</th>
      <th>A-443654</th>
      <th>A-484954</th>
      <th>A-770041</th>
      <th>A-83-01</th>
      <th>Acadesine</th>
      <th>ACY-1215</th>
      <th>...</th>
      <th>Vemurafenib</th>
      <th>VER-155008</th>
      <th>Vorapaxar</th>
      <th>vorinostat:carboplatin (1:1 mol/mol)</th>
      <th>vorinostat:navitoclax (4:1 mol/mol)</th>
      <th>VU0155056</th>
      <th>WZ4002</th>
      <th>WZ8040</th>
      <th>YL54</th>
      <th>Zebularine</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TNF</th>
      <td>-0.647627</td>
      <td>-1.171123</td>
      <td>-0.798656</td>
      <td>-1.056030</td>
      <td>-1.482396</td>
      <td>-0.351407</td>
      <td>-1.223713</td>
      <td>-0.432469</td>
      <td>-0.355250</td>
      <td>-1.175542</td>
      <td>...</td>
      <td>-0.150800</td>
      <td>-0.713235</td>
      <td>-0.225855</td>
      <td>-0.834043</td>
      <td>-1.694794</td>
      <td>-0.321265</td>
      <td>-0.451579</td>
      <td>-0.623499</td>
      <td>-0.315067</td>
      <td>-0.188699</td>
    </tr>
    <tr>
      <th>TLR4</th>
      <td>0.424753</td>
      <td>0.587436</td>
      <td>0.552789</td>
      <td>0.811383</td>
      <td>1.290940</td>
      <td>0.170548</td>
      <td>0.649595</td>
      <td>0.153363</td>
      <td>0.155758</td>
      <td>0.671745</td>
      <td>...</td>
      <td>0.261201</td>
      <td>0.585952</td>
      <td>0.462414</td>
      <td>0.527991</td>
      <td>0.756867</td>
      <td>0.326152</td>
      <td>0.669734</td>
      <td>0.668688</td>
      <td>0.348900</td>
      <td>0.281724</td>
    </tr>
    <tr>
      <th>CX3CL1</th>
      <td>0.367396</td>
      <td>0.850156</td>
      <td>0.397639</td>
      <td>0.463275</td>
      <td>1.145052</td>
      <td>0.093489</td>
      <td>0.278140</td>
      <td>0.087945</td>
      <td>0.085774</td>
      <td>0.503575</td>
      <td>...</td>
      <td>-0.033222</td>
      <td>0.333330</td>
      <td>0.384190</td>
      <td>0.375685</td>
      <td>0.785719</td>
      <td>0.193030</td>
      <td>0.303128</td>
      <td>0.316458</td>
      <td>0.240599</td>
      <td>0.139607</td>
    </tr>
    <tr>
      <th>NCKAP1L</th>
      <td>-0.209777</td>
      <td>-0.407460</td>
      <td>-0.355179</td>
      <td>-0.555746</td>
      <td>-0.655136</td>
      <td>-0.109196</td>
      <td>-0.420126</td>
      <td>-0.102194</td>
      <td>-0.131356</td>
      <td>-0.715699</td>
      <td>...</td>
      <td>-0.008355</td>
      <td>-0.403133</td>
      <td>-0.178399</td>
      <td>-0.604647</td>
      <td>-0.783917</td>
      <td>-0.156935</td>
      <td>-0.544116</td>
      <td>-0.704079</td>
      <td>-0.086063</td>
      <td>-0.150185</td>
    </tr>
    <tr>
      <th>VAV1</th>
      <td>-0.113726</td>
      <td>-0.504159</td>
      <td>-0.213520</td>
      <td>-0.656914</td>
      <td>-0.371759</td>
      <td>-0.107983</td>
      <td>-0.040083</td>
      <td>-0.144649</td>
      <td>-0.099499</td>
      <td>-0.107369</td>
      <td>...</td>
      <td>-0.077634</td>
      <td>-0.122759</td>
      <td>-0.110199</td>
      <td>-0.115432</td>
      <td>0.150526</td>
      <td>-0.045423</td>
      <td>0.037265</td>
      <td>-0.052557</td>
      <td>-0.074338</td>
      <td>-0.039058</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 790 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb206"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a>attribution_data_all_genes_var.to_excel(resultsdir<span class="op">+</span><span class="st">'attribution_data_all_genes.xlsx'</span>, index<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="179">
<div class="sourceCode cell-code" id="cb207"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb207-1"><a href="#cb207-1" aria-hidden="true" tabindex="-1"></a>top_genes <span class="op">=</span> {}</span>
<span id="cb207-2"><a href="#cb207-2" aria-hidden="true" tabindex="-1"></a>num_genes <span class="op">=</span> <span class="dv">50</span></span>
<span id="cb207-3"><a href="#cb207-3" aria-hidden="true" tabindex="-1"></a><span class="co"># loop through each column in the dataframe</span></span>
<span id="cb207-4"><a href="#cb207-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> col <span class="kw">in</span> attribution_data_all_genes_var.columns:</span>
<span id="cb207-5"><a href="#cb207-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb207-6"><a href="#cb207-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get the top n scores and their index (i.e., row name)</span></span>
<span id="cb207-7"><a href="#cb207-7" aria-hidden="true" tabindex="-1"></a>    top_scores <span class="op">=</span> <span class="bu">abs</span>(attribution_data_all_genes_var[col]).nlargest(num_genes)</span>
<span id="cb207-8"><a href="#cb207-8" aria-hidden="true" tabindex="-1"></a>    top_scores_idx <span class="op">=</span> top_scores.index</span>
<span id="cb207-9"><a href="#cb207-9" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb207-10"><a href="#cb207-10" aria-hidden="true" tabindex="-1"></a>    top_genes[col] <span class="op">=</span> top_scores_idx.values.tolist()</span>
<span id="cb207-11"><a href="#cb207-11" aria-hidden="true" tabindex="-1"></a>        <span class="co"># 'scores': top_scores.values.tolist(),</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="180">
<div class="sourceCode cell-code" id="cb208"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb208-1"><a href="#cb208-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create DataFrame from dictionary</span></span>
<span id="cb208-2"><a href="#cb208-2" aria-hidden="true" tabindex="-1"></a>df_top_genes <span class="op">=</span> pd.DataFrame.from_dict(top_genes, orient<span class="op">=</span><span class="st">'index'</span>)</span>
<span id="cb208-3"><a href="#cb208-3" aria-hidden="true" tabindex="-1"></a>df_top_genes.columns  <span class="op">=</span> [<span class="st">'Gene</span><span class="sc">{}</span><span class="st">'</span>.<span class="bu">format</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, num_genes<span class="op">+</span><span class="dv">1</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="181">
<div class="sourceCode cell-code" id="cb209"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb209-1"><a href="#cb209-1" aria-hidden="true" tabindex="-1"></a>df_top_genes.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="181">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>Gene1</th>
      <th>Gene2</th>
      <th>Gene3</th>
      <th>Gene4</th>
      <th>Gene5</th>
      <th>Gene6</th>
      <th>Gene7</th>
      <th>Gene8</th>
      <th>Gene9</th>
      <th>Gene10</th>
      <th>...</th>
      <th>Gene41</th>
      <th>Gene42</th>
      <th>Gene43</th>
      <th>Gene44</th>
      <th>Gene45</th>
      <th>Gene46</th>
      <th>Gene47</th>
      <th>Gene48</th>
      <th>Gene49</th>
      <th>Gene50</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>(-)-Parthenolide</th>
      <td>TNF</td>
      <td>TLR4</td>
      <td>CX3CL1</td>
      <td>ROR2</td>
      <td>TBX1</td>
      <td>WNT4</td>
      <td>DRAXIN</td>
      <td>BCL2</td>
      <td>FOXB1</td>
      <td>NKX2-5</td>
      <td>...</td>
      <td>HEY1</td>
      <td>IL33</td>
      <td>NCKAP1L</td>
      <td>SLC7A3</td>
      <td>NOTCH4</td>
      <td>CCR7</td>
      <td>IL24</td>
      <td>ADAMTS20</td>
      <td>RFLNA</td>
      <td>ADORA1</td>
    </tr>
    <tr>
      <th>(5Z)-7-Oxozeaenol</th>
      <td>TNF</td>
      <td>PTPRO</td>
      <td>SFRP1</td>
      <td>KCNE1</td>
      <td>CX3CL1</td>
      <td>SLC38A5</td>
      <td>TBX1</td>
      <td>SLCO2B1</td>
      <td>RIPOR2</td>
      <td>SFRP2</td>
      <td>...</td>
      <td>CAPN3</td>
      <td>VAV1</td>
      <td>GBP5</td>
      <td>ABCB1</td>
      <td>CLDN5</td>
      <td>AIM2</td>
      <td>ADAMTS20</td>
      <td>TSPO2</td>
      <td>RFLNA</td>
      <td>IL2RG</td>
    </tr>
    <tr>
      <th>5-Fluorouracil</th>
      <td>TNF</td>
      <td>TLR4</td>
      <td>SLC38A5</td>
      <td>TBX1</td>
      <td>ROR2</td>
      <td>WNT4</td>
      <td>CX3CL1</td>
      <td>DRAXIN</td>
      <td>FOXB1</td>
      <td>ANGPT1</td>
      <td>...</td>
      <td>PTPRO</td>
      <td>CCL26</td>
      <td>SFRP1</td>
      <td>WNT9B</td>
      <td>TLR2</td>
      <td>KCNB1</td>
      <td>PAK3</td>
      <td>INHBB</td>
      <td>CCL5</td>
      <td>ZIC1</td>
    </tr>
    <tr>
      <th>681640</th>
      <td>TNF</td>
      <td>TLR4</td>
      <td>VAV1</td>
      <td>PDGFRA</td>
      <td>BCL2</td>
      <td>LAPTM5</td>
      <td>TBX1</td>
      <td>IL2RG</td>
      <td>ANGPT1</td>
      <td>YAP1</td>
      <td>...</td>
      <td>CR1</td>
      <td>SLCO2B1</td>
      <td>PF4</td>
      <td>WNT9B</td>
      <td>AGTR1</td>
      <td>IL23A</td>
      <td>ADA2</td>
      <td>SLFN11</td>
      <td>JPH4</td>
      <td>WNT4</td>
    </tr>
    <tr>
      <th>A-443654</th>
      <td>TNF</td>
      <td>TLR4</td>
      <td>TBX1</td>
      <td>CX3CL1</td>
      <td>DRAXIN</td>
      <td>WNT4</td>
      <td>FOXB1</td>
      <td>PTPRO</td>
      <td>SLC38A5</td>
      <td>SFRP1</td>
      <td>...</td>
      <td>GHSR</td>
      <td>HES5</td>
      <td>IL2RG</td>
      <td>ADD2</td>
      <td>CLEC7A</td>
      <td>RFLNA</td>
      <td>SLCO2B1</td>
      <td>ITGB3</td>
      <td>FGF9</td>
      <td>CCL26</td>
    </tr>
    <tr>
      <th>A-484954</th>
      <td>TNF</td>
      <td>TLR4</td>
      <td>SLC38A5</td>
      <td>TBX1</td>
      <td>DRAXIN</td>
      <td>SFRP1</td>
      <td>WNT4</td>
      <td>LDHC</td>
      <td>PIK3CG</td>
      <td>FOXB1</td>
      <td>...</td>
      <td>ANXA1</td>
      <td>AEBP1</td>
      <td>ERBB2</td>
      <td>IL33</td>
      <td>LAPTM5</td>
      <td>NOD2</td>
      <td>HAMP</td>
      <td>FUT7</td>
      <td>TLR2</td>
      <td>SLAMF8</td>
    </tr>
    <tr>
      <th>A-770041</th>
      <td>TNF</td>
      <td>SLCO2B1</td>
      <td>TLR4</td>
      <td>PCSK9</td>
      <td>WNT4</td>
      <td>WNT3A</td>
      <td>PTGS2</td>
      <td>GJA1</td>
      <td>SLC38A5</td>
      <td>SYT12</td>
      <td>...</td>
      <td>GBP5</td>
      <td>ROBO2</td>
      <td>ACP5</td>
      <td>SEMA3D</td>
      <td>TNFRSF1B</td>
      <td>MGARP</td>
      <td>SERPINE1</td>
      <td>IFNB1</td>
      <td>FGF9</td>
      <td>NR1H4</td>
    </tr>
    <tr>
      <th>A-83-01</th>
      <td>TNF</td>
      <td>BTK</td>
      <td>TBX1</td>
      <td>SLC38A5</td>
      <td>WNT4</td>
      <td>CCL5</td>
      <td>SPI1</td>
      <td>FOXJ1</td>
      <td>DRAXIN</td>
      <td>TLR4</td>
      <td>...</td>
      <td>ADRA2A</td>
      <td>PCSK9</td>
      <td>DRD2</td>
      <td>SLAMF8</td>
      <td>IL12RB1</td>
      <td>CCR2</td>
      <td>CD300A</td>
      <td>PTPRO</td>
      <td>CD80</td>
      <td>NTRK2</td>
    </tr>
    <tr>
      <th>Acadesine</th>
      <td>TNF</td>
      <td>TBX1</td>
      <td>TLR4</td>
      <td>SLC38A5</td>
      <td>WNT4</td>
      <td>DRAXIN</td>
      <td>PIK3CG</td>
      <td>LDHC</td>
      <td>SFRP1</td>
      <td>NCKAP1L</td>
      <td>...</td>
      <td>ERBB2</td>
      <td>HAMP</td>
      <td>INHBB</td>
      <td>SLAMF8</td>
      <td>TWIST2</td>
      <td>TNFSF18</td>
      <td>SOX8</td>
      <td>CHRNB2</td>
      <td>CX3CL1</td>
      <td>ITGB7</td>
    </tr>
    <tr>
      <th>ACY-1215</th>
      <td>TNF</td>
      <td>NCKAP1L</td>
      <td>TLR4</td>
      <td>DRAXIN</td>
      <td>TBX1</td>
      <td>PTPRO</td>
      <td>IL2RG</td>
      <td>SFRP1</td>
      <td>CCR7</td>
      <td>YAP1</td>
      <td>...</td>
      <td>LHX3</td>
      <td>GPR37L1</td>
      <td>PAK3</td>
      <td>CHRNB2</td>
      <td>BHMG1</td>
      <td>IL4</td>
      <td>CRTAM</td>
      <td>SOCS1</td>
      <td>LAPTM5</td>
      <td>SMOC2</td>
    </tr>
  </tbody>
</table>
<p>10 rows × 50 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb210"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb210-1"><a href="#cb210-1" aria-hidden="true" tabindex="-1"></a>df_top_genes.to_excel(computer<span class="op">+</span><span class="st">'SparseGO_FirstPaper/First_Revision_eBioMedicine/Supplementary/top_genes_exp.xlsx'</span>, index<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="for-a-specific-drug" class="level3">
<h3 class="anchored" data-anchor-id="for-a-specific-drug">For a specific drug</h3>
<div class="cell" data-execution_count="174">
<div class="sourceCode cell-code" id="cb211"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb211-1"><a href="#cb211-1" aria-hidden="true" tabindex="-1"></a>combobox <span class="op">=</span> interactive(f, drug<span class="op">=</span>widgets.Combobox(options<span class="op">=</span><span class="bu">list</span>(attribution_data_all_genes.columns[<span class="dv">1</span>:])))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="175">
<div class="sourceCode cell-code" id="cb212"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb212-1"><a href="#cb212-1" aria-hidden="true" tabindex="-1"></a>display(combobox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"6de302386d11478db9c9cc2bb80cf0f5","version_major":2,"version_minor":0}
</script>
</div>
</div>
<div class="cell" data-execution_count="184">
<div class="sourceCode cell-code" id="cb213"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb213-1"><a href="#cb213-1" aria-hidden="true" tabindex="-1"></a>selected_drug_name <span class="op">=</span> combobox.result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="183">
<div class="sourceCode cell-code" id="cb214"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb214-1"><a href="#cb214-1" aria-hidden="true" tabindex="-1"></a>selected_drug_attribution_gene <span class="op">=</span> (attribution_data_all_genes_var.loc[:,selected_drug_name])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="186">
<div class="sourceCode cell-code" id="cb215"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a>selected_drug_attribution_gene <span class="op">=</span> selected_drug_attribution_gene.sort_values( ascending<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="187">
<div class="sourceCode cell-code" id="cb216"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb216-1"><a href="#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.graph_objects <span class="im">as</span> go</span>
<span id="cb216-2"><a href="#cb216-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-3"><a href="#cb216-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming you have a list of gene scores for a given drug</span></span>
<span id="cb216-4"><a href="#cb216-4" aria-hidden="true" tabindex="-1"></a>gene_scores <span class="op">=</span> selected_drug_attribution_gene.values[<span class="dv">0</span>:<span class="dv">20</span>]</span>
<span id="cb216-5"><a href="#cb216-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-6"><a href="#cb216-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an array of x-coordinates</span></span>
<span id="cb216-7"><a href="#cb216-7" aria-hidden="true" tabindex="-1"></a>gene_names <span class="op">=</span> selected_drug_attribution_gene.index[<span class="dv">0</span>:<span class="dv">20</span>]</span>
<span id="cb216-8"><a href="#cb216-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-9"><a href="#cb216-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the color scale for the bars</span></span>
<span id="cb216-10"><a href="#cb216-10" aria-hidden="true" tabindex="-1"></a>color_scale <span class="op">=</span> <span class="st">'Blues'</span></span>
<span id="cb216-11"><a href="#cb216-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-12"><a href="#cb216-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a bar plot using Plotly and assign the color scale</span></span>
<span id="cb216-13"><a href="#cb216-13" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> go.Figure(data<span class="op">=</span>go.Bar(x<span class="op">=</span>gene_names, y<span class="op">=</span>gene_scores, marker<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span>gene_scores, colorscale<span class="op">=</span>color_scale)))</span>
<span id="cb216-14"><a href="#cb216-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-15"><a href="#cb216-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Customize the plot layout</span></span>
<span id="cb216-16"><a href="#cb216-16" aria-hidden="true" tabindex="-1"></a>fig.update_layout(</span>
<span id="cb216-17"><a href="#cb216-17" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"Top Gene Scores for "</span> <span class="op">+</span> selected_drug_name,</span>
<span id="cb216-18"><a href="#cb216-18" aria-hidden="true" tabindex="-1"></a>    xaxis_title<span class="op">=</span><span class="st">"Gene"</span>,</span>
<span id="cb216-19"><a href="#cb216-19" aria-hidden="true" tabindex="-1"></a>    yaxis_title<span class="op">=</span><span class="st">"Score"</span>,</span>
<span id="cb216-20"><a href="#cb216-20" aria-hidden="true" tabindex="-1"></a>    xaxis<span class="op">=</span><span class="bu">dict</span>(tickmode<span class="op">=</span><span class="st">'linear'</span>),</span>
<span id="cb216-21"><a href="#cb216-21" aria-hidden="true" tabindex="-1"></a>    yaxis<span class="op">=</span><span class="bu">dict</span>(<span class="bu">range</span><span class="op">=</span>[<span class="dv">0</span>, <span class="bu">max</span>(gene_scores)]),</span>
<span id="cb216-22"><a href="#cb216-22" aria-hidden="true" tabindex="-1"></a>    font<span class="op">=</span><span class="bu">dict</span>(family<span class="op">=</span><span class="st">"Roboto"</span>),</span>
<span id="cb216-23"><a href="#cb216-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># paper_bgcolor='rgba(0,0,0,0)',  # Transparent background</span></span>
<span id="cb216-24"><a href="#cb216-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot_bgcolor='rgba(0,0,0,0)'    # Transparent plot area</span></span>
<span id="cb216-25"><a href="#cb216-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb216-26"><a href="#cb216-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb216-27"><a href="#cb216-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the plot in a web-based environment</span></span>
<span id="cb216-28"><a href="#cb216-28" aria-hidden="true" tabindex="-1"></a>fig.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Unable to display output for mime type(s): application/vnd.plotly.v1+json</code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>